{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install openai"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install torch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pip install -r requirement.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "from dotenv import load_dotenv\n",
    "import os\n",
    "\n",
    "# Load environment variables from .env file\n",
    "load_dotenv()\n",
    "\n",
    "# Access the API key\n",
    "openai_api_key = os.getenv('OPEN_AI_API_KEY')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from dotenv import load_dotenv\n",
    "load_dotenv()\n",
    "OPEN_AI_API_KEY = os.getenv(\"OPEN_AI_API_KEY\")\n",
    "MODEL=\"gpt-3.5-turbo\"\n",
    "# MODEL=\"mixtral:8x7b\"\n",
    "MODEL=\"llama3.1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The quadratic formula is a mathematical formula that provides the solutions to a quadratic equation of the form:\n",
      "\n",
      "ax² + bx + c = 0\n",
      "\n",
      "where a, b, and c are constants, and x is the variable.\n",
      "\n",
      "The quadratic formula is:\n",
      "\n",
      "x = (-b ± √(b² - 4ac)) / 2a\n",
      "\n",
      "This formula works by first calculating the value inside the square root (called the discriminant), which is b² - 4ac. If this value is positive, there are two real solutions to the equation. If it's zero, there is one repeated solution. And if it's negative, there are no real solutions.\n",
      "\n",
      "The formula then takes the square root of this value and combines it with -b and the denominator (2a) to get the final values for x.\n",
      "\n",
      "Here's a breakdown of what each part of the formula does:\n",
      "\n",
      "* -b: This is the opposite of the coefficient of x, which accounts for the linear term.\n",
      "* √(b² - 4ac): This calculates the square root of the discriminant, which determines the number and nature of the solutions.\n",
      "* ± (plus/minus): This sign indicates that there are two possible values for x, depending on whether you take the positive or negative square root.\n",
      "* 2a: This is the denominator that scales the result to get the final solution.\n",
      "\n",
      "The quadratic formula can be derived by factoring a quadratic expression and then canceling out common factors. However, it's often easier to just memorize the formula and use it directly!\n"
     ]
    }
   ],
   "source": [
    "from langchain_openai.chat_models import ChatOpenAI\n",
    "from langchain_openai.embeddings import OpenAIEmbeddings\n",
    "from langchain_community.llms import Ollama\n",
    "from langchain_community.embeddings import OllamaEmbeddings\n",
    "\n",
    "if MODEL.startswith(\"gpt\"):\n",
    "    model=ChatOpenAI(api_key=OPEN_AI_API_KEY,model=MODEL)\n",
    "    embeddings=OpenAIEmbeddings()\n",
    "else:\n",
    "    model=Ollama(model=MODEL)\n",
    "    embeddings=OllamaEmbeddings()\n",
    "\n",
    "print(model.invoke(\"what is the quadratic formula?\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"Here's one:\\n\\nWhy did the cat join a band?\\n\\nBecause it wanted to be the purr-cussionist!\\n\\nI hope that one made you howl with laughter (or at least purred with amusement)!\""
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain_core.output_parsers import StrOutputParser\n",
    "parser = StrOutputParser()\n",
    "chain = model | parser\n",
    "chain.invoke(\"Tell me a joke about cats\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 0}, page_content='Ph. D Dissertation on  \\n \\n“DEVELOPMENT OF DISTINCT SNP GENOTYPIC \\nARRAYS FOR ASSOCIATION MAPPING OF \\nGERMPLASM AND BI -PARENTAL POPULATION \\nIN MULBERRY”  \\n \\n \\nSubmitted by  \\n \\nSINSHA P S  \\nUSN -6RV18MBT01  \\n \\nRESEARCH CENTRE  \\n \\n \\nDepartment of Biotechnology  \\nR.V. College of Engineering, Bengaluru -560059  \\n \\nUnder the Guidance of  \\nDr. VIDYA NIRANJAN  \\nProfessor and Head of Department - Biotechnology  \\nR.V. College of Engineering'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 1}, page_content='DECLARATION  \\n \\n \\n \\n \\n \\nI, SINSHA P S  bearing the USN 6RV18MBT01, hereby declare that the research work entitled \\n“DEVELOPMENT OF DISTINCT SNP GENOTYPIC ARRAYS FOR ASSOCIATION \\nMAPPING OF GERMPLASM AND BI -PARENTAL POPULATION IN MULBERRY ” carried \\nout by me has not previously formed the basis for the award of any degree or diploma or certificate \\nfrom any other university. The thesis is based on the individual original research work. The thesis  \\ndoes not contain any material that infringes the copyright of any individual or organization. The text, \\ntables, equations, figures, charts and graphs taken from sources such as research articles, books, \\nperiodicals, websites have been cited appropriately.  \\n \\n \\n \\n \\nSignature of the Research Scholar  \\n \\n \\n \\n \\n \\nSINSHA P S  \\nUSN: 6RV18MBT01  \\nDepartment of Biotechnology  \\nRVCE, Bengaluru – 560059.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 2}, page_content='3'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 3}, page_content='4  ACKNOWLEDGEMENTS  \\nWith profound appreciation and reverence, I would like to extend my heartfelt gratitude to my \\nPh.D. project advisor, Dr. Vidya Niranjan , Professor and Head of the Department of \\nBiotechnology and Centre of Excellence in Computational Genomics. Her invaluable assistance, \\nunwavering encouragement, and extensive expertise have been instrumental in guiding me \\nthroughout the entirety of my doct oral journey.  \\nI would also like to express my sincere acknowledgment to Dr. K N Subramanya , Principal of \\nR V College of Engineering, for fostering an environment conducive to the successful execution \\nof my research project. Furthermore, I am grateful to the teaching and non -teaching staff of the \\nDepartment of Biotechnology for their unwavering assistance and support in the laboratory, \\nwhich proved vital in the completion of my research work.  \\n   \\nI would like to express my sincere gratitude to my lab mates at Centre of Excellence Computational \\nGenomics for their unwavering support and assistance throughout the course of my doctoral research. \\nTheir collaboration, encouragement, and camaraderie have been invaluable to me during this journey.  \\nI am deeply thankful to Lavanya C, Spoorthi R Kulkarni, Shreya Satyanarayan Bhat, Akshay \\nUttarkar C, Anagha S Setlur, Chandrashekar K, S Pooja  for their willingness to share their \\nexpertise, provide constructive feedback, and engage in stimulating discussions. Their diverse \\nperspectives and dedication to excellence have enriched my research experience immeasurably.  \\nI am also grateful for the countless hours spent together in the lab, troubleshooting experiments, \\nanalyzing data, and brainstorming ideas. The sense of camaraderie and teamwork fostered within our \\nlab has been instrumental in overcoming challenges and achieving milestones.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 4}, page_content='5  I am immensely grateful to the Central Sericulture Research and Training Institute, Central \\nSilk-Board, Mysore , for their invaluable assistance. I would like to extend my sincere \\nappreciation to Dr. V. Sivaprasad , the project coordinator, as well as the other esteemed \\ninvestigators, Dr. S. Manthira Moorthy, Dr. K.P Arunkumar, Dr. Kusuma L, and Dr. R. \\nSumathy , for generously providing us with silkworms for RAD sequencing. Their support and \\ncontribution have been instrumental in the success of our research, a nd I am deeply thankful for \\ntheir assistance.  \\nWe express our sincere gratitude to the Department of Biotechnology  for their invaluable \\nsupport in facilitating the project titled \" Genetic Enhancement of Mulberry by Genomics \\nApproach: A Multi Component Network Project .\" Their unwavering commitment to \\nadvancing scientific research has been instrumental in driving our exploration towards innovative \\nsolutions in agricultural genomics.  \\n \\nFinally, I would like to express my gratitude to my family and friends for their unwavering \\nmoral support during the research work.  \\n \\n \\n \\nSINSHA P S  \\n6RV18MBT01'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 5}, page_content='6  ABSTRACT  \\nThis research presents a comprehensive analysis of four Mulberry genotypes, each harboring unique \\ntraits crucial for sericulture and broader agricultural applications. Through meticulous investigation, \\ngenetic variations, including single nucleotide polymorphisms (SNPs) and simple sequence repeats \\n(SSRs), were uncovered in the Thailand Male genotype, renowned for its high -quality leaves. These \\nfindings offer insights into the genetic factors associated with superior leaf yield, laying the groundwork \\nfor selective breeding to develop Mulberry cultivars with consi stently nutrient -rich leaves for \\nsilkworms. Additionally, the study elucidated the genetic basis for the resistance to root rot diseases in \\nthe Assam Bola genotype, originating from India, potentially facilitating the development of disease -\\nresistant Mulbe rry cultivars, thereby promoting environmentally sustainable practices. The research \\nexamined the drought -resistant S1 genotype, aiming to understand its exceptional resilience under \\nwater scarcity conditions, critical for ensuring a stable leaf supply ami dst increasing drought \\noccurrences due to climate change. Additionally, a comprehensive SNP and SSR database was \\ngenerated, serving as a valuable resource for future genetic studies in Mulberry. The study also \\ninvestigated effective nitrogen utilization in  Mulberry, utilizing the Punjab Local genotype, a vital \\naspect in enhancing agricultural productivity while minimizing environmental impact. Utilizing state -\\nof-the-art methods, such as SNP and SSR markers, a vast array of genetic data was generated, \\nencomp assing approximately 50 million SNPs. To streamline the data, consensus generation was \\nconducted, facilitating the development of diagnostic tools and genetic linkage maps. Moreover, \\nprimers were designed for chip development, enabling further genetic anal ysis and enhancement of \\nMulberry cultivars. Phenotype -genotype correlation studies were successfully achieved, providing \\nvaluable insights into the genetic mechanisms underlying key agronomic traits. Additionally, \\nphenotype -phenotype correlation studies fu rther enhanced our understanding of the complex \\ninteractions between genetic variations and phenotypic traits in Mulberry. Overall, this research \\ncontributes to the advancement of Mulberry agriculture by identifying genetic markers, elucidating \\ngenetic lin kages, and creating diagnostic tools, ultimately enhancing economic viability and \\nenvironmental responsibility in the sericulture industry.  \\n \\nKey words:  Morus indica, Markers, S ingle Nucleotide Polymorphism, Simple sequence repeats, \\nGenotype, Phenotype, Genetic mapping'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 6}, page_content='7  CHAPTER -WISE EXPLANATION  \\n \\nThis thesis is presented in nine different chapters which are as follows:  \\n \\nChapter -1 Introduction  \\nIntroduction summarizes the importance of Mulberry, Specific details of Morus indica , Economic \\nimportance, Medicinal value and the Phenotypic details of Morus indica  \\nChapter -2 Review of Literature  \\nPresents the previous research on related topics, the literature gap, steps that have current study, and \\nmotivation and scope of study.  \\nObjectives  \\nPresents the frame research objectives of the study.  \\nChapter -3 Research Methodology  \\nIncludes all the materials and methods that were used in the study to arrive at the expected outcomes.  \\nChapter -4 Results and Discussion  \\nPresents comprehensive results that were obtained from each methodology, including some value -\\nadded results. The results include the quality checks, adaptor trimming, alignment, variant calling, \\nanalysis of mutational profiles, data normalization, analysis  of unique variants, database development \\nand design of mulberry array with markers. Discussions for each result by comparing with previous \\nresults are also presented.  \\nChapter 5- Conclusions and future perspectives  \\nSummarizes the findings of this research work and outlines specific conclusions drawn from the \\nexperimental efforts. Further, the future scope of the present work is provided.  \\nChapter 6- Reference s \\nAll the reference has been cited'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 7}, page_content=\"8   \\nLIST OF PUBLICATIONS  \\n \\n \\n1)  Sinsha Vikhin,  Lavanya C, Shri Ganapathi V Raman, Anirudh R URS, Shashank Rao  \\nVidya Niranjan “COMPUTATIONAL TOOLS FOR UNVEILING MULBERRY'S GENETIC \\nARCHITECTURE THROUGH LINKAGE MAPPING: A COMPREHENSIVE REVIEW”, \\nGradiva Review Journal, ISSN NO: 0363 -8057, Volume 9, ISSUE 8, 2023, DOI: \\n10.37897.GRJ. 2022.V9I8.23.513311  \\n \\n2)  Sinsha Vikhin, Vidya Niranjan , Lavanya C, Spoorthi R Kulkarni, “Unlocking the Genetic \\nDiversity of Morus indica  through Molecular Marker Analysis”, JOURNAL OF \\nTECHNOLOGY, ISSN NO: 1012 -3407, VOL 13, ISSUE 10, DOI: 10.61350/v13 -105416\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 8}, page_content='9  Sl.No.  TITLE  PAGE  \\nNUMBER  \\nI. TITLE PAGE  i \\nII. CERTIFICATE  ii \\nIII. ACKNOWLEDGEMENT  iii \\nIV. ABSTRACT  iv \\nV. CHAPTER -WISE EXPLANATION  vi \\nVI. LIST OF PUBLICATIONS  vii \\nVII. TABLE OF CONTENTS  viii \\nVIII.  OVERVIEW OF THE STUDY  xv \\nChapter -1 \\n1.0 Introduction  16 \\n1.1 Taxonomy and major species  16 \\n1.2 Introduction to Morus indica  17 \\n1.3 Importance of feed in a sericulture  18 \\n1.4  Morus indica harvest in India  19 \\n1.5 Quantitative trait loci (QTL)  19 \\n1.6 Importance of Morus indica  20 \\n1.6.1  Medicinal importance  21 \\n1.6.2  Commercial importance of Morus indica  22 \\n1.6.3  Economic importance of Morus indica  22 \\n1.7 Breeding technologies of mulberry  24 \\n1.8 Genomic resources of mulberry  25 \\n1.9 Exploring the genome of Morus indica  27 \\n1.9.1  Whole genome sequence of Morus indica  27 \\n1.9.2  The repetitive content in a draft genome  28 \\n1.9.3  Genes that code for proteins  28 \\n1.9.4  Identification of single nucleotide polymorphism in Morus indica  \\n 28 \\n1.9.5  Simple sequence repeats  29 \\n1.10 Phenotypes of mulberry in the research  30 \\n1.10.1  Problems encountered by mulberry cultivators  30 \\n1.10.2  Drought stress  32'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 9}, page_content='10  1.10.3  Leaf yield  32 \\n1.10.4  Nitrogen use efficiency  33 \\n1.10.5  Root rot conditions  34 \\n1.11 Identification of biomarkers for the four phenotypes  35 \\n1.12 Genetic linkage and mapping studies  36 \\n2.0 Review of literature  37 \\n2.1 Polyploidy in plants  39 \\n2.2  Exploring polyploid genome in plants  39 \\n2.2.1  Methods of analyzing genome  39 \\n2.2.2  In situ hybridization  \\n 40 \\n2.2.3  Molecular marker -based genetic mapping  40 \\n2.2.4  Methylation sensitive molecular markers  40 \\n2.2.5  Comparative genome analysis  41 \\n2.2.6  High -Throughput DNA Sequencing and High -Resolution Melting \\n(HRM) Analysis  41 \\n2.3  Current strategies of Sequence alignment and SNP calling in \\nPolyploid plants  \\n 41 \\n2.4 Third generation sequencing  42 \\n2.5 Best practices in the identification of molecular markers  42 \\n2.5.1  \\n Define clear objectives  \\n42 \\n2.5.2  \\n Choose appropriate marker types  \\n42 \\n2.5.3  Utilize high quality genomic data  43 \\n2.5.4  Consider genetic diversity  43 \\n2.5.5  Validation of markers  43 \\n2.5.6  Evaluate marker transferability  43 \\n2.5.7  Quality control measures  43 \\n2.5.8  Consider genomic context  43 \\n2.5.9  Document methodology and parameters  44 \\n2.5.10  Community standards and resources  44 \\n2.5.11  Ethical consideration  44 \\n2.6  SNP Genotyping chip for genetic studies and Breeding 44'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 10}, page_content='11  applications  \\n \\n2.7  Prerequisites for a genotyping chip  45 \\n2.8  Linkage mapping studies  46 \\n2.9  Plant linkage mapping with genetic markers  48 \\n2.10 Computational requirements for linkage mapping studies  \\n 49 \\n2.11 Motivation  50 \\n2.12  Research Gap  52 \\n3.0 Research methodology  54 \\n3.1 Retrieval of raw mulberry data  55 \\n3.2 Raw data pre -processing  56 \\n3.3 Quality checks of the reads  57 \\n3.4 Gapped alignment  57 \\n3.5 Refining the alignments and conversion by SAM tools  57 \\n3.6 Variant processing  58 \\n3.7 Variant calling  59 \\n3.8 Variant filtration  59 \\n3.9 Identification of SSR Markers  60 \\n3.10 Database development  60 \\n3.10.1  Frontend development  61 \\n3.10.2  Next.js and Tailwind CSS  61 \\n3.10.3  Backend development  62 \\n3.10.4  Advantages  62 \\n3.10.5  Hosting platform  62 \\n3.10.6  Database management  62 \\n3.10.7  Database Structure and Attributes  62 \\n3.10.8  Database Size and Performance  63 \\n3.10.9  Advantages of mysql  on Aiven Cloud  63 \\n3.10.10  Hosting platform  64 \\n3.11  Consensus generation  \\n 64 \\n3.12 Designing the primer sequences  66'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 11}, page_content='12  3.13 Linkage groups  67 \\n3.14 Chip development  67 \\n4.0 Results and Discussion  68 \\n4.1  Raw Data Pre -processing  69 \\n4.2 Quality Checks and Reads  69 \\n4.3 Gapped alignment  71 \\n4.4 Sam to Bam conversion  71 \\n4.5 Variant processing  72 \\n4.6 Variant calling  72 \\n4.7 Identifying SSRs  73 \\n4.8  Database development  74 \\n4.9 Consensus generation  75 \\n4.10 Primer design  75 \\n4.11 Genotype/phenotype relation  76 \\n4.12 Phenotype/phenotype relation  78 \\n4.13 Genotyping chip generation  79 \\n4.14 Utilizing Genotype -Phenotype Correlation and Specific snps for \\nInformed Breeding Techniques in Agriculture  80 \\n5.0  Conclusion and Future Aspects  82 \\n5.1  Conclusion  82 \\n5.2 Future aspects  83 \\n6.0 References  85'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 12}, page_content='13  LIST OF FIGURES  \\n \\nFigure Number  Title  Page Number  \\n Overview of the Study  xv \\nFigure -1: Nutritional content of the mulberry leaves  20 \\nFigure 2  Medi cinal importance of Mulberry  21 \\nFigure 3  Mulberry production in India  24 \\nFigure 4  Genome i nformation of Morus indica  27 \\nFigure 5  Challenges face d by mulberry cultivators  31 \\nFigure 6  Requirements for designing a genotype chip  46 \\nFigure 7  Overview of the methodology followed  54 \\nFigure 8  Protocol for mapping and SNP Calling  58 \\nFigure 9  Best Practices  for GATK pipelin e 59 \\nFigure 10  The query flow for Fetching data  61 \\nFigure 11  The flow for database development  64 \\nFigure 12  Protocol for consensus development  66 \\nFigure 13  Quality control for sequenced data  68 \\nFigure 14  Sequence mean quality score  69 \\nFigure 15  Sequence counts  70'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 13}, page_content='14  Figure 16  Graph for sample types and corresponding SNPs  72 \\nFigure 17  Frequency count of SSRs  73 \\nFigure 18  Top SSR Motifs  74 \\nFigure 19  Contig file after consen sus generation  75 \\nFigure 20  Primers designed  76 \\nFigure 21  Genotype -Phenotype association  77 \\nFigure 22  Correlation matrix of agric ultura l factors  78 \\nFigure 23  Marker -Trait association or linkage mapping  81 \\n \\n \\nLIST OF TABLES  \\n \\nTable Number  Title  Page Number  \\nTable -1: The genotype details  55 \\nTable 2  Phenotype details of the sample collected  56 \\nTable 3  Genotype and alignment percentages  71'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 14}, page_content='15   \\nOVERVIEW OF THE STUDY'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 15}, page_content='16  CHAPTER 1  \\n1.0 INTRODUCTION  \\nThe mulberry, belonging to the Moraceae  genus, is a fast -growing, woody, hardy perennial plant. It has \\nseveral crucial qualities, including a larger leaf output, a shorter gestation time, and better environmental \\ntolerance. More than 15 species of the genus are mostly found in temperate, tropic al, and subtropical \\nclimates in Asia, Africa, and North America. The following species: Morus alba, nigra, rubra, and \\naustralis  More than 15 species of the genus are mostly found in temperate, tropical, and subtropica l \\nclimates in Asia, Africa, and North America. Among these, Morus alba, Morus nigra, Morus rubra, \\nand Morus australis,  Morus atropurpurea , Morus cathayana , and Morus notabilis  are some of the more \\nnotable ones [1].  \\n1.1. TAXONOMY AND MAJOR SPECIES  \\nThe fruits, which mature to white, pink, red, or purple, resemble blackberries in several ways. \\nIndividuals can be dioecious or monoecious, having both male and female flowers (bearing only male \\nor female flowers) One metric ton of mulberry leaves is thoug ht to be required for the raising of \\nsilkworms that emerge from one case of eggs and produce roughly 25 kg to 30 kg of high -quality \\ncocoons. Various species of mulberry, including Morus indica and Morus mesosygia , among others, \\nare renowned for their econo mic importance. Mulberry leaves are commonly utilized to nourish Bombyx \\nmori silkworms, essential for silk production. Through a process wherein the silkworm converts \\nmulberry leaf protein into sericin and fibroin, the foundation of silk is established. This silk is then \\ncommercially utilized in the creation of fine clothing. Globa lly, mulberry silk constitutes approximately \\n90% of all raw silk production, significantly benefiting countless individuals [2]. Additionally, the \\nmulberry plant offers various nu tritional and therapeutic advantages. Many nations have historically \\nemployed mulberry leaves and fruits, abundant in protein and vitamins, as both animal feed and human \\nsustenance. Mulberry cultivation primarily supports silkworm production, serving as th e exclusive food \\nsource during their larval stage. Therefore, maintaining a robust production of mulberry leaves \\naccording to scientific standards is crucial for sustainable sericulture. Mulberry trees are deciduous, \\nfeaturing alternately positioned, tooth ed, and occasionally lobed leaves. Clusters of small blooms, called \\ncatkins, adorn the branches, each developing into a fruit cluster known as a \"multiple.\" These fruits, \\nranging in color from white to purple, bear resemblance to blackberries. Mulberry tre es can exhibit \\ndioecious or monoecious characteristics, with some possessing separate male and female flowers while \\nothers have both on the same plant. It\\'s estimated that one metric ton of mulberry leaves is necessary to \\nrear silkworms hatching from a sin gle batch of eggs, yielding approximately 25 kg to 30 kg of premium -\\nquality cocoons. The genus Morus has 68 species, the majority of which are found in Asia [3 -5]. Over'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 16}, page_content=\"17  a thousand different types are grown in China. Descended from four primary species —Guangdong \\nMulberry (Morus atropurpurea ), Mountain Mulberry ( Morus bombycis ), White Mulberry ( Morus \\nalba), and Lu Mulberry ( Morus multicaulis )—mulberry varieties exhibit diversity. In India, prominent \\nspecies include M. indica, M. alba, M. serrata, and M. laevigata , predominantly found in the northern \\nregions. Most cultivated types belong to either M. indica or M. alba . The Central Sericulture Research \\nand Training Institute i n Mysore, India, maintains a collection of 223 mulberry cultivars, encompassing \\nindigenous, alien, unidentified, and top hybrid varieties. Across the former Soviet Union's Republics, \\nM. alba and M. indica are widely recognized. Through selective breeding m ethods such as open \\npollination, controlled hybridization, selection, and mutation breeding, numerous variations —many of \\nwhich are polyploid —have been developed, surpassing a thousand in some cases. In Brazil alone, there \\nare approximately 90 distinct M. a lba variants [6,7].  \\n1.2 INTRODUCTION TO MORUS INDICA  \\nMorus indica , belonging to the Morus genus within the Moraceae  family of flowering plants, is widely \\nrecognized as part of the mulberry family. Morus indica is a deciduous tree. Indigenous to the temperate \\nand subtropical Himalayan region, it is now grown in various regions including India, China, Japan, \\nand East Africa [8] Morus indica , as together with more mulberry family members, is frequently \\ncharacterized as either a small tree or a shrub,  rarely exceeding 10 –15 meters (33 –49 feet). When young, \\nthe br anches are covered in fine, soft hairs known as; however, the plant loses these hair as they grow \\nolder. The branches are a light gray -brown color. The leaves range from 4 –12.5 centimeters (1+1⁄2 –5 \\ninches) long and 2.5 –7.5 cm (1 –3 in) wide, and are attache d to the tree via petioles. The leaves \\nthemselves are usually ovate but sometimes lobed, coming to a narrow point, making them somewhere \\nbetween caudate and acuminate. The leaves are retuse to slightly cordate, having a small lobe at the \\nbase. They are sho rtly serrated, with each tooth narrowing to a thin point, making them apiculate. The \\nleaf color is dark green, with a paler underside covered in fine hairs.   Morus indica is a monoecious \\nflowering plant, having male and female flowers growing on the same tree, although often on distinct \\nbranches [8]. The male inflorescence is narrow, between 9 –11.5 millimeters (3⁄8 –7⁄16 in) long, and \\ncovered in fine hairs. The female flower s are subglobose, or just shy of spherical. They measure 6 –9.5 \\nmm (1⁄4 –3⁄8 in) long. The stigma of these flowers is about 3.5 mm long with dense, short hair. The \\nfemale flower, after being fertilized, forms a fleshy compound fruit known as a syncarp. This syncarp, \\nwhich is black when fully ripe, looks like that of Morus nigra, commonly kno wn as a black mulberry. \\noccupies a pivotal position in the realm of sericulture, embodying an intricate interplay of biological, \\neconomic, ecological, and cultural significance. Morus indica leaves serve as the main source of \\nsustenance for silkworms ( Bomb yx mori ) throughout their life stages. These leaves provide a balanced \\nmix of proteins, carbohydrates, vitamins, and minerals, which cater to the specific dietary requirements\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 17}, page_content=\"18  of silkworm larvae [9]. This nutritional blend supports their robust growth and development and \\nsignificantly impacts the silk's quality and quantity. The symbiotic partnership between Morus indica \\nand silkworms has nurtured centuries of traditional wisdom, fostering a thriving sericulture industry \\nthat contributes substantially to regional economies. Beyond its economic impact, Morus indica \\ncatalyzes rigorous scientific research, driving innovations in cultivation techniques, leaf quality \\nenhancement, and disease resistance. Moreover, this interdependence extends to ecological \\nconsiderations, where the plant's innate disease -resistant compounds contribute to silkworm health and \\nrobustness. Cultivating Morus indica aligns with sustainability principles, utilizing a renewable plant \\nresource, and carries cultural and historical weight in regions with deep sericultural roots. Ultimately, \\nMorus indica’s role transcends its botanical identity, weaving a tapestry of tradition, innovation, and \\necological harmony tha t is intrinsic to the intricate fabric of sericulture.  \\n1.3 IMPORTANCE OF FEED IN SERICULTURE  \\nThe specialized insect Bombyx mori predominantly consumes the rapidly growing perennial plant \\nMorus indica (L.), commonly known as Indian mulberry and belonging to the Moraceae  family. This \\nplant holds significant importance in sericulture [10]. Feed plays a crucial role in sericulture by directly \\nimpacting the growth, development, and silk quality produced by silkworms. Mulberry leaves, the \\nprimary food source for silkworms, ma y result in decreased silk production if certain nutrients are \\nlacking. The nutr itional value of these leaves is influenced by various agro climatic factors, and any \\ndeficiency in nutrients hampers the silk production of silkworms. A diet rich in nutrients promotes the \\noptimal growth and development of silkworm larvae, as well as the production of eggs. To meet the \\nnutritional demands of numerous insects, supplementary sources of nutrients are necessary. This \\nfacilitates the evaluation of different fortification agents' significance and effects on silkworm nutrition. \\nRecent efforts hav e been made to enhance both the quantity and quality of silk through various \\ninitiatives. These initiatives include nutrient supplementation, antibiotic spraying, application of \\njuvenile hormone (JH), plant -based products, steroids, JH -mimicking techniques , and the use of plant \\nextracts. In an endeavor to enhance both the quantity and quality of silk, mulberry leaves have been \\nenriched with various nutrients for silkworm consumption. The practice of fortifying and supplementing \\nmulberry leaves is a relative ly recent advancement in sericulture research. The sericulture industry has \\nstrategically utilized nutritional interventions to optimize the functioning of silk glands, thereby \\nimpacting economic performance and silk production [11]. Mulberry leaves are fo rtified with an array \\nof minerals, antibiotics, plant extracts, and hormones to augment their nutritional profile. As the \\nexclusive food source for silkworm caterpillars, mulberry holds paramount importance in sericulture \\n[12]. Sericulture represents a piv otal sector in agriculture, deeply interconnected with the cultivation of \\nmulberry and the production of silk [13]. Factors such as intercropping with pulses, integrated pest and\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 18}, page_content=\"19  disease management, and external variables like the COVID -19 pandemic collectively influence \\nmulberry yield [14 -16]. Notably, mulberry is often interplanted with pulses, particularly chickpeas.  \\n 1.4 MORUS INDICA HARVEST IN INDIA  \\nMulberry ( Morus indica L.) is a foliage crop that is commonly cultivated in North -Eastern India with \\nthe goal of growing silkworms ( Bombyx mori L.) in rainfed areas. [17 -20]. Abiotic factors that can \\ndrastically reduce mulberry output include dryness, salt, and high temperatures. The species Morus \\nindica is particularly noteworthy in the context of generating high yield and quality mulberry leaves in \\nacid soil of North Eastern India. Mulberries are the sole crop farmed in many regions of India, and an \\nanalysis of the soil fertility there indicated acidic s oils with varying levels of organic carbon and nutrient \\nshortag es [21,22]. Given that the genome of the domesticated Indian mulberry ( Morus indica cv. K2) \\nhas been sequenced, comprehending repetitive DNA, protein -coding genes, and the study [23] \\nhighlighted the widespread nutritional inadequacies in the upper Brahmaputra area of northeastern \\nIndia's soils. Agronomic traits include good propagation, e arly sprouting, rapid growth, high biomass \\noutput, wide adaptability to varied agro -climatic conditions, sensitivity to organic and inorganic \\nfertilizers, and resistance to  diseases and pests are important mulberry breeding needs. In other words, \\nthe most important factors are the silkworm's taste and the capacity to produce excellent cocoons. Thus, \\nit is crucial to exploit the mulberry genetic resources that are presently a vailable to develop mulberry \\nvarieties that are suitable for certain soil types prevalent in different agro -climatic zones. Mulberry \\nfarming has been practiced in India for a very long time, going all the way back to the beginning of \\ntime. Chinese traders and travelers contributed the talent [24,25]  \\n1.5 NUTRITIONAL COMPOSITION ON MULBERRY MORUS INDICA  \\nA single mulberry fruit typically weighs between 2.14 to 4.07 grams. Among its therapeutic components \\nare natural and amino sugars [26]. Mulberry fruit contains various polysaccharides that significantly \\ninfluence human physiology [27,28]. These fruit -derived polysaccharides have been associated with \\nseveral beneficial bioactivities, including antioxidant, hypoglycemic [29,30], anti -obesity [31], anti -\\ninflammatory, and anti -apoptotic effects [32,33]. Recent research has revealed that mulberry fruit \\ncontains  nearly all essential amino acids. Spectroscopic techniques have identified six novel forms of \\nmorusimic acid, labeled A, B, C, D, E, and F which are branched -chain amino acids known to accelerate \\nmuscle growth [34]. Studies indicate that mulberry fruits c ontain 7.55% saturated lipids and 87.5% \\nunsaturated fatty acids, with linoleic acid comprising the majority (79.4%) and palmitic acid following \\nat 8.6% [35]. Turkish mulberry fruits exhibit a total lipid content of 57.3%, with palmitic acid \\nconstituting 22 .4%. Mulberry fruits contain a higher proportion of polyunsaturated fatty acids compared \\nto monounsaturated or saturated fatty acids [36]. In comparison to strawberries (72%), mulberries (76%) \\nand ziziphus jujuba (68 –72%) have a greater range of polyunsatu rated fatty acids [37,38]. Linoleic acid,\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 19}, page_content='20  a key polyunsaturated fatty acid abundant in mulberry fruit, plays a critical role in human growth, health \\npromotion, and disease prevention [39,40].   \\n \\nFigure 1: Nutritional Content of Mulberry Leaves  \\n \\n1.6 IMPORTANCE OF MORUS INDICA  \\n1.6.1 MEDICINAL IMPORTANCE  \\nThe fast -growing perennial plant Morus indica (L.), commonly known as Indian mulberry, belongs to \\nthe Moraceae  family and serves as the primary food source for the monophagous insect Bombyx mori , \\nmaking it highly relevant in sericulture [41,42]. Ancient Ayurvedic texts document its therapeutic \\nattributes, utilizing its fruits, leaves, roots, bark, and latex in diverse formulations aimed at addressing \\nvarious ailments. The presence of a plethora o f bioactive compounds, including steroids, phenolics, \\nalkaloids, reducing sugars, tannins, flavonoids, antho cyanins, terpenoids, cardiac glycosides, saponins, \\nproteins, amino acids, and carbohydrates, underscores its traditional use as a natural remedy for \\nantibacterial, antifungal, anti -hyperglycemic, and anti -hypolipidemic conditions.  \\n  \\nRecent studies, supported by citations [43] through [44], have unveiled the antibacterial prowess of'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 20}, page_content='21  Morus indica . Compounds like Kuwanon C and G, identified in the root bark extract, exhibit notable \\nantibacterial activity. Particularly strong antibacterial effects are observed in hydro methanolic stem \\nbark extracts, suggesting their potential as effective antimicrob ial agents against gastrointestinal \\nillnesses [45].  \\n \\nFigure 2: Medicinal Importance of Mulberry  \\n \\nMoreover, Morus indica leaves have been found to possess antibacterial properties effective against \\noral microbes [46]. Phytochemicals within Morus indica , such as phenolic compounds, flavonoids, and \\nanthocyanins, contribute to its antimicrobial capabilities. Further research, as detailed in citations [47] \\nthrough [48], has demonstrated Morus indica’s effectiveness against various bacterial species, including \\noral pathogens. Its ability to impede DNA replication, metabolic processes, and disrupt cell membranes \\ncontributes to this  efficacy. Additionally, its antibacterial properties are augmented by its capacity to \\nneutralize free radicals and alleviate oxidative stress, thereby mitigating damage to bacterial cells. \\nMoreover, Morus indica exhibits anti -inflammatory properties due to the presence of active compounds \\nsuch as phenolic compounds, flavonoids, and anthocyanins. These compounds contribute to its anti -\\ninflammatory effects, offering potential applications in combating various inflam matory conditions.  \\n Inhibition of inflammat ory mediators: The active ingredients in Morus indica can decrease \\ninflammation by preventing the generation of inflammatory mediators including prostaglandins and \\ncytokines [49]. Free radicals can produce oxidative stress, which can lead to inflammation. Morus \\nindica possesses antioxidant activity that can help avoid this.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 21}, page_content=\"22  1.6.2 COMMERCIAL IMPORTANCE OF MORUS INDICA  \\nThe commercial value of Indian mulberry, also known as Morus indica , is significant, particularly \\nwithin the sericulture sector where it serves as the primary food source for the monophagous insect \\nBombyx mori [50]. Mulberry's impact extends to the rural economy [51]. Morus indica fruit shows \\npromise for inclusion in food compositions due to its abundance in bioactive compounds and its anti -\\nquorum sensing properties [52]. Additionally, advancements such as an effective in vitro regeneration \\nprocess enable i ts potential for genome editing and genetic modification [53]. The acquisition of a high -\\nquality genome sequence for Morus indica provides a comprehensive resource for genomic research \\nand agricultural advancement. The bioactive components of Indian mulberry have been harnessed for \\ntheir medicinal properties, suggesting potential applications as nutraceuticals for various health \\nconc erns. Moreover, mulberry finds its place in numerous cosmetic products and functional foods [54 -\\n56]. The plant's significance extends to the food and pharmaceutical industries [57]. The potent \\nantioxidant activity observed in mulberries hints at their potential therapeutic applications, including in \\nconditions like cancer and cardiovascular disease. They also show promise in prev enting fatty liver \\ndisease and managing disorders linked to hyperlipidemia by aiding in lipid reduction [58]. Morus \\nindica’s commercial importance within the sericulture sector remains paramount, given Bombyx mori’s \\nheavy reliance on it. The demand for pro ducts derived from mulberries, boasting beneficial biological \\neffects, underscores its significance in various industries. Mulberry plays a crucial role in the global \\neconomy and holds multifaceted importance across diverse sectors [53 -57]. \\n1.6.3 ECONOMIC IMPORTANCE OF MULBERRY ( MORUS INDICA ) IN \\nINDIA  \\nMulberry ( Morus indica ) is a cornerstone of India's economic landscape, playing a crucial role across \\nsectors such as sericulture, agriculture, horticulture, and traditional medicine. Its multifaceted \\ncontributions sustain livelihoods, foster income generation, drive exports, a nd promote sustainable \\nagricultural practices, establishing it as an indispensable asset for economic growth and rural \\ndevelopment in the country.  \\n The sericulture industry, deeply entrenched in Indian tradition, relies heavily on mulberry leaves as the \\nprimary source of nutrition for silkworms. This sector not only offers employment opportunities to rural \\ncommunities but also holds significant sway over the nation's economy. India ranks among the largest \\nsilk producers globally, with states like Karnataka, Andhra Pradesh, and Tamil Nadu emerging as major \\nsilk-producing regions. The sericulture industry, encompassing silk farming, cocoon production, a nd \\nsilk processing units, generates substantial income for a significant portion of the population [58].  \\n Beyond silk production, mulberry trees play a vital role in Indian agriculture, benefiting sectors like \\ntraditional medicine, horticulture, and livestock farming. With its diverse contributions supporting\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 22}, page_content=\"23  sustainable agricultural practices, export growth, income generation, and livelihoods, Morus indica \\nholds a pivotal and multifaceted economic function in India. The nation's economy heavily relies on its \\ncontributions across various industries, enhancing the quality of life for millions of people. The \\nsericulture sector, a longstanding and successful ind ustry in India, predominantly feeds silkworms on \\nmulberry leaves, contributing significantly to the nation's economy while offering employment \\nprospects in rural areas. Mulberry leaves, nutrient -rich fodder, also benefit the dairy and animal \\nhusbandry industries. Moreover, mulberry cultivation aids in sustainable farming methods by enhancing \\nsoil fertility and reducing erosion. Practices like mulberry crop ro tation further bolster overall \\nagricultural production, further solidifying its importance in India's economic landscape.  \\nAgricultural Significance: Some mulberry species are farmed for their juicy, flavorful fruits in India. In \\naddition to being eaten raw, these fruits are often used to produce jams, jellies, and confections. Selling \\nmulberry fruits is one way for farmers in  mulberry -growing regions like Punjab and Haryana to augment \\ntheir revenue. The nursery and landscaping industries gain from the ornamental attractiveness of \\nmulberry trees [59].  \\n Traditional medicine and herbal remedies: The leaves, bark, and fruits of the mulberry tree are \\nemployed for their curative characteristics in Indian traditional medicine systems like Ayurveda. Due \\nto the expanding demand for natural and conventional trea tments, the manufacturing and marketing of \\nherbal medications and nutritional supplements containing components obtained from mulberries has \\ngrown significantly in economic importance. The mulberry tree has a significant economic impact on \\nIndia. Employmen t possibilities are created as a result, particularly in rural regions where sericulture is \\npracticed extensively. Along with silk production, other income sources include selling mulberry fruits, \\nraising cattle, and expanding herbal medicine market. India 's involvement in the international silk \\nmarket helps the country earn foreign currency and expand its economy [60].\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 23}, page_content='24   \\nFigure 3: Mulberry Production and Silk Industry in India  \\n \\n1.7 BREEDING TECHNOLOGIES OF MULBERRY  \\nMulberry breeding methods have evolved significantly, encompassing various techniques to enhance \\nthe yield, quality, and resilience of this versatile plant. It takes many years (approximately 15 -20) to \\ndevelop a new variety of mulberry because it is a pere nnial woody plant. Breeding targets should be set \\nwith a long -term view. To date, breeding targets have been, for example, high yield, high nutritional \\nvalue and resistance against diseases and pests. But today, new targets have been added to cope with \\nchanges in the sericultural system, such as large numbers of silkworm reared and adaptability to densely \\nplanted fields suitable for mechanical harvesting. Crossing is the major breeding method adopted for \\nthe development of new mulberry varieties. The aim wa s to develop a variety with high quality, high \\nyield and resistance Conventional breeding involves hybridization and selection, where different \\nmulberry varieties are crossed to create offspring with desired traits. These offspring are rigorously \\nscreened and bred further to establish new varieties with improved characteristics [61]. Mutagenesis \\nexposes mulberry plants to mutagenic agents, like radiation or chemicals, to induce DNA mutations. \\nSubsequent screening identifies plants with favorable traits for further breeding. Molecular breeding'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 24}, page_content='25  relies on molecular markers to select plants based on their DNA, rather than physical traits, aiding in \\noptimizing mulberry leaf yield and quality.[59] Plant tissue culture fosters disease -free and trait -specific \\nplant propagation in a sterile environment,  facilitating mass production of desired mulberry traits. \\nGenetic engineering manipulates mulberry DNA to enhance leaf yield and quality, as well as abiotic \\nstress tolerance [60]. Collectively, these breeding technologies illustrate the promising future of  \\nmulberry farming, driven by continuous advancements in breeding techniques [62].  Mulberry \\npropagation is generally carried out by grafting and by cutting methods. Root grafting prevails because \\nit is easy to handle and the grafted saplings have a high survivability. The cutting method can be with \\nhard wood (using the branches grown in  the previous year) and soft wood (using the spring sprouted \\nshoots). With mulberry varieties of poor rooting ability, treatment with plant hormones is advised to \\nstimulate ro oting. Recently, tissue culture derived saplings have also been produced [ 63]. \\n1.8 GENOMIC RESOURCES OF MULBERRY  \\nThe need for mulberry genetic resources has been acknowledged in the expansion of the mulberry \\nindustry and sustainable silk production [64,65]. Research endeavors have focused on assessing genetic \\ndiversity and variation among mulberry genotypes to enhanc e leaf productivity and other desirable \\ntraits. Modern methodologies like inter simple sequence repeat (ISSR) analysis have been employed to \\nidentify genetic variations among mulberry genotypes from diverse regions [60]. Preservation and \\nutilization of wil d mulberry genetic resources, which harbor genes for crucial traits like drought and \\nsalt resistance, have been emphasized [66].  \\n Utilizing biotechnological approaches such as molecular markers, cloning of essential functional genes, \\nand transgenic experiments, the genetic resources of mulberry have been explored, characterized, \\nconserved, and utilized. These efforts significantly c ontribute to advancing knowledge and leveraging \\nmulberry genetic resources for sustainable sericulture and crop improvement. Given its significance as \\na major food source for silkworms, mulberry (Morus spp.) holds considerable importance in India [67]. \\nRecent interest in mulberry genomics stems from its potential to offer valuable resources for functional \\nand translational genomics. The high -quality draft genome sequence of the Indian mulberry cultivar K2 \\n(Morus indica ), with significant repetitive content, stands out as a notable genetic resource for \\nmulberries in India [68]. This genome serves as a crucial asset for research in functional and \\ntranslational genomics. Advances in mulberry genetics and genomics have been  facilitated by \\nsequencing the draft ge nome of Morus notabilis  and various transcriptome and genomic resources. \\nStudies on mulberry genome size, genetic diversity, and phenotypic variation have provided insights \\ninto genetic differences and their impact on trait plasticity [69].  \\nThe synthesis of the first T2T gap -free reference genome of mulberry, enabled by integrating high -\\ncoverage, precise long -read sequences, offers a vast resource for exploring the structure and'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 25}, page_content=\"26  development of polycentric chromosomes. Furthermore, ISSR research has contributed to assessing \\ngenetic divergence across mulberry cultivars and variations in India, supporting effective conservation \\nand utilization of mulberry genetic resources [70]. Rese arch on Arabidopsis SHN1 expression in Indian \\nMulberry ( Morus indica L.) has furnished insights into genetic engineering techniques aimed at \\nenhancing leaf surface wax content and reducing post -harvest water loss.  \\n In summary, India's rich and invaluable mulberry genomic resources present opportunities for in -depth \\nexploration of mulberry genetics and genomes, promising significant contributions to functional and \\ntranslational genomics efforts [71]. Draft genomic se quence of Morus indica  \\nThe Indian mulberry's ( Morus indica ) draught genome sequence is a complete tool for research in \\nfunctional and translational genomics. Multiple methods were used to generate the genome sequence, \\nwhich produced a high -quality sequence with a gene completeness of 96.5 percent. [72] The genome  \\nstudy of M. indica revealed many protein -coding genes, repetitive DNA sequences, and species -specific \\ngenes [73]. Understanding the biological processes in mulberry was made possible by transcriptome \\nprofiling, which ide ntified tissue -specific and differential gene expression [74]. Morus notabilis, a \\nspecies of wild mulberry, had its genome sequence examined as well, and a web -based database called \\nMorusDB was created to make it simple to access mulberry genetic data [75] . The discovery of growth -\\nregulating factors (GRFs) in mulberries advances our knowledge of how this significant crop grows and \\ndevelops [76]. Earlier work on the whole -genome sequencing of Morus indica was completed by the \\nSchool of Computational & Integr ative Sciences at Jawaharlal Nehru University in New Delhi. This \\nserved as the foundation for our efforts,  with over 90% of them being confirmed by transcript data, the \\ngenome sequence analysis found 27,435 high -confidence protein -coding genes and 49.2% of repetitive \\nDNA. 4.8% of the genes in the M. indica genome were found to be species -specific when compared  to \\nother plant genomes. Across numerous accessions, transcriptome profiling revealed tissue -specific and \\nvariable expression of protein -coding genes,  with around 4.7 percent and 2 –5 percent of genes \\nimplicated in various biological activities. Following the whole genome resequencing of 21 \\naccessions/species, 2.5 million single nucleotide polymorphisms and 0.2 million insertions/deletions \\nwere found. Th e information and findings from this study served as a thorough resource for our \\nmulberry genomes research and its advancement [77].\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 26}, page_content='27  1.9 EXPLORING THE GENOME OF MORUS INDICA  \\n1.9.1 WHOLE GENOME SEQUENCE OF MULBERRY MORUS INDICA  \\n27,435 high -confidence protein -coding genes were found in the genome sequencing study, with 49.2% \\nof them having repetitive DNA, and more than 90% of them having transcript evidence. The M. indica \\ngenome contains 4.8 percent of species -specific genes, acco rding to comparison with other plant \\ngenomes. Across numerous accessions, transcriptome profiling revealed tissue -specific and variable \\nexpression of protein -coding genes, with around 4.7 percent and 2 –5 percent of genes implicated in \\nvarious biological ac tivities. A total of 2.5 million single nucleotide polymorphisms and 0.2 million \\ninsertions/deletions were found during whole genome resequencing of 21 accessions/species. Our \\ngenomics research on mulberry and its advancement benefited greatly from the inf ormation and findings \\nfrom this work. [77]  \\n \\nFigure 4: Genome information for Morus indica  \\n \\n1.9.2 THE REPETITIVE CONTENT IN A DRAFT GENOME  \\nIn many biological processes and genome evolution, repetitive sections of the genome are significant. \\nThey can influence structural variation, genomic stability, and gene regulation. The repeated sections \\nin the Indian mulberry genome study were found usin g a mix of data from many distinct methods, \\nincluding Illumina sequencing, single -molecule real -time sequencing, chromosomal conformation \\ncapture, and optical mapping. Repetitive DNA makes about 49.2% of the Indian mulberry genome, \\naccording to the researc h. It is crucial to identify repeating sections because they can provide light on'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 27}, page_content=\"28  how the genome is organized and structured and may also include functional components. The \\nannotation and characterization of genes, as well as the investigation of genome evolution and genetic \\ndiversity, can all benefit from an understanding of the repeti tive areas. The Indian mulberry genome \\nhas repeated sections, which is relevant information for future functional and translational genomics \\nstudies of this significant agricultural plant [78].  \\n1.9.3 GENES THAT CODE FOR PROTEINS  \\nIn a number of investigations, the farmed Indian mulberry Morus indica’s protein -coding areas have \\nbeen found and described. 27,435 high -confidence protein -coding genes were found in the M. indica \\ngenome, and transcript evidence backed up more than 90% of them [79]. These genes are involved in \\ntissue -specific and varied expression across several accessions, and they are essential for a variety of \\nbiological functions. Furthermore, a cDNA library made from M. indica revealed a significant number \\nof genes encoding  enzymes involved in diverse secondary metabolites, opening up new study directions \\nfor mulberry natural product chemistry. The role that protein -coding regions play in plant stress \\nresponses and cellular metabolism makes them significant. The research Cen tre of Jawaharlal Nehru \\nUniversity carried out [80]. Using the repeat -masked genome sequence, a mix of de novo prediction, \\nsimilarity -based searches, and evidence -based searches were used to identify the protein coding regions \\nin the Indian mulberry genome . 39 Mb (7.7%) of the genome's sequence included 27,435 unique genes \\nthat the investigation anticipated. Multiple transcript isoforms were predicted for at least 1546 genes \\n[81]. The predicted genes' average mRNA length was 1430 bp. With an average exon co unt of 6, almost \\n87 percent of the genes were expected to be multi -exonic. Based on similarity searches in several \\ndatabases, functional annotation of predicted genes was carried out. In all, 27,138 genes (98.9%) \\nshowed notable sequence similarity with pro teins included in the UniProt Viridiplantae database. \\nAdditionally, it was discovered that 24,939 (90.9%) of the genes included at least one conserved Pfam \\ndomain. Through whole -genome resequencing of 21 mulberry accessions/species, it was possible to \\nidentify DNA polymorphisms and create a high -density SNP map. This found 2.5 million single \\nnucleotide polymorphisms and 0.2 million insertions/deletions [82].  \\n1.9.4 IDENTIFICATION OF SINGLE NUCLEOTIDE POLYMORPHISM IN \\nMORUS INDICA  \\nThe draught genome contains comprehensive information on the SNPs found in Morus indica . To find \\nSNPs in Morus indica , the whole genomes of 21 mulberry accessions/species were sequenced. High -\\nquality reads were filtered out of the raw data, representing extremely high coverage of the mulberry \\ngenome for each genotype. [83] A total of 95.3 -99.7% of high -quality reads were  successfully mapped \\nto the M. indica draught genome assembly. 2,480,073 high -quality SNPs and 2,14,704 insertions and \\ndeletions (InDels) w ere found in total. SnpEff was used to annotate DNA polymorphisms in coding\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 28}, page_content='29  areas as synonymous, non -synonymous, and large -effect (v4.3t). Boxplot analysis was used to estimate \\nthe frequency (per kb) distribution of SNPs and InDels in the coding and 2 kb upstream areas, and the \\ngenes with substantially higher frequency of SNPs/InD els above the upper fence in the boxplot were \\nselected. Utilizing Cytoscape, comparative GO enrichment maps for various gene sets were also \\ncreated. [84] Numerous applications, including genetic mapping, association research, and population \\ngenetics, depen d on the detection of SNPs in the Morus indica plant. [85] SNPs can be used as markers \\nfor genetic mapping to locate the genes responsible for phenotypes. They can also be used to identify \\ngenetic variants associated with complex traits or diseases in association studies. In population genetics, \\nStudyi ng the genetic diversity and population organization using SNPs can reveal information about \\nthe evolutionary background and adaption of a species [86] Whole -genome resequencing of 21 \\nmulberry accessions/speci es was used to discover SNPs before high -quality reads were filtered and \\nmapped to the draught M. indica genome assembly. Utilizing tools like SnpEff and Cytoscape, the SNPs \\nwere annotated and examined for frequency distribution and gene enrichment.  \\nAn important resource for functional and translational genomics studies aimed for mulberry \\nimprovement is the discovery of SNPs in Morus indica . SNPs may be employed to examine the genetic \\ndiversity and structure of mulberry populations, generate molecular markers for breeding plans, and \\ndiscover genes linked to desired features [87].  \\n1.9.5 SIMPLE SEQUENCE REPEATS  \\nMicrosatellites, also known as simple sequence repeats (SSRs), are short repetitive DNA sequences \\nscattered throughout the genome. These SSRs serve as important genetic markers with diverse \\napplications such as marker -assisted selection, population genetic s, and genomic mapping. Utilizing a \\ncombination of technologies including Illumina, PacBio single -molecule real -time sequencing, and \\nBionano optical mapping, the draft genome sequence of Morus indica was analyzed to identify SSRs. \\nWithin the M. indica  geno me, 121,212 SSRs were identified, occurring at an average frequency of one \\nSSR per 3.2 kb. The most common SSR motifs included mononucleotide (57.5%), dinucleotide \\n(28.7%), trinucleotide (11.5%), tetranucleotide (1.7%), pentanucleotide (0.4%), and hexanucl eotide \\n(0.4%) repeats. These SSRs serve as molecular markers for genetic mapping and population genetics \\nresearch.  \\nRepetitive regions within the Morus indica genome were identified using various tools such as \\nRepeatMasker and RepeatModeler to locate SSRs. The distribution frequency and different motif types \\nof the identified SSRs were annotated and analyzed. These SSRs offer diverse utilities including \\nmarker -assisted selection, population genetics, and genomic mapping. They enable the examination of \\ngenetic diversity and population structure in mulberry populations, aid in the discovery of genes \\nassociated with desirab le traits and facilitate the development of molecular markers for breeding'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 29}, page_content=\"30  programs. The exploration of SSRs in Morus indica holds significant promise for advancing functional \\nand translational genomics in mulberry improvement. These SSRs offer avenues to investigate the \\ngenetic diversity and structure of mulberry populations, identify genes associated with desired traits, \\nand d evelop molecular markers for breeding strategies. The draft genome sequence of Morus indica , \\ngenerated through a fusion of four distinct technologies —namely, Illumina, PacBio single -molecule \\nreal-time seque ncing, chromosomal conformation capture (Hi -C), and Bionano optical mapping —\\nserved as the foundation for SSR discovery [88].  \\n1.10 PHENOTYPES OF MULBERRY STUDIES IN THIS RESEARCH  \\nThe production of mulberries (Morus spp.) is crucial to the sericulture sector since they serve as the \\nmain food supply for silkworms (Bombyx mori), which in turn helps the silk manufacturing sector. This \\nstudy begins a thorough investigation of mulberry phenotypes, including drought stress tolerance, leaf \\nyield, root rot resistance, and nitrogen consumption efficiency. The goal is to further knowledge of \\nmulberry genetics and aid in the creation of improved mulberry cultivars. Du e to their crucial roles in \\nmulberry farming and their capacity to completely alter sericulture techniques, t phenotypic features \\nhave been chosen with consideration.  \\n1.10.1 PROBLEMS ENCOUNTERED BY MULBERRY CULTIVATORS  \\nThe above -mentioned conditions frequently present mulberry farmers with a number of difficult \\nproblems, such as drought stress, root rot infections, and insufficient nitrogen usage efficiency. Reduced \\nwater availability brought on by drought stress can res ult in lower mulberry leaf production and, as a \\nresult, a reduction in the amount of food available to silkworms. In especially in areas susceptible to \\nwater constraint, this has a direct influence on the silk production cycle and, eventually, the revenue of \\nsericulture producers [89]. Diseases that cause root rot are still another serious issue. These diseases, \\nwhich cause wilting, leaf discoloration, and decreased plant vigor, can completely destroy mulberry \\ntrees. The ensuing fall in leaf yield puts the nutritional requirements of silkworms in danger, perhaps \\nresulting in less silk being produced. Additionally, the need for chemical treatments to battle root rot \\ninfections adds to farmers' expenses and might have a negative impact on the environment. The \\ndifficulties experienced by mulberry producers are further exacerbated by suboptimal nitrogen usage \\nefficiency [90]. Ineffective nitrogen absorption and use can result in higher fertilizer usage, rising \\nproduction costs, and nitrogen runoff that contribute s to environmental damage [91]. Diseases that cause \\nroot rot are still another serious issue. These diseases, which cause wilting, leaf discoloration, and \\ndecreased plant vigor, can completely destroy mulberry trees. The ensuing fall in leaf yield puts the  \\nnutritional requirements of silkworms in danger, perhaps resulting in less silk being produced. \\nAdditionally, the need for chemical treatments to battle root rot infections adds to farmers' expenses \\nand might have a negative impact on the environment. The  difficulties experienced by mulberry\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 30}, page_content=\"31  producers are further exacerbated by suboptimal nitrogen usage efficiency. [80] Ineffective nitrogen \\nabsorption and use can result in higher fertilizer usage, rising production costs, and nitrogen runoff that \\ncontributes to environmental damage [92]. Farme rs face a complicated conundrum that highlights the \\nneed of tackling nitrogen management in mulberry farming as they struggle with the combined weight \\nof economic inefficiency and environmental responsibility. Together, these problems highlight the \\ncomplex  interactions between environmental stresses and agricultural production in the mulberry \\nindustry. In order to successfully navigate these difficulties, farmers must strike a balance between the \\nneed for increased yield, cost -effective methods, and sustain able resource management, all the while \\nprotecting the silk production process, which is closely related to the availability and quality of mulberry \\nleaves. For the sericulture industry's long -term viability and the industry's overall environmental effect,  \\nas well as the financial well -being of farmers, it is imperative to address these issues via research and \\ninnovation.  \\n \\nFigure 5: Illustration of the various challenges faced by mulberry farmers\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 31}, page_content=\"32  1.10.2 DROUGHT STRESS  \\nThe primary focus of this research revolves around the issue of drought stress, given its significant \\nimplications for mulberry production. The escalation in the frequency and severity of drought events, \\nattributed to climate change and erratic weather pat terns, underscores the urgency of understanding \\nmulberry's genetic resilience to drought stress. This comprehension holds immense importance as \\nmulberry's ability to endure drought conditions is pivotal for sustaining sericulture production, \\nespecially in regions with limited water resources [93]. The exacerbation of water scarcity due to \\nclimate change could disrupt traditional sericulture practices, highlighting the necessity for drought -\\ntolerant mulberry cultivars. Thus, this study aims to identify the g enes and mechanisms underlying \\ndrought tolerance in various mulberry varieties to develop cultivars capable of thriving in arid \\nenvironments. Beyond its implications for sericulture, mulberry farming plays a crucial role in \\nagroforestry, providing liveliho ods for local communities in numerous regions. By cultivating \\ndrought -resistant mulberry cultivars, farmers can mitigate the impact of environmental uncertainties \\non their income, reducing their vulnerability to crop failures [94]. Rigorous phenotyping met hods will \\nbe employed to evaluate mulberry plants' response to water scarcity, examining variables such as \\nstomatal conductance, relative water content, and leaf water potential.  \\n 1.10.3 LEAF YIELD  \\nLeaf yield stands as a pivotal trait in this study, bearing significant importance not only for sericulture \\nbut also for agriculture. It directly correlates with the quantity of mulberry leaves a plant produces, \\nplaying a crucial role in determining the ou tput of silkworms (Bombyx mori) [95]. The essence of leaf \\nyield lies in its direct impact on silkworms' development, silk production, and overall well -being, given \\nthat mulberry leaves serve as their primary food source [96]. Thus, it becomes imperative fo r the \\nsericulture sector to unravel the genetic components responsible for high leaf production. High leaf \\nyield offers a multitude of advantages. Firstly, it ensures a consistent and ample supply of food for \\nsilkworms, thereby accelerating their growth ra tes and silk output. Ultimately, the aim of sericulture is \\nto enhance silk cocoon production, a goal attainable through increased leaf yield. Moreover, a reliable \\nand high -quality leaf supply reduces stress among silkworms, bolstering their resilience and overall \\nhealth. Financially speaking, high leaf production benefits sericulture producers by optimizing silk \\nproduction capacity and increasing revenue. Furthermore, it may extend the silk -producing season, \\nenabling growers to capitalize on market demands [97].  \\nTechnologies employed for evaluating leaf yield typically involve non -destructive techniques, given \\nthe necessity to retain leaves for ongoing silkworm feeding. Methods such as leaf area and dry weight \\nmeasurements are commonplace, with image analysis soft ware often utilized to estimate leaf area by \\ndetermining the total leaf surface area. Harvested leaves are dried to remove moisture before being\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 32}, page_content='33  weighed to estimate their dry weight, offering quantitative insights into leaf production without causing \\nharm to the plant. The study of leaf yield holds significance beyond sericulture, as mulberry cultivars \\nwith high yields can enhance overall agricultu ral production. Increased leaf output can serve as \\nadditional green fodder for animals or organic mulch in agroforestry systems, enriching soil fertility \\nand crop health. Mulberry leaves, being a rich source of nutrients, can also be utilized in various le af-\\nbased products such as tea or herbal treatments. Ultimately, leaf yield emerges as a crucial trait with \\nfar-reaching implications for agriculture and sericulture alike [88]. Understanding the genetic factors \\ncontributing to high leaf yield facilitates t he development of mulberry cultivars that not only boost silk \\nproduction and enhance financial returns for sericulture farmers but also potentially contribute to \\nbroader agricultural sustainability through increased fodder and nutrient -rich leaf production  [98].  \\n1.10.4 NITROGEN USE EFFICIENCY  \\nExploring Nitrogen Use Efficiency (NUE) within the context of mulberry cultivation holds paramount \\nimportance due to its profound implications for agriculture and sustainability. NUE plays a pivotal role \\nin maximizing agricultural productivity, efficient resource utilization, and environmental preservation. \\nNitrogen stands as a vital nutrient essential for plant growth and development, making the analysis of \\nNUE a crucial and intricate aspect of mulberry farming [99]. Given its far -reaching consequences, NUE \\nassumes critical significance in the realm of mulberry cultivation. NUE is relevant to mulberry \\ncultivation as it directly impacts crop yield, influencing silkworm productivity, a key factor in \\nsericulture. Nitrogen, being a crucial macronutrient, plays a pivotal role in promoting plant growth and \\nleaf formation, rendering mulberry leaves an ideal foo d source for silkworms. Higher NUE signifies the \\nability of mulberry plants to enhance leaf production while efficiently utilizing available nitrogen \\nresources, ensuring a consistent and abundant supply of premium mulberry leaves. This, in turn, \\nenhances s ilkworm health and production, thereby contributing significantly to the sustainability of the \\nsericulture sector [100].  \\nFurthermore, investigating NUE in mulberry cultivation facilitates cost -effective resource management \\nand environmental sustainability. Inefficient nitrogen utilization can lead to financial losses, given that \\nnitrogen fertilizers constitute a significant input cost in agriculture. Precision nitrogen management, \\nfacilitated by NUE research, reduces fertilizer usage, conserves resources, and mitigates the adverse \\nenvironmental impacts of nitrogen runoff, soil acidification, and groundwater contamination. Thi s \\nfosters a more responsible and environmentally friendly approach to mulberry farming, aligning with \\nprinciples of sustainable agriculture and environmental stewardship. Moreover, NUE research \\naddresses broader concerns related to agricultural sustainabil ity and climate resilience. Nitrogen \\npollution, largely attributable to agricultural activities, exacerbates numerous environmental issues. By \\nimproving NUE, mulberry cultivation can mitigate nitrogen losses into the environment, thereby'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 33}, page_content='34  mitigating problems like eutrophication and air pollution. Additionally, effective nitrogen utilization \\nbecomes crucial for crop resilience in the face of climate change -induced uncertainties in temperature \\nand precipitation patterns. High NUE mulberry cul tivars enable sericulture to adapt to climatic \\nchallenges by thriving in changing environmental conditions [101].  \\n \\n1.10.5 ROOT ROT CONDITIONS  \\nThis study centers on an in -depth examination of the root rot fungal condition affecting mulberry \\n(Morus spp.), addressing a significant challenge in mulberry cultivation. Root rot infections, caused by \\nvarious pathogenic fungi, pose a serious threat to th e health and productivity of mulberry plants. A \\ndetailed exploration of this issue underscores the urgent need to expand our understanding of root rot \\nfungal infections, their implications, appropriate assessment methodologies, and the broader context of \\nprior research in this domain. Effective management of these diseases necessitates a comprehensive \\nunderstanding of the root rot fungal condition. These infections manifest through symptoms such as \\nwilting, leaf yellowing, root discoloration, and overall we akened plants. By identifying the specific \\nfungal pathogens responsible for root rot conditions through visual inspections and laboratory analyses, \\ntargeted treatments can be devised to mitigate their impact and prevent crop losses. Given that mulberry \\nleaves serve as the primary food source for silkworms (Bombyx mori) in sericulture, addressing root \\nrot infections is crucial for ensuring a consistent supply of high -quality mulberry leaves, thereby \\npromoting silkworm health and production, essential for the  sustainability of the sericulture sector. \\nFurthermore, addressing the root rot fungal issue in mulberry plants aligns with ecologically friendly \\nagricultural practices. Developing mulberry cultivars that naturally resist these infections obviates the \\nneed  for synthetic fungicides, contributing to environmentally responsible farming practices. By \\ninvestigating the genetic basis of resistance, this study holds the potential to develop robust and durable \\nmulberry cultivars, particularly relevant amid the impe rative for responsible resource management in \\nagriculture and changing environmental conditions. The evaluation of the root rot fungal condition \\nemploys a multimodal approach, encompassing visual observation of disease symptoms, isolation and \\nidentificatio n of fungal pathogens from infected plant samples using microscopy and molecular \\ntechniques, and assessment of disease severity. Advanced molecular techniques such as DNA \\nsequencing and PCR enable precise identification and classification of root rot patho gens, shedding \\nlight on the diversity of fungi affecting mulberries. Drawing on prior research in the field of root rot \\ndiseases, including studies on pathogenicity mechanisms, host specificity, and management strategies \\nacross various crops, this research  expands our understanding of the unique challenges faced by \\nmulberry growers and silk producers. It builds upon this foundation to foster the resilience of the \\nsericulture industry over the long term by developing disease -resistant mulberry varieties.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 34}, page_content='35  In conclusion, investigating the root rot fungal condition in mulberry cultivation is a critical endeavor \\nwith implications that extend beyond crop health and productivity, encompassing the sustainability of \\nagriculture and the prosperity of the sericultur e sector.  \\n \\n1.11 IDENTIFICATION OF BIOMARKERS FOR THE FOUR PHENOTYPES \\nOF MULBERRY  \\nThe principal discovery of this study, with significant implications for sericulture and sustainable \\nagriculture, involves the identification of biomarkers associated with leaf yield, root rot resistance, and \\nnitrogen utilization efficiency in mulberry far ming. Biomarkers serve as physiological or molecular \\nindicators linked to the presence or severity of a particular trait or condition. It is imperative to identify \\nthese biomarkers due to factors such as precision breeding, crop management, and the develop ment of \\nrobust, disease -resistant mulberry cultivars. They can function as early indicators of stress and plant \\nhealth, particularly in plants capable of withstanding drought stress, facilitating proactive management. \\nThese biomarkers may include physiolog ical indicators such as leaf water potential or relative water \\ncontent, or they may be specific gene expressions associated with stress response pathways. \\nTechnologies like RNA -Seq can be employed to identify candidate genes whose expression patterns \\ncorre late with drought tolerance, shedding light on the molecular mechanisms underlying this trait. \\nPrevious research in crops like maize and rice has unveiled drought -related biomarkers that can inform \\nmulberry studies. Biomarkers offer a means to anticipate a nd select high -yielding mulberry cultivars \\nduring breeding programs focusing on leaf production. They could be based on genes that enhance leaf \\nyield or linked to critical physiological processes influencing leaf development. Molecular approaches \\nsuch as t ranscriptomics and marker -assisted selection can be utilized to discover and validate these \\nbiomarkers. Although research on mulberry -specific leaf yield biomarkers is limited, insights from \\nstudies on other crops provide valuable information on the geneti c basis of high -yielding traits. \\nIdentifying mulberry plants with innate resistance to root rot pathogens necessitates the use of \\nbiomarkers for root rot resistance. These biomarkers may encompass biochemical indicators associated \\nwith pathogen interaction s, disease resistance mechanisms, or genes involved in defense responses. \\nGenomic methods like genome -wide association studies (GWAS) and molecular techniques for \\nidentifying disease resistance genes can aid in identifying these biomarkers. Previous studie s on other \\ncrops have identified biomarkers linked to resistance against various diseases, offering guidance for \\nidentifying similar biomarkers in mulberry. Efficient nitrogen management in mulberry agriculture \\nrelies on indicators for nitrogen utilization  efficiency. These biomarkers may include physiological \\nsignals indicating nitrogen status in the plant and genes involved in nitrogen absorption, assimilation, \\nor transport. Molecular methods such as gene expression analysis and metabolomics can help disc over'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 35}, page_content='36  nitrogen -related biomarkers. While research on mulberry -specific biomarkers for nitrogen utilization \\nefficiency is nascent, lessons from analogous studies in crop species like rice and wheat can inform the \\nsearch for relevant biomarkers in mulberries. The discovery of biomarkers for these traits in mulberry \\nrepresents a cutting -edge endeavor with the potential to enhance crop management, breeding programs, \\nand sericulture practices. These biomarkers serve as valuable tools for meticulous selection, improved  \\ncrop management, and the development of resilient and high -yielding mulberry cultivars. Despite \\npotential limitations in prior research on biomarkers for these specific traits in mulberry, the wealth of \\ninformation from analogous studies in other crops pr ovides a solid foundation for this crucial research \\nendeavor [102].  \\n1.12 GENETIC LINKAGE & MAPPING STUDIES  \\nGenetic linkage mapping is an indispensable method in genetic studies that delineates the locations of \\ngenes and genetic markers on chromosomes, as well as their proximity. This technique holds immense \\nimportance for unraveling the genetic underpinnings of  various traits, such as yield, disease resistance, \\nand stress tolerance, within the context of mulberry ( Morus indica ), a tree species of significant \\neconomic importance for the sericulture industry. Introducing linkage mapping in mulberry is \\nimperative t o advance breeding programs, enhance crop management practices, and augment overall \\nmulberry productivity [103].  \\n The significance of mulberry linkage mapping lies in its provision of a blueprint of the mulberry \\ngenome and identification of genes and markers associated with desirable traits. This genetic roadmap \\nfacilitates the improvement of mulberry cultivars throu gh selective breeding, yielding varieties that are \\nhigh-yielding and resistant to diseases. Moreover, linkage mapping aids in the development of \\nsustainable sericulture methods by elucidating the genetic pathways underlying complex traits like \\ndrought tole rance, leaf yield, and nitrogen use efficiency. Understanding the genetic structure of \\nmulberry enables researchers and breeders to expedite the creation of robust and productive cultivars, \\nessential for the expansion of the silk industry.  \\nMulberry linkage mapping can be conducted using various methods, with molecular markers being the \\nprimary tools. Commonly employed markers for linkage mapping encompass simple sequence repeats \\n(SSRs), single nucleotide polymorphisms (SNPs), and amplified f ragment length polymorphisms \\n(AFLPs) [104]. These markers are selected based on their genome -wide distribution and polymorphic \\ncharacteristics. Typically, genetically diverse mulberry individuals are crossed, and the resulting \\nprogeny are genotyped using m olecular markers. Subsequently, data from this genotyping is processed \\nusing specialized software and statistical techniques to unveil marker -trait associations and construct \\ngenetic maps.  \\nIn summary, linkage mapping in the mulberry species Morus indica serves as a crucial genetic research'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 36}, page_content='37  tool with significant implications for sustainable agriculture and the sericulture sector. By elucidating \\nthe genetic basis of key traits, linkage mapping facilitates the development of superior mulberry \\ncultivars, ultimately enhancing silk production and ensuring the financial viability of the sericulture \\nindustry. Leveraging a variety of molecular markers and analytical methods, linkage mapping is \\nindispensable for comprehending the complexities of the mulberry genome and expanding breeding \\nprograms to me et the evolving needs of the silk industry. In our research, we focus on genotypes, each \\nbringing unique benefits to our study. Thailand Male, a renowned Mulberry genotype, is favored in \\nsericulture due to its reputation for producing high -quality leaves. Its robust leaf yield and adaptability \\nto diverse environmental conditions offer an opportunity to explore the genetic mechanisms underlying \\nsuperior leaf production. Our research, which includes Thailand Male, aims to identify genetic \\nvariations like SNPs  and SSRs associated with high leaf yield. These markers could be instrumental in \\nselective breeding programs to develop superior Mulberry cultivars that consistently provide nutrient -\\nrich leaves for silkworms.  \\nAssam Bol, originating from the Assam region of India, is intriguing for its inherent resistance to root \\nrot diseases. Root rot infections pose a significant challenge in Mulberry production, often leading to \\nreduced plant health and productivity. By study ing Assam Bol, we can investigate the genetic factors \\ncontributing to its resistance to root rot pathogens. This exploration may unveil genetic, and biomarkers \\nlinked to root rot resistance, paving the way for breeding initiatives aimed at developing disea se-\\nresistant Mulberry cultivars. Such research holds practical implications, potentially reducing the \\nreliance on chemical fungicides and promoting environmentally sustainable Mulberry agriculture.  \\nS1, known for its exceptional drought resistance, is another crucial genotype in our study. \\nUnderstanding the genetic basis of drought resistance in Mulberry is vital in the face of increasing \\ndrought conditions due to climate change. Our investigation int o S1 aims to uncover genetic variations, \\nsuch as SNPs and SSRs, associated with its drought resistance. This knowledge could aid in the \\ndevelopment of drought -resistant Mulberry cultivars, ensuring a consistent leaf supply even in water -\\nlimited areas. This  study not only supports the sericulture sector but also aligns with sustainable farming \\npractices amidst changing climatic patterns.  \\nPunjab Local, tailored to the regional characteristics of the Punjab area, plays a significant role in \\nresearching nitrogen use efficiency in Mulberry. Nitrogen is crucial for plant development, and efficient \\nnitrogen use is essential for enhancing agricul tural productivity while mitigating negative \\nenvironmental impacts. By analyzing Punjab Local, we seek to identify genetic and biomarkers for \\neffective nitrogen utilization in Mulberry. This understanding may inform strategies for optimal \\nnitrogen manageme nt, minimizing resource wastage and environmental degradation. Furthermore, it \\ncould promote ethical agricultural practices and enhance the economic feasibility of Mulberry farming.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 37}, page_content='38  The inclusion of multiple Mulberry genotypes in our study broadens the scope of investigation and \\nenables a comprehensive genetic examination of traits vital to the sericulture sector and agriculture at \\nlarge. We aim to elucidate the genetic foundations of  these traits using advanced methods such as SNP \\nand SSR markers, genetic linkage mapping, and sophisticated molecular tools like DNA microarray \\nchips. These tools also hold potential as diagnostic aids for identifying Mulberry crop issues, \\nsignificantly e nhancing the adaptability, productivity, and sustainability of the sericulture industry.  \\nOur selection of four Mulberry genotypes represents a comprehensive and strategic approach to our \\nstudy. The findings from our research could have implications for sustainable farming practices and the \\nsericulture industry, particularly in the face of chan ging environmental conditions. By uncovering \\ngenetic markers, deciphering genetic linkage maps, and developing diagnostic tools, our study has the \\npotential to revolutionize Mulberry agriculture, enhancing its economic viability and environmental \\nstewardsh ip.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 38}, page_content=\"39  CHAPTER 2  \\nREVIEW OF LITERATURE  \\n2.1 POLYPLOIDY IN PLANTS  \\nPolyploidy refers to the heritable condition in which an organism possesses more than two complete \\nsets of chromosomes. This phenomenon is observed in various organisms, including plants, certain \\nspecies of fish, and amphibians [105]. The elevated chromoso mal count observed in these plants is \\nattributed to a genome duplication event. There are two primary types of polyploids distinguished by \\nthe origin of this duplication event: autopolyploid, which arise from the multiplication of a diploid \\ngenome within a  single species, and allopolyploids, which occur when two distinct haploid genomes \\nare combined through hybridization and subsequently doubled [106].  \\nDespite the differentiation between autopolyploids and allopolyploids, the process of doubling genetic \\nmaterial generates a genomic buffering effect that provides advantages to both types of polyploids. This \\nbuffering mechanism safeguards the organism agai nst the detrimental impacts of epigenetic alterations \\nand changes in DNA sequences, thereby ensuring minimal or no adverse effects on its viability. \\nFurthermore, this process promotes genetic diversity and facilitates the evolution of the genome [95,96]. \\nPlant polyploidy offers significant benefits to humans, particularly considering that many major \\nagricultural crops worldwide are polyploid. Recent genomic investigations have revealed multiple \\ninstances of genome duplication in various eukaryotic lineages,  including yeast, vertebrates, and even \\norganisms like Arabidopsis thaliana, previously believed to have a diploid small -genome structure \\n[107].  \\n 2. 2 EXPLORING POLYPLOID GENOMES IN PLANTS  \\n2.2.1 METHODS OF ANALYZING GENOME   \\nThe combination of genetic mapping, molecular cytogenetics, sequence analysis, and comparative \\nanalysis has enabled the discovery of numerous new understandings regarding the evolution of ploidy \\nacross various timescales. These insights range from the orig ins of the plant kingdom to events such as \\nintra- and interspecific hybridization linked to plant domestication and breeding. In -depth examination \\nof chromosomes through techniques like in situ hybridization, along with comprehensive molecular \\nmarker analy ses spanning entire genomes, has yielded compelling evidence regarding the mechanisms \\nunderlying genomic modification.  \\nThe integration of genetic mapping, molecular cytogenetics, sequencing, and comparative analysis has \\nprovided valuable insights and expanded our understanding of ploidy evolution across various temporal \\nscales. This encompasses investigations spanning the plant kingdom's foundational stages to the\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 39}, page_content='40  occurrences of intra - and interspecific hybridization events linked to both plant domestication and \\nbreeding processes. A robust understanding of the mechanisms driving genomic modifications has been \\nsubstantially enriched through the application of in sit u hybridization techniques for physical \\nchromosome analysis and comprehensive genome -wide molecular marker analyses [108].  \\n2.2.2 In Situ  Hybridization  \\nIt proves notably proficient in precisely discerning chromosomes and facilitating the mapping of \\nrepetitive DNA sequences as well as unique sequences onto the chromosome(s). Through the utilization \\nof DNA probes tagged with fluorescent labels, fluorescent in situ hybridization (FISH) is employed, \\nallowing visualization via a fluorescence microscope. Genomic in situ hybridization (GISH) employs \\nthe complete genomic DNA of a species as a probe on chromosomes, enabling discrimination of the \\nentire genome rathe r than specific sequences. Investigation into the distribution of four tandem repeats \\nin allotetraploids, such as Tragopogon mirus, Tragopogon miscellas, and their diploid progenitors, \\nrevealed evidence suggesting that chromosomal rearrangements did not oc cur after polyploidization, \\nhighlighting additive patterns in polyploids [109].  \\n2.2.3 Molecular Marker -Based Genetic Mapping  \\nPolyploid genetic mapping presents unique challenges in contrast to diploid species, primarily \\nstemming from the intricate segregations and statistical methodologies involved. The accuracy of \\ngenetic distance measurements necessitates substantial segregati ng populations to address these \\ncomplexities effectively [100]. Important discoveries emerged from the application of Amplified \\nFragment Length Polymorphism (AFLP) analysis in Arabidopsis, Brassica, and Spartina [110] or when \\ntraditional Restriction Fragme nt Length Polymorphism markers (RFLP) were used in B. napus, Draba \\nnorvegica, and T. miscellus, as well as wheat [111].  \\n2.2.4 Methylation Sensitive Molecular Markers  \\nA method akin to AFLP, employing restriction enzymes with identical recognition sites but varying \\nsensitivity to DNA methylation (isoschizomers), has proven effective and reliable for discerning \\ngenome -wide DNA methylation patterns [110]. The technique, re ferred to as Methylation -Sensitive \\nAmplified Polymorphism (MSAP), operates by targeting the 5′CCGG sequence through the \\nisoschizomers HpaII and MspI, both sensitive to the methylation status of either the outer or inner'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 40}, page_content='41  cytosine residues. Remarkable results were observed in newly formed polyploids through this \\nmethodology, as demonstrated by Madlung et al. in Arabidopsis [111].  \\n2.2.5 Comparative Genome Analysis  \\nIn Brassica oleracea, around 75% of the genes underwent loss as evidenced by a comparative \\nexamination of the genomic segments. Those genes that remained were not distributed randomly; \\ninstead, they displayed dosage sensitivity. After whole -genome duplicat ion (WGD), there was a notable \\nover-retention of duplicates belonging to transcription factors and signal transduction pathway \\nelements. Conversely, following smaller -scale duplications, the retention rates of these functional gene \\ncategories were comparat ively reduced [112]. In four separate polyploid wheat lineages, repeated \\ndeletions of the Puroindoline (Pin) gene at the Grain Hardness (Ha) locus were identified [113]. Various \\nmethods for detecting single nucleotide polymorphisms (SNPs) have been employe d to explore plant \\npolyploidy, as demonstrated by the research [114]. The Illumina Golden Gate assay revealed a \\nconsiderable abundance of single nucleotide polymorphisms (SNPs) in tetraploid and hexaploid wheat. \\nRecently, Allen et al. further investigated this phenomenon [115]. In 6225 distinct reference sequences \\nof hexaploid bread wheat, the Illumina GAIIx data identified upwards of 14,000 potential single \\nnucleotide polymorphisms (SNPs). Similarly, the Illumina sequencing platform detected over one \\nmilli on SNPs in elite inbred lines of maize [116].  \\n2.2.6 High -Throughput DNA Sequencing and High -Resolution Melting (HRM) \\nAnalysis  \\nGeneral solutions for the genetic analysis of polyploids are provided by high -throughput DNA \\nsequencing combined with computational analysis [117]. The high level of heterozygosity made \\nattempts to sequence the heterozygous diploid potato genome (RH89 -039-16) difficult [118]. The \\nwoodland strawberry (Fragaria vesca) genome was sequenced to avoid the challenges associated with \\nsequencing the polyploid genome of the cultivated strawberry (Fragaria x ananassa) [119].  \\n2.3 Current strategies of Sequence alignment and SNP calling in Polyploid plants  \\nNGS technologies, introduced in 2004, brought about a significant boost in sequencing throughput and \\ncost reduction [120]. However, their short sequence read output also heightened the complexity of \\nfragment assembly. Despite this challenge, NGS finds wide -ranging applications in various fields such \\nas genome sequencing and resequencing, metagenomics, RNA -sequencing (transcriptomics), and \\npersonalized genomics (personal medicine). Notably, it allows genome sequencing even with lower'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 41}, page_content=\"42  DNA concentrations. Recent NGS platforms include Roche's 454 or pyrosequencing (with read lengths \\nup to 700 bp), Life Technologies SOLiD (50 bp), Illumina's HiSeq (two by 250 bp), and MiSeq (two \\nby 300 bp), and Life Technologies' Ion Torrent/Proton (200 bp ). Unlike Sanger sequencing, NGS \\nmethods eliminate the need for DNA cloning, simplifying the process and enhancing adaptability to \\nvarious biological phenomena. This enables massive parallelization at reduced costs. However, NGS \\npresents drawbacks such as the requirement for special assembly algorithms due to short sequence \\nlength, less accurate base calling compared to Sanger sequencing, and lower -quality assemblies than \\nthose generated from Sanger sequencing [121].  \\nThe genomes of polyploid plant species, including Gossypium hirsutum  (cotton) and the hexaploid \\nTriticum aestivum  (wheat), have been sequenced using Illumina technology [122,123].  \\n 2.4 Third Generation Sequencing  \\nThird -generation sequencing technologies, characterized by long reads, represent the latest \\nadvancements in genome sequencing. These techniques yield longer read lengths, typically measured \\nin kilobases (Kb) rather than bases (bp), while also reducing sequ encing costs and simplifying \\npreparation and sequencing protocols [124]. This emerging technology offers numerous advantages, \\nbut it also presents drawbacks like high error rates and the requirement for exceptionally high -quality \\nDNA samples. Nevertheless,  by generating a substantial quantity of long reads to aid in resolving \\nchallenging regions within the genome, these methods show promise in addressing the hurdles \\nassociated with sequencing and assembling large, repetitive, and intricate plant genomes. Wi thin third -\\ngeneration sequencing approaches, long -read sequencing and long -range scaffolding technologies are \\nthe two main categories of techniques utilized [125].  \\n  \\n2.5 Best practices in the identification of molecular markers  \\nThe identification of molecular markers is a crucial step in various genetic studies, and following best \\npractices ensures the reliability and accuracy of these markers. Here are key points:  \\n2.5.1 Define Clear Objectives  \\nBefore initiating the process of marker identification, researchers must establish clear objectives for \\ntheir study. A comprehensive understanding of the traits or genetic variations being targeted facilitates \\nthe selection of suitable molecular markers.  \\n2.5.2 Choose Appropriate Marker Types  \\nVarious studies may necessitate diverse molecular markers, such as Single Nucleotide Polymorphisms\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 42}, page_content='43  (SNPs) or Simple Sequence Repeats (SSRs), contingent upon factors like the organism under \\ninvestigation, the desired extent of genetic variation, and the accessibility of genomic data.  \\n2.5.3 Utilize High -Quality Genomic Data  \\nThe efficacy of marker identification hinges on the caliber of genomic data. Researchers ought to \\nemploy high -throughput sequencing techniques or other sophisticated methodologies to produce precise \\nand thorough genetic data.  \\n2.5.4 Consider Genetic Diversity  \\nChoosing polymorphic markers is vital for capturing population diversity, enabling the differentiation \\nbetween individuals or groups and offering valuable insights for genetic inquiries.  \\n2.5.5 Validation of Markers  \\nIdentified markers should undergo rigorous validation to ensure their accuracy and reproducibility. This \\ninvolves testing the markers across diverse individuals or populations to confirm their association with \\nthe target traits.  \\n2.5.6 Evaluate Marker Transferability  \\nIn cases where markers are intended for use across different populations or species, assessing their \\ntransferability becomes crucial. Markers that can be reliably applied in diverse genetic backgrounds are \\nhighly valuable.  \\n2.5.7 Quality Control Measures  \\nImplementing quality control measures during marker identification is essential. This includes filtering \\nout low -quality data, ensuring consistency in experimental conditions, and applying stringent criteria \\nfor marker selection.  \\n2.5.8 Consider Genomic Context  \\nMolecular markers do not act in isolation; they are part of the larger genomic context. Researchers \\nshould consider the genomic environment to understand the potential effects of neighboring genes and \\nregulatory elements on marker functionality.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 43}, page_content='44   \\n2.5.9 Document Methodology and Parameters  \\nTransparency in methodology is vital for reproducibility. Researchers should thoroughly document the \\nmethods used for marker identification, along with any specific parameters or thresholds applied during \\nthe process.  \\n2.5.10  Community Standards and Resources  \\nAdhering to community standards and utilizing available genetic databases or resources can enhance \\nthe robustness of marker identification. Collaboration with other researchers and leveraging shared \\nknowledge contributes to the establishment of best practices.  \\n2.5.11 Ethical Considerations  \\nResearchers should adhere to ethical guidelines in marker identification, ensuring that studies are \\nconducted responsibly and with consideration for potential societal implications.  \\nIn conclusion, the best practices in the identification of molecular markers involve setting clear \\nobjectives, selecting appropriate marker types, utilizing high -quality genomic data, considering genetic \\ndiversity, validating markers, evaluating transferab ility, implementing quality control measures, \\nconsidering the genomic context, documenting methodologies, adhering to community standards, and \\naddressing ethical considerations. Following these practices enhances the credibility and utility of \\nidentified m olecular markers in various genetic studies.  \\n2.6 SNP Genotyping chip for genetic studies and Breeding applications  \\n In the realm of genetic studies and breeding applications, SNP genotyping chips have emerged as \\ninvaluable tools for understanding genetic variations and facilitating targeted breeding efforts. These \\nchips are designed to analyze Single Nucleotide Polymor phisms (SNPs), which are subtle genetic \\ndifferences occurring at a single nucleotide level. The adoption of SNP genotyping chips enables \\nresearchers to efficiently identify and study these variations across the genome [126].  \\n The primary advantage of SNP genotyping chips lies in their ability to simultaneously assess many \\nSNPs, providing a comprehensive view of the genetic landscape. This technology is particularly useful \\nin genetic studies where researchers aim to identify as sociations between specific genetic markers and \\ntraits of interest, ranging from disease susceptibility to agricultural characteristics.  \\nFor breeding applications, SNP genotyping chips play a pivotal role in developing improved plant'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 44}, page_content='45  varieties or animal breeds. By identifying SNPs associated with desirable traits, breeders can selectively \\nchoose individuals with the desired genetic makeup, enhancing the efficiency of the breeding process \\n[127]. This targeted approach allows for the dev elopment of crops with increased yield, resistance to \\ndiseases, or improved nutritional content, contributing to advancements in agriculture. The success of \\ngenetic studies and SNP genotyping chip applications is intricately linked to breeding techniques a nd \\nagricultural practices. Different agricultural practices can significantly influence the genetic makeup of \\nplants, including crops like mulberry. For example, cultivation methods, soil management, and \\nirrigation practices can impact the genetic diversit y within a population. These factors may contribute \\nto the selection of specific genetic traits over time, influencing the distribution of SNPs and markers. \\nUnderstanding the agricultural practices employed in mulberry cultivation is essential for interpre ting \\nthe genetic data obtained from SNP genotyping chips accurately. Selective breeding techniques also \\nplay a crucial role in shaping the genetic composition of cultivated plants. If specific breeding practices \\nhave been employed to enhance certain traits  in mulberry, such as disease resistance or silk production, \\nthe associated genetic markers may be detectable through SNP genotyping [128]. Conversely, \\ntraditional breeding methods, which rely on natural processes, may result in a more diverse genetic \\nlandscape. Furthermore, the impact of environmental factors, such as climate and geographic location, \\non mulberry cultivation cannot be overlooked. These factors may contribute to the adaptation of certain \\ngenetic variations, potentially influencing the distri bution of SNPs. In conclusion, the success of genetic \\nstudies and breeding applications, particularly those involving SNP genotyping chips, is closely tied to \\na comprehensive understanding of breeding techniques and agricultural practices. The interaction \\nbetween these factors shapes the genetic makeup of cultivated plants, influencing the presence and \\ndistribution of SNPs and markers. A holistic approach that considers both genetic data and the context \\nof cultivation practices is essential for leveraging t he full potential of SNP genotyping chips in \\nadvancing genetic studies and breeding programs [129].  \\n 2.7 Prerequisites for the genotyping chip  \\nIn the process of developing a genotyping chip, researchers carefully select only a limited number of \\nSingle Nucleotide Polymorphisms (SNPs) to serve as markers. The choice of these specific SNPs is not \\narbitrary; rather, it is guided by stringent criteria . The selection process involves developing consensus \\nsequences, essentially creating a simulated genome sequence that closely mimics the actual genome. \\nFrom these mimicked sequences, confidential SNPs are derived. The criteria for choosing SNPs for the \\ngenotyping chip include quality control (QC) measures and in silico  criteria. Quality control is a vital \\nstep in ensuring that the selected SNPs meet certain standards. Researchers set threshold values to \\ndetermine which SNPs cut. This threshold value acts as a filter, allowing only those SNPs that surpass'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 45}, page_content='46  a specified quality benchmark to be included in the genotyping chip [130]. This rigorous QC process \\nserves as a safeguard against errors and inaccuracies, ensuring the reliability of the selected SNPs for \\ndownstream analyses. The relationship between the t hreshold values for SNP selection and polyploidy \\nlevels is significant, especially in the case of highly polyploid organisms like the mulberry. Polyploidy, \\nthe condition of having multiple sets of chromosomes, can complicate genetic analyses. By \\nimplementi ng stringent QC criteria, researchers indirectly address polyploidy issues. The QC measures \\nhelp in filtering out noise and inaccuracies that may arise due to the complexity of polyploid genomes, \\nensuring that only high -quality SNPs are chosen for the geno typing chip. In the context of VCF (Variant \\nCall Format) files, which store genetic variations data, there are QC and Pass/Fail columns. If a SNP \\nfalls below the predefined threshold during QC analysis, it is marked as \"Fail,\" indicating that it did not \\nmeet the quality standards and should be excluded from further analyses. This additional layer of QC, \\nintegrated into the genotyping process, enhances the reliability of the selected SNPs and contributes to \\nthe overall success of the genotyping chip developm ent, particularly in the context of addressing \\nchallenges associated with polyploidy, as observed in the case of the highly polyploid mulberry [131].  \\n \\nFigure 6: The picture depicts the requirements for designing a genotyping chip  \\n \\n2.8 Linkage mapping studies  \\nIn the exploration of linkage mapping studies, researchers delved into the intricate relationships between \\ngenetic variations and observable traits. To unravel these connections, they considered four distinct \\ngenotypes and phenotypes within their samples. Each genotype represents a specific genetic makeup, \\nwhile phenotypes encompass the observable characteristics or traits exhibited by the samples. The'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 46}, page_content='47  inclusion of gene ontology, which categorizes genes based on their functions and roles, provided \\nadditional insights into the biological significance of the genetic variations under investigation. With a \\nwealth of Single Nucleotide Polymorphism (SNP) data derived from consensus sequences, the \\nresearchers embarked on the task of linking this genetic information to the observed traits, particularly \\nfocusing on stress tolerance tests. These tests aimed to identify specific SNPs associated with stress \\ntolerance , shedding light on the genetic basis of resilience in the studied samples. The process of \\nconnecting SNPs to gene ontology and linking them to traits involved a meticulous examination of the \\ngenetic landscape. Researchers sought to understand how variatio ns in the SNP data correlated with the \\nfunctions and roles of specific genes as defined by gene ontology. This comprehensive approach \\nallowed for the identification of genetic markers that played a crucial role in stress tolerance, offering \\nvaluable insigh ts into the molecular mechanisms governing this trait. Subsequently, the researchers \\nconducted linkage mapping studies to further elucidate the connections between the identified SNPs \\nand stress tolerance traits. Linkage mapping involves creating a map tha t illustrates the physical \\nlocations of genes on chromosomes and their associations with specific traits. By integrating SNP data \\nwith the results of stress tolerance tests and considering gene ontology, the linkage mapping studies \\naimed to provide a holis tic view of the genetic landscape associated with stress tolerance in the studied \\nsamples. This integrated approach not only enhances our understanding of the genetic factors \\ncontributing to stress tolerance but also lays the groundwork for more targeted a nd effective strategies \\nin breeding programs and molecular studies [132].  \\n In traditional linkage mapping, the goal is to identify the location of genes on chromosomes. However, \\nin this study on mulberry, the challenge emerged due to the absence of well -defined chromosomes for \\nmulberry plants. Unlike some organisms where chromos omes are clearly identified, mulberry lacks \\nsuch definite chromosomal information. Consequently, the usual approach of mapping loci on \\nchromosomes was not feasible in this study. The absence of chromosomal mapping in the research \\nhighlights a gap in our un derstanding of the mulberry genome. Despite the efforts to connect SNPs with \\ngene ontology and conduct linkage mapping studies, the lack of chromosomal information limits the \\ncomprehensive mapping of genetic loci [133].  \\n Looking ahead, future studies could explore various avenues to address the absence of chromosomal \\nmapping in mulberry. One approach involves further genomic research to uncover the structure and \\norganization of mulberry chromosomes. Advances in technology  and methodologies for genome \\nsequencing may contribute to revealing the chromosomal landscape of mulberry plants. Additionally, \\ncollaborations with experts in genomics and bioinformatics could provide valuable insights into \\ntechniques that may aid in reso lving the challenge of chromosomal mapping in mulberry. While the \\ncurrent study faced limitations in achieving chromosomal mapping, future research endeavors hold the'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 47}, page_content='48  potential to enhance our understanding of mulberry genetics. The pursuit of novel methodologies and \\nthe integration of cutting -edge genomic tools may pave the way for comprehensive chromosomal \\nmapping studies in mulberry plants, thereby advancing our knowl edge of their genetic architecture \\n[134].  \\n \\n2.9 Plant Linkage mapping with genetic markers  \\nPlant linkage mapping with genetic markers is a crucial aspect of understanding the inheritance of traits \\nin plants. In this process, researchers analyse the association between genetic markers and specific traits, \\nunravelling the genetic basis of various characteristics in plants. Genetic markers, such as Single \\nNucleotide Polymorphisms (SNPs) or Simple Sequence Repeats (SSRs), serve as signposts along the \\nplant genome. These markers act like landmarks, helping researchers identify the location of genes \\nresponsible for traits. Linkage mapping essentially involves tracking the inheritance patterns of these \\nmarkers within plant populations [135].  \\n  \\nOne significant step in plant linkage mapping is the selection of appropriate genetic markers. \\nResearchers carefully choose markers that are distributed across the genome and exhibit variability \\nwithin the plant population. This diversity ensures that the markers can effectively highlight genetic \\ndifferences related to specific traits of interest. Once the genetic markers are chosen, they are applied \\nto plant populations with known traits. The goal is to observe how the markers segregate (get passed \\non) alo ng with the traits from one generation to the next. By studying the co -inheritance patterns, \\nresearchers can map the locations of genes associated with plant characteristics. In the context of plant \\nbreeding programs, linkage mapping is invaluable. It help s plant breeders identify regions of the \\ngenome that are linked to desirable traits, allowing for the targeted development of new plant varieties \\nwith improved features. For example, if a genetic marker is consistently inherited with a trait like \\ndisease r esistance, breeders can selectively choose plants with that marker to enhance resistance in \\nsubsequent generations [136].  \\n In some cases, linkage mapping might be challenging due to factors like the complexity of plant \\ngenomes or the presence of genetic variations. Advanced technologies, such as Next -Generation \\nSequencing (NGS), have significantly enhanced the precision and e fficiency of linkage mapping \\nstudies, enabling researchers to delve deeper into the genetic architecture of plants. In conclusion, plant \\nlinkage mapping with genetic markers is a powerful tool in deciphering the genetic basis of plant traits. \\nIt contribute s to our understanding of inheritance patterns, aids in targeted plant breeding, and facilitates \\nthe development of improved crop varieties with desirable characteristics. The integration of cutting -\\nedge technologies further strengthens the accuracy and sc ope of plant linkage mapping studies, opening'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 48}, page_content=\"49  new avenues for advancements in plant genetics and breeding [137].   \\n \\n2.10 Computational requirements for linkage mapping studies  \\nComputational requirements play a crucial role in facilitating linkage mapping studies, especially in the \\ncontext of unravelling the genetic basis of traits in various organisms. These studies involve the analysis \\nof large datasets, complex statistical computations, and intricate bioinformatics analyses, demanding \\nsignificant computational resources. One fundamental aspect of computatio nal requirements in linkage \\nmapping studies is the handling of massive genomic datasets. As researchers utilize advanced \\ntechnologies like Next -Generation Sequencing (NGS), they generate vast amounts of genetic data. \\nEfficient storage solutions and powerfu l computing infrastructure are essential for managing and \\nprocessing these large datasets. High -performance computing clusters or cloud -based platforms often \\nbecome indispensable for handling the computational load associated with genomic data analysis [13 8]. \\nThe algorithms employed for linkage mapping are computationally intensive, requiring substantial \\nprocessing power. These algorithms assess the genetic markers' segregation patterns and calculate \\nstatistical probabilities to identify regions of the genome a ssociated with specific traits. Parallel \\ncomputing capabilities, which allow multiple calculations to be performed simultaneously, can \\nsignificantly accelerate the execution of these algorithms. Additionally, bioinformatics tools are integral \\nto extracting  meaningful insights from genomic data. Variant calling, genotype imputation, and quality \\ncontrol procedures involve complex computational procedures. Adequate computational resources are \\nnecessary to execute these bioinformatics analyses accurately. Moreo ver, the visualization of linkage \\nmaps and the interpretation of results often involve software applications that demand substantial \\ncomputational capabilities. The size and complexity of the genome being studied also impact \\ncomputational requirements. For  organisms with large and intricate genomes, such as some plants, \\nanimals, or humans, the computational demands escalate. As researchers aim for higher resolution and \\naccuracy in mapping genetic loci, the computational infrastructure must be robust enough to handle the \\nintricacies of such genomes. As technologies evolve, the computational demands of linkage mapping \\nstudies continue to grow. The integration of machine learning approaches, advanced statistical methods, \\nand sophisticated algorithms further acc entuates the need for powerful computational resources. \\nCollaboration with computational experts, access to supercomputing facilities, and utilization of cloud \\ncomputing services are common strategies employed by researchers to meet these computational \\nrequirements. In summary, computational requirements for linkage mapping studies encompass the \\nmanagement of large genomic datasets, execution of complex algorithms, utilization of bioinformatics \\ntools, and adaptation to the size and complexity of the studied  genome. Adequate computational \\nresources are essential for the successful execution and interpretation of linkage mapping studies,\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 49}, page_content='50  contributing significantly to advancements in genetics and genomics research [139].  \\n \\n2.11 MOTIVATION OF THE STUDY  \\nThe motivation behind this research stems from a deep -seated commitment to revolutionize mulberry \\ncultivation practices, particularly in the context of Indian varieties of Morus indica . The current \\nagricultural landscape necessitates a multidimensional exploration into the genomic makeup of these \\nvarieties to enhance their cultivation methodologies effectively [140,141].  \\nOur endeavor  is driven by the pressing need to address key challenges faced by mulberry growers, \\nincluding resilience to drought conditions, optimization of leaf yield, enhancement of nitrogen \\nefficiency, and resistance to root rot. These challenges significantly impa ct agricultural productivity \\nand sustainability, underscoring the importance of genetic insights in informing targeted breeding \\nefforts and selection strategies.  \\nFurthermore, the lack of a comprehensive reference genome or datasets for these unique Morus indica \\nvarieties presents a significant gap in knowledge. This research seeks to bridge that gap by employing \\ncutting -edge techniques such as genotyping, consensus generation, linkage mapping, SNP generation, \\nSSR markers, and microarray chip development. By pioneering the utilization of a draft reference \\ngenome for computational analyses, we aim to unlock vital ge netic information that can drive \\ntransformative advancements in mulberry cultivation.  \\nAt the heart of this research lies a commitment to innovation and progress. The development of a \\nspecialized gene chip tailored to mulberry cultivation represents a groundbreaking advancement, \\npromising to provide growers with actionable insights to optimi ze their practices and enhance crop \\nyields. Moreover, the establishment of an expansive genetic database will serve as a cornerstone for \\nfuture research endeavors , fostering continued exploration into the genetic intricacies of Morus indica  \\n[142] . \\nDespite the formidable challenges posed by the scarcity of genetic data and the absence of a reference \\ngenome, our dedication remains unwavering. We believe that the anticipated findings of this research \\nhave the potential to catalyze  a paradigm shift in mulberry cultivation practices, not only within India \\nbut also on a global scale. By fostering sustainability and resilience in sericulture industries worldwide, \\nwe aim to make a meaningful and lasting impact on agricultural landscapes . \\nIn essence, this research is driven by a passion for innovation, a commitment to addressing pressing \\nagricultural challenges, and a vision for a more sustainable and resilient future for mulberry cultivation.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 50}, page_content='51   \\n \\n2.12 SCOPE OF THE STUDY  \\nThe proposed research aims to pioneer a multidimensional exploration into the genomic landscape of \\nIndian varieties of Morus indica , with a particular emphasis on enhancing mulberry cultivation \\npractices. By employing cutting -edge techniques such as genotyping, consensus generation, linkage \\nmapping, SNP generations, SSR markers, and microarray chip development, the research endeavors to \\nunravel the genetic intricacies underlying four distinct phenotypes: S1, Thailand Male, Punjab Local, \\nand Assama Bola. Add itionally, the investigation will delve into four specific genotypes characterized \\nby traits crucial for agricultural productivity: resilience to drought conditions, optimization of leaf yield, \\nenhancement of nitrogen efficiency, and resistance to root rot . Given the dearth of a reference genome \\nor comprehensive datasets for these unique Morus indica varieties, the research represents a pioneering \\neffort in utilizing a draft reference genome to undertake sophisticated computational analyses. The \\noverarching  objective is to harness genomic insights to engineer a specialized gene chip, tailored to \\naugment mulberry cultivation practices and stimulate agricultural growth. The research objectives are \\nmultifaceted and encompass various critical facets. Firstly, th e endeavor seeks to unravel the intricate \\ngenetic makeup of the target phenotypes, aiming to discern common genetic elements through \\nconsensus sequence generation. Subsequently, the construction of detailed genetic maps will elucidate \\nthe intricate relatio nship between key traits and molecular markers, including the identification of \\nSingle Nucleotide Polymorphisms (SNPs) and Simple Sequence Repeats (SSRs) [143,144]. This phase \\nholds the potential to unveil crucial insights into the genetic architecture gov erning mulberry traits, \\nfacilitating targeted breeding efforts and informed selection strategies. A pivotal aspect of the research \\ninvolves the development of a bespoke microarray chip, engineered to enable rapid and comprehensive \\ngene expression profiling . This transformative technology promises to revolutionize mulberry \\ncultivation practices by providing growers with actionable insights to optimize cultivation \\nmethodologies and enhance crop yields. Furthermore, the research aims to establish an expansive \\ngenetic database, integrating genetic mapping data, microarray chip results, analysis outcomes. This \\ncomprehensive repository will serve as a cornerstone for ongoing research endeavors, providing a \\nrobust foundation for further exploration into the genetic  intricacies of Morus indica . Despite the \\nformidable challenges posed by the scarcity of a draft reference genome and the limited existing data \\non these specific Morus indica varieties, the research remains steadfast in its commitment to making \\nsubstantial contributions to the burgeoning field of genomics and mulberry cultivation. The anticipated \\nfindings hold the promise of catalyzing a paradigm shift in mulberry cultivation p ractices, not only \\nwithin India but also across global agricultural landscapes, f ostering sustainability and resilience in'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 51}, page_content='52  sericulture industries worldwide [145 -148].  \\n \\n2.13 RESEARCH GAP  \\nDespite the importance of Morus indica  in agriculture and sericulture, there is a notable absence of a \\ncentralized database containing the complete genome sequence of Indian varieties. Existing genomic \\nresources primarily focus on non -Indian mulberry species, which may not fully capture the ge netic \\ndiversity present in Indian populations. The availability of molecular markers tailored to Indian \\nmulberry varieties is limited. Existing marker databases predominantly comprise markers designed for \\nnon-Indian genetic backgrounds, which may not accur ately reflect the genetic architecture of Indian \\nMorus indica  populations [149,150]. As a result, the utility of these markers for assessing genetic \\ndiversity and guiding crop improvement efforts in Indian contexts is compromised. Despite \\nadvancements in genomic research, there is a significant gap in translating re search findings into \\nactionable insights for farmers. Farmers lack access to information regarding the traits of interest and \\nstrategies for crop improvement based on genomic insights [151]. The absence of user -friendly tools, \\nsuch as a chip for assessing genetic diversity, further hinders farmer engagement and adoption of \\ngenomic technologies in mulberry cultivation. Bridging the gap in genomic resources and molecular \\nmarker information for Morus indica  is crucial for accelerating crop improvement initiatives and \\nenhancing agricultural productivity in India. Establishing an Indian -specific database will not only \\nadvance our understanding of the genetic basis of desirable traits but also empower fa rmers with \\nactionable insights for sustainable mulberry cultivation. This research endeavor represents a vital step \\ntowards harnessing the full potential of genomic technologies to benefit Indian agriculture \\nand sericulture [152,153].'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 52}, page_content='53   \\n \\n \\nOBJECTIVES OF THE STUDY  \\n \\n \\nBased on the Research gap and the Scope of the study, 3 Main objectives were framed \\nfor the Research study.  \\n1) Development of database with the curated SNP and  SSR Markers  \\n2) Construction of the linkage maps with the  Markers  \\n3) Development of SNP Genotyping  Chip'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 53}, page_content='54  CHAPTER 3  \\nRESEARCH METHODOLOGY  \\n \\n \\nFigure 7: Overview of the Methodology followed'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 54}, page_content=\"55  OBJECTIVE 1: Development of database with the curated SNP and SSR Markers  \\n3.1 Whole genome sequencing of the Mulberry data  \\nWhole genome sequencing (WGS) with Illumina technology offers a powerful approach for exploring \\nthe genetic makeup of Morus indica, the mulberry tree. This species holds immense value in silk \\nproduction and traditional medicine. Illumina's high -throughput sequencing provides extensive \\ngenomic coverage rapidly and affordably. Its accuracy ensures precise reconstruction of the genome, \\ncrucial for identifying genetic variations like SNPs. Moreover, the scalability of Illumina sequencing \\nallows researchers to t ailor their study objectives, whether uncovering genetic markers or tracing \\nevolutionary paths. Ultimately, this method enhances our understanding of Morus indica's  genetic \\nlandscape, unlocking opportunities for agriculture and medicine.  \\nThe whole genome sequencing of the Mulberry was carried out using Illumina sequencer. Total of 16 \\ngenotypes were sequence and 4 were selected for the analysis due to their correlation with the \\nphenotypes selected i.e, Root rot, Nitrogen use efficiency, Drought and Leaf yield. The details of which \\nare shared in the table 2.  \\nK2 is renowned for its high yield and adaptability to diverse climates, stemming from a cross between \\nlocal Indian rice varieties. V1 stands out for its disease resistance, particularly against brown spot and \\nsheath blight, appealing to farmers seeking rob ust yields. S13 excels in drought resistance, crucial for \\ncultivation in arid regions. AR12 showcases resistance against pests like the rice borer, ensuring crop \\nprotection. Similarly, C176, S34, S1635, and S1 offer disease and pest resistance, enhancing y ield \\nreliability. RFS175 further fortifies against pests, particularly the rice borer. M. multicaulis , or Mung \\nbean, enriches soil fertility through nitrogen fixation, boosting overall crop productivity. Himachal \\nLocal, native to Himachal Pradesh, India, exhibits resilience against diseases and pests. BR -8, like other \\nvarieties, combines disease resistanc e with high yield potential. These varieties collectively empower \\nfarmers with options tailored to environmental challenges, ensuring sustainable agricultu ral practices \\nand food security  [154] . The correlation of the genotype and phenotype are mentioned in the table 1.  \\n \\nTable 1 The Genotype details  \\n       \\nGenotype \\nname  Drought    Leaf Yield  NUE  Root rot  \\nK2 Medium  Low - Moderately Resistant  \\nV1  High  High  High  Highly susceptible  \\nS13  Medium  Medium  Moderate  Resistant  \\nAR12  Not included  Medium to high  - Resistant  \\nC176  Not included  High  - Moderately Resistant  \\nS34  Not included  Medium  High  Resistant\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 55}, page_content='56  S1635  Not included  High   Resistant  \\nS1  Not included  Medium  High  Highly susceptible  \\nRFS175  High  High  Low Resistant  \\nM. \\nmulticaulis  Not included   - Highly Resistant  \\nHimachal \\nLocal  Not included  Low - Resistant  \\nBR-8  Not included  Low - Moderately Resistance  \\n \\n3.2 Retrieval of raw mulberry data  \\nThe high -depth sequencing of multiple mulberry accessions via Illumina platform was performed by \\nJawaharlal University.  A total of 21 genotypes were sequenced and deposited in the Morus indica \\nGenome project ( http://tgsbl.jnu.ac.in/MindGP/ ). Out the 21 genotypes 4 were selected for analysis \\nwhich had impact on the phenotypes i.e. Root rot, Leaf yield, Nitrogen use efficiency and Drought. The \\ndetails of the phenotypes are mentioned in the table below:  \\nTable 2 Phenotype details of the samples collected  \\nGenotype  Drought  Leaf Yield  Nitrogen use \\nEfficiency  Root rot  \\nS1 Low High  Medium  Highly \\nSusceptible  \\nThailand Male  Low  Medium  Moderate  Highly \\nSusceptible  \\n \\nPunjab Local  High  Medium  Nil Highly \\nSusceptible  \\n \\nAssama Bola  Medium  High  Low Moderate  \\n \\n3.3 Quality checks of the reads  \\nWe wanted to make sure the raw data we gathered was up to scratch, so we ran it through a quality \\nassessment using tools like FastQC and MultiQC. FastQC is great for giving us a quick overview of'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 56}, page_content=\"57  any issues in the raw sequencing data. We checked all four genotypes of Morus indica to see if there \\nwere any problems like inconsistent quality across sequences or unusual nucleotide content. After \\nrunning these checks, we knew which datasets were good to go for further analysis. FastQC also helped \\nus spot any over -represented sequences, which might indicate contamination from adapters or primers. \\nThis detailed report gave us insight into whether we needed to do any additional steps like trimming \\nbases, cl ipping adapters, or filtering reads before moving on to alignment.  \\n 3.4 Gapped alignment  \\nTo make sure we accurately and quickly matched up the pre -processed reads with the reference \\ngenome, we used a process called alignment mapping. There are different tools available for this task, \\nbut we chose BWA (Burrows -Wheeler Aligner) because it's known for its speed and efficiency. BWA \\nuses a clever technique called the Burrows -Wheeler Transformation along with the Smith -Waterman \\ndynamic programming algorithm to align the short reads with the reference genome.  It has three \\ndifferent algorithms for this job: BWA -backtrack, BWA -MEM, and BWA -SW. By default, BWA looks \\nfor alignments within a certain distance from the query sequence, but it can also be adjusted to find \\nlonger gaps if needed, although this might slow  things down a bit and increase the chance of incorrect \\nmatches. The reason we went with BWA for this task is that it's really fast. It can handle about two \\nmillion high -quality short reads and map them against the mulberry genome in just around 30 minutes . \\n3.5 Refining the alignments and conversion by SAM tools  \\nAfter aligning the cancer exome short reads to the reference genome, we wanted to enhance the quality \\nof the alignments and minimize the chance of errors in variant calling. To achieve this, we ran the \\nalignment output through several refining steps using tools designed for processing SAM/BAM files. \\nThere are various software options available for handling SAM/BAM files, including SAM tools, \\nGenome Analysis Toolkit (GATK), and Picard. These tools, developed by institutes like Broad and \\nSanger, are commonly used for operations on alignment data. In our study, we refined the alignments \\nfor all twenty datasets. This involved tasks such as sorting the reads, recalibrating quality scores, \\nmarking PCR and optical duplicates, realigning indels, and filtering the re ads. SAM tools is a versatile \\ntoolkit specifically designed for manipulating and analyzing alignment data in SAM/BAM format. It \\ncan handle tasks like converting from other alignment formats, merging and sorting alignments, \\nremoving PCR duplicates, calling variants like SNPs, and providing per -position information. We used \\nSAM tools to convert the short -read alignment output from SAM to BAM format. Then, we sorted and \\nmerged the BAM files. SAM format is widely used due to its simplicity and flexibility, whil e BAM, \\nbeing a binary representation, offers a more compact size and faster retrieval of alignments. By \\nleveraging positional indexing, sorting, and merging, SAM/BAM files allow for efficient processing of \\ngenomic regions without loading the entire file in to memory. This approach enables an integrated\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 57}, page_content='58  analysis of genome sequence data, separating the alignment step from downstream analyses. Therefore, \\nour study employed SAM tools to convert SAM to BAM format, facilitating subsequent analyses.  \\n \\nFigure 8: Protocol for Mapping and SNP calling  \\n3.6 Variant processing  \\nFirst, before identifying any mutations or variants from the sequenced data, we go through some initial \\nsteps. This includes indexing the reference genome using tools like BWA and SAM tools, as well as \\nchecking the quality of the sequences using FASTQC. Af ter that, we merge the alignments in BAM \\nformat. Once we have these preliminary steps done, we move on to variant calling and processing. In \\nour study, we started this process using two main tools: PICARD and GATK. GATK, short for Genome \\nAnalysis Toolkit, is a framework designed to make it easier to develop effective tools for analysing \\nDNA from next -generation sequencing (NGS). It offers a wide range of data access patterns needed for \\nmost analyses, and its structure allows for optimization in terms of pre cision, memory efficiency, and \\nstability. GATK also provides tools for calculating coverage and identifying single nucleotide \\npolymorphisms (SNPs). In our study, we used GATK to get rid of PCR duplicates, which is essential \\nfor efficient and accurate analy sis, especially in large -scale projects. On the other hand, PICARD is a \\nset of command -line tools commonly used for handling high -throughput sequencing data and formats \\nlike SAM, BAM, CRAM, and VCF. Its file formats are well -defined in the HTS -specs reposi tory. In \\nour study, we used PICARD to mark PCR duplicates and perform local realignment and base quality \\nrecalibration using parameters suited for our analysis. We also used PICARD for tasks like sorting \\nSAM files, recalibrating bases, and building BAM ind exes. Once we analyzed the covariates, we \\nproceeded to call the variants.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 58}, page_content='59  3.7 Variant calling  \\nAfter processing the variants, we proceeded to call them. To do this, we first piled up the mapped data \\nand counted the bases using SAMtools. Then, we used GATK to identify and extract the SNPs and \\nindels from the variants. Once we had all the variants ide ntified, we focused on calling the SNPs. Next, \\nwe extracted the SNPs and indels and filtered them to ensure accuracy. We then used a tool called \\nsnpEFF to annotate the filtered SNPs and indels. SnpEFF is handy because it not only annotates the \\nvariants but  also predicts the effects of these genetic changes, including any alterations in amino acids. \\nSnpEFF takes in variant call format (VCF) files containing the predicted variants, such as SNPs and \\nindels, and outputs its annotations in the default VCF format . In our study, we carefully analyzed the \\nidentified SNPs and scrutinized the output from snpEFF. One of the advantages of using snpEFF is its \\nspeed —it can annotate up to 1,000,000 mutations per minute, which is quite impressive. Additionally, \\nit supports over 2500 genomes, ranging from plants and animals to bacteria and fungi. Another benefit \\nis its seamless integration with GATK and Galaxy toolkits. Overall, we found snpEFF to be a valuable \\ntool for both variant calling and annotation in our study.  \\n \\nFigure 9: Best Practices  for GATK pipeline for calling the variants  \\n \\n3.8 Variant Filtration  \\nUnderstanding the link between genotype and phenotype is a key scientific challenge, and the ability \\nto anticipate phenotypes based on molecular genotypes is critical in molecular breeding. Whole genome \\nduplications have affected the history of all floweri ng plants and pose difficulties in determining the \\nlink between genotype and phenotype, particularly in polyploid species. Although single nucleotide'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 59}, page_content=\"60  polymorphisms (SNPs) have become popular tools for genetic mapping, finding and using SNPs in \\npolyploids has proven challenging.  These findings are expected to differ amongst species; hence we \\nsuggest a set of best practices for SNP calling in polyploids.  \\nFiltering criteria used in variant calling analysis, particularly for single nucleotide polymorphisms \\n(SNPs). Each criterion targets specific aspects of variant quality and potential artifacts within the data. \\nVariants with low confidence due to factors li ke poor quality by depth (QD < 2.0), high strand bias (FS \\n> 60.0 and SOR > 4.0), low mapping quality (MQ < 40.0), and discrepancies in mapping quality \\nbetween reference and alternate alleles (MQRankSum < -12.5) are filtered out. Additionally, variants \\nwher e the alternate allele is found at the ends of reads more frequently than the reference allele \\n(ReadPosRankSum < -8.0) are also removed. These filtering steps help ensure the reliability of variant \\ncalls by removing potential false positives caused by tech nical artifacts or sequencing errors.  \\n \\n3.9 Identification of SSR Markers  \\nGenome Wide Microsatellite Analyzing Tool was used for analyzing the gene wise Markers  \\nData Preparation  \\nThe genomic data utilized in this study were obtained in FASTA format, providing the foundational \\nsequences for the organism under investigation. The GMATA software was subsequently installed on \\nthe local computer system to facilitate the analysis.  \\nRunning GMATA  \\nTo initiate the SSR identification process, a terminal or command prompt was accessed on the \\ncomputer. The terminal was navigated to the directory where the GMATA software was installed, \\nensuring that the tool was readily accessible for analysis.  \\nIdentifying SSRs  \\nThe cornerstone of SSR identification within GMATA is the Microsatellite identification tool (MISA) \\nmodule. This algorithm systematically detects SSRs within the genomic sequences based on user -\\ndefined parameters. Specifically, repeat unit length and the m inimum number of repeats were \\ncustomized to suit the requirements of our investigation . \\n \\n3.10 DATABASE DEVELOPMENT  \\n3.10.1 Frontend Development  \\nThe frontend of the Mulberry Database project facilitates seamless retrieval of genetic data based on \\nspecific key -value pairs, enhancing user interaction and data exploration. Here's a detailed overview of \\nthe frontend architecture, functionalities, and u ser experience:\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 60}, page_content=\"61  Architecture  \\n3.10.2  Next.js and Tailwind CSS : The frontend leverages Next.js and Tailwind CSS to create a \\nresponsive and visually appealing user interface. Next.js enables server -side rendering for \\nenhanced performance, while Tailwind CSS provides a utility -first approach for streamlined \\nstyling.  \\n● Axios for HTTPS Requests : Frontend -to-backend communication is managed via Axios, \\nenabling asynchronous retrieval of genetic data based on user input.  \\n● Data Retrieval Based on Key -Value Pairs  \\nThe frontend enables users to retrieve genetic data by specifying key -value pairs, such as selecting the \\n'id' attribute and providing the value 'Mi00088'.  \\nFor instance, upon submitting the search query for the specified key -value pair, the frontend displays \\nthe following result:  \\n \\n \\nFigure 10 The query flow for fetching data (gene sequences) from database  \\nThe frontend interface restricts the display to values pertinent to the search query, ensuring that users \\nreceive focused and relevant genetic data matching their criteria.  \\nUsers can easily select attributes and input corresponding values to retrieve specific genetic sequences, \\nenhancing the efficiency and effectiveness of data exploration and analysis.  \\n• Hosting Platform  \\nThe frontend application is hosted on Netlify , a reliable hosting platform offering continuous \\ndeployment and scalability capabilities.\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 61}, page_content=\"62  Netlify's seamless deployment process ensures that updates and enhancements to the frontend are \\npromptly reflected in the live application environment, providing users with a consistent and up -to-date \\nexperience. The application is live on mulbery_application  \\n3.10.3 Backend Development  \\nThe backend of the Mulberry Database project, powered by Node js  and Express js , orchestrates \\ncommunication between the frontend and the MySQL database hosted on Aiven Cloud.  Here's an \\noverview of the backend architecture, functionalities, and hosting platform:  \\nArchitecture  \\nNode.js and Express : The backend utilizes Node js and Express js  to create a robust and scalable server \\nenvironment, capable of handling incoming HTTPS  requests from the frontend.  \\nMySQL Database Interaction : The backend interacts with the MySQL database hosted on Aiven \\nCloud, executing queries and managing data retrieval and insertion operations.  \\n3.10.4 Advantages  \\n● Efficient Data Management : The backend facilitates seamless data management, allowing \\nusers to upload genetic data via CSV files  and add values to the database securely. The backend \\nensures data integrity and consistency, maintaining the reliability of genetic information stored \\nin the database.  \\n● Security and Access Control : Authentication mechanisms implemented in the backend enforce \\nsecurity measures, preventing unauthorized access and manipulation of sensitive genetic data. \\nBy implementing access controls, the backend enhances data confidentiality and privacy.  \\n3.10.5 Hosting Platform  \\nThe backend application is deployed on Render.com ’s free service tier, providing a reliable and \\nscalable hosting platform for Node.js applications. Render.com's managed infrastructure simplifies \\ndeployment and maintenance tasks, ensuring high availability and performance of the backend server.  \\n 3.10.6 Database Management  \\nThe Mulberry Database project relies on MySQL hosted on Aiven Cloud to manage and store genetic \\ndata efficiently. Here's a detailed breakdown of the database architecture, features, and advantages:  \\n3.10.7 Database Structure and Attributes  \\nThe Mulberry Database comprises a single table named MULBERRY_DB with the following \\nattributes:  \\nsl: An integer field serving as the primary key with auto -increment functionality.  \\ngene : A varchar field with a maximum length of 255 characters, allowing for the storage of gene \\nidentifiers.\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 62}, page_content=\"63  from1 : An integer field representing the starting position of a genetic sequence.  \\nto1: An integer field representing the ending position of a genetic sequence.  \\nid: A varchar field with a maximum length of 255 characters, potentially storing additional identifiers.  \\ns1: A text field for storing textual genetic data.  \\nAssama_Bola : A text field for specific genetic attributes.  \\nThailand_Male : A text field for specific genetic attributes.  \\nPunjab_Local : A text field for specific genetic attributes.  \\nssr_repetitions : A varchar field for storing repetitions of specific genetic sequences.  \\nssr_motif : A varchar field for storing motifs of specific genetic sequences.  \\nGene_ontology_IDs : A text field for storing gene ontology identifiers.  \\nMolecular_Function : A text field for storing molecular function descriptions.  \\n3.10.8 Database Size and Performance  \\nSize: The Mulberry Database table contains approximately 26,964 records , comprising genetic data \\ntotaling around 1.2 gigabytes . This size demonstrates the scale and complexity of the genetic \\ninformation managed by the database.  \\nPerformance : MySQL on Aiven Cloud offers robust performance and scalability, ensuring optimal \\nquery execution and data retrieval even with large datasets. With horizontal scalability and automatic \\nresource provisioning, the database seamlessly accommodates growing da ta volumes and user demands \\nwithout sacrificing performance.  \\n3.10.9 Advantages of MySQL on Aiven Cloud  \\n• Scalability and Performance : Aiven Cloud's MySQL service provides horizontal scalability, \\nenabling the Mulberry Database to handle increasing data volumes and user traffic efficiently. \\nAutomatic resource provisioning ensures optimal performance and responsiveness, even under \\nheavy loads.  \\n• Data Security and Compliance : Aiven Cloud implements advanced security measures, \\nincluding encryption at rest and in transit, to safeguard the confidentiality and integrity of \\ngenetic data stored in the MySQL database. Compliance with regulatory standards such as \\nGDPR and HIPAA ensur es adherence to data protection requirements and user privacy.  \\n• Managed Services : Aiven Cloud's managed database services simplify database administration \\ntasks, including monitoring, backups, and maintenance. Automated management features \\nreduce operational overhead and ensure the reliability and availability of the Mulberry Database .\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 63}, page_content=\"64  3.10.10 Hosting Platform  \\nThe MySQL database is hosted on Aiven Cloud, a leading provider of managed database services. \\nAiven Cloud's scalable infrastructure and comprehensive management capabilities make it an ideal \\nplatform for hosting critical databases like the Mulberry Databas e. With high availability, data \\ndurability, and security features, Aiven Cloud ensures the reliability and integrity of genetic data stored \\nin the database.  \\n \\n \\nFigure 11: The flow for the Database development  \\n \\nObjective 2: Construction of the linkage maps with the Markers  \\n3.11 Consensus Generation  \\nTo generate a consensus sequence based on variant calls from a VCF (Variant Call Format) file and a \\nreference genome, several steps are typically involved. Here's a write -up outlining the methods \\nfollowed for consensus generation based on the provided comm ands:  \\nVariant Calling and Filtering  \\nThe first step involves variant calling from aligned sequencing data (e.g., from DNA sequencing reads). \\nVariant calling identifies positions in the genome where individuals in the sample differ from the \\nreference genome.  \\nThe file vcf likely contains variant calls, specifically Single Nucleotide Polymorphisms (SNPs), filtered \\nto retain high -quality variants. This file may have been generated using tools such as GATK,\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 64}, page_content='65  Compression and Indexing : \\nThe bgzip  command is used to compress the VCF file and give the output in the form of vcf.gz  \\nFollowing compression, the tabix  command is applied to index the compressed VCF file for efficient \\nretrieval of variants. The index is generated in VCF format.  \\nReference Genome Preparation : \\nThe mulberry reference genome is prepared for use in generating the consensus sequence. This file \\nlikely contains the entire genomic sequence of the organism under study, broken down into contigs or \\nscaffolds.  \\n \\nConsensus Sequence Generation : \\nThe samtools faidx  command indexes the mulberry reference genome to facilitate random access to \\nspecific regions of the genome.  \\nFollowing indexing, the bcftools consensus  command is utilized to generate a consensus sequence for \\na specific region of interest. In this case, the command generates a consensus sequence for \"Super -\\nScaffolds\" based on the provided reference genome and variant calls from the SNP and indel VCF file s \\n(the filtered snps in vcf.gz ) \\nThe output of this command is redirected ( >) to a FASTA file named Contig .fa, which contains the \\nconsensus sequence for the specified region.  \\nIn summary, the methods involve variant calling and filtering, compression and indexing of VCF files, \\npreparation of the reference genome, and generation of a consensus sequence based on the reference \\ngenome and variant calls using specific commands from t he bcftools and samtools software packages. \\nThese steps enable the creation of a consensus sequence that reflects the most likely genomic sequence \\nfor the specified region, considering both SNPs and indels detected in the sequencing data.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 65}, page_content='66   \\nFigure 12: Workflow for Consensus generation with the SNPs  \\n \\n3.12 Designing the primer sequences  \\nOnce the consensus has been developed, the primers were designed for the SNPs obtained using \\nPrimer3 design suite. To obtain the right set of primer sequences, the following parameters were set:  \\na) Primer minimum size was set to 18 and maximum was set to 30  \\nb) The melting temperature was set to 52  \\nc) Primer GC content was set to 20  \\nThe parameters were set and 25 bases before and after the SNP position (considered as flanking \\nsequences) was given as a sequence for the design of primers.  \\n \\n3.13 Gene ontology studies and genotype -phenotype relation  \\nGene Ontology (GO) analysis is an essential component of our study, facilitating a systematic \\nexploration of gene function and biological processes associated with key phenotypic traits in Morus \\nindica  varieties. QuickGO, a user -friendly online tool provided by the Gene Ontology Consortium, is \\nutilized for this purpose. The Pathways relating to the phenotypes considered in the study were \\nconsidered. A correlation matrix, depicting the relation among the phenotypes was drawn.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 66}, page_content=\"67  3.14 LINKAGE GROUPS  \\nTo prepare for genetic mapping, we begin by creating a mapping population, typically through crosses \\nbetween different Morus indica varieties. This population, often bi -parental (like F2, RILs, or DH \\nlines), ensures genetic diversity for robust linkage mapping outcomes. It's crucial to ensure a \\nsufficiently large population size to capture the genetic variation adequately. Next, we gen otype the \\nmapping population using molecular markers suitable for linkage mapping, such as SSRs or SNPs. \\nThese markers are chosen based on their polymorphism and genome coverage to provide \\ncomprehensive genetic information. The marker data is then formatte d to be compatible with QTL \\nIciMapping Version 4.2, the software we'll use for analysis. With the marker data in hand, we utilize \\nQTL IciMapping Version 4.2 to construct linkage maps. This involves selecting appropriate mapping \\nalgorithms and parameters to  accurately represent the genetic distances between markers. The quality \\nof the constructed linkage maps is assessed, ensuring marker order and map distances are reliable for \\nsubsequent analysis.  \\n \\nObjective 3: Development of SNP Genotyping Chip  \\n3.15 CHIP DEVELOPMENT  \\nTo begin designing an effective SNP genotyping assay, it's essential to clearly define the objectives of \\nthe assay. This involves identifying the specific SNPs of interest, understanding their genomic \\npositions, allelic variants, and any relevant flanking sequences. Accuracy and completeness of the SNP \\ndata are crucial for precise assay design.  \\nOnce the SNP data is gathered, accessing the Design Studio Assay Design Tool becomes the next step. \\nThis tool, provided by the manufacturer or developer, offers a platform specifically tailored for SNP \\ngenotyping. Logging in and accessing the Assay Design Tool within Design Studio facilitates the \\ncustomization of assay parameters according to the requirements of the study.  \\nAssay parameters such as SNP positions, primer sequences, probe sequences (if applicable), amplicon \\nsize, and melting temperature (Tm) need to be specified. It's important to consider the genotyping \\nplatform and the specific requirements for primer and pro be design. For instance, if using a PCR -based \\nassay or microarray probes, compatibility with these platforms should be ensured for optimal assay \\nperformance.  \\nBy carefully defining objectives, preparing accurate input data, and accessing the appropriate assay \\ndesign tool with tailored parameters, researchers can lay the foundation for designing SNP genotyping \\nassays that meet their research goals effectively and  accurately.\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 67}, page_content='68  CHAPTER 4  \\nRESULTS AND DISCUSSION  \\n4.1 Raw Data Pre -processing  \\nThe raw sequencing data obtained from Morus indica , encompassing four distinct genotypes, \\nunderwent a meticulous quality control process. The initial quality assessment revealed. Subsequently, \\nthe raw data was subjected to quality filtering and adapter trimming to ensure data integrity and \\nreliability.  \\n4.2 Quality Checks and Reads  \\nQuality assessment of the raw sequencing data retrieved from all four genotypes of Morus indica  was \\nconducted using FastQC and MultiQC toolkit. FastQC provided a comprehensive analysis, evaluating \\nvarious aspects of the raw sequence data, including mean sequence quality per base position, per read, \\nnucleotide content per base position, GC distributi on, and other key metrics.  \\n \\nFigure 13: Quality Control Status for Sequencing Data'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 68}, page_content='69   \\n \\n \\n  \\n  \\n  \\n  \\n  \\n  \\n  \\n  \\n Figure 14: The image depicts the distribution of mean sequence quality scores, which are \\nassigned to individual DNA sequences obtained from high-throughput sequencing devices. It \\nillustrates that as the mean sequence quality (Phred Score) rises, the number of sequences \\nsharply declines. This indicates a scarcity of sequences with higher quality scores. The graph \\neffectively represents the pred ominance of good -quality scores among the sequences utilized in  \\nthe study.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 69}, page_content='70   \\n  \\n  \\n  \\n  \\n  \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 1 5: Sequence Counts of FastQC  \\n \\n \\nThe x -axis corresponds to sequence  (genotypes) , ranging from 0 to 80 million reads. The blue portion \\nof each bar represents unique reads. These are individual sequences that occur only once in the dataset. \\nUnique reads are essential for accurate data analysis, providing genuine information about the s ample. \\nThe black portion of each bar represents duplicate reads. These are sequences that appear more than \\nonce in the dataset. Duplicate reads can arise due to technical artifacts during sequencing or PCR \\namplification biases. Assessing the balance betwee n unique and duplicate reads is crucial for data \\nquality. High -quality data should have a reasonable proportion of unique reads and minimal \\nduplication. DNA sequencing generates millions of reads, resulting in gigabytes of data. Efficient \\nanalysis tools ar e necessary to make sense of this high -throughput sequencing (HTS) data.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 70}, page_content='71  4.3 Gapped Alignment  \\nIn the present investigation, we conducted a proficient alignment mapping of Morus indica  exome \\nreads, which were pre -processed, to the genome utilizing the Burrows -Wheeler Aligner (BWA). By \\nemploying BWA -backtrack, BWA -MEM, and BWA -SW algorithms, our methodology ensured \\naccurate alignments within an edit distance of 2 to the query sequence. F urthermore, we assessed the \\ndata using Bowtie2, a highly efficient tool recognized for its proficiency in aligning short reads to \\nextensive genomes, including that of Morus indica . The resulting SAM -formatted output is compatible \\nwith various downstream tools, especially the Genome Analysis Toolkit (GATK), facilitating \\nsubsequent comparative genomics studies. This strategic alignment lays the foundation for in -depth \\nanalyses, such as variant calling, offering valuable insights into the genomic intricacies of Morus \\nindica . Importantly, all samples showcased robust alignment, each achieving a notable alignment rate \\nof 85% or higher.  \\n \\nTable 3: The table clearly shows the genotypes and their corresponding alignment percentages  \\nSample ID  Genotypes   Alignment Percentage \\n(%) \\nSRR14506990  S1 87% \\nSRR14506999  Punjab Local  85% \\nSRR14507001  Assama Bola  86% \\nSRR14507002  Thailand Male  85% \\n \\n4.4 Sam to Bam conversion  \\nImproved the alignment output to ensure high -quality alignments and reduce the risk of false variant \\ncalls in the genomic data of Morus indica  (mulberry). We utilized popular software such as SAM tools, \\nGenome Analysis Toolkit (GATK), and Picard to perform various operations on the SAM/BAM files \\ngenerated from the alignment of mulberry from twenty Morus indica  datasets. These operations \\ninvolved sorting, recalibrating quality scores, marking PCR and optical duplicates, realigning indels, \\nand filtering reads . SAM tools were particularly important for converting the alignment output to BAM \\nformat, as well as for sorting and merging BAM files. The choice of the SAM/BAM format was based \\non its simplicity, flexibility, and compatibility with data from different s equencing platforms. The \\nbinary BAM format, known for its small size and positional indexing, enabled quick retrieval of'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 71}, page_content=\"72  alignments in specific regions, ensuring efficient downstream processing. This strategic use of SAM \\ntools demonstrates an integrated approach to the thorough examination and analysis of large -scale \\ngenomic datasets, laying a strong groundwork for subsequen t analyses, including variant calling and \\ncomprehensive exploration of the genomic landscape of Morus indica . \\n4.5 Variant Processing  \\nThe study successfully processed the variants using a combination of PICARD and GATK. PCR \\nduplicates were identified using PICARD, while local re -alignment and base quality recalibration were \\nperformed using GATK with appropriate parameters. Sort Sam and b uilding BAM indexes were also \\nexecuted using PICARD. Once the covariates were analyzed, the variants were called.  \\n4.6 Variant Calling  \\nFollowing the variant processing steps, the identified variants were called utilizing SAMtools for pileup \\nof mapped data and base counting, followed by variant calling using GATK. The total variants were \\nidentified, with a specific focus on SNPs. Subsequen tly, the SNPs and indels were extracted and \\nsubjected to filtering. The filtered SNPs and indels underwent annotation using snpEFF, a tool adept at \\nannotating variants and predicting their effects on known genes. The analysis revealed comprehensive \\ninsight s into the identified SNPs, facilitated by snpEFF's efficient annotation capabilities. Leveraging \\nits compatibility with various genomes and rapid processing speed, snpEFF proved to be a valuable \\nasset in annotating many mutations per minute.  \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 1 6: The graph represents the sample types and the corresponding SNP generated concerning \\nthe sample  \\n \\n \\n \\n \\n05,00,00010,00,00015,00,00020,00,00025,00,00030,00,00035,00,00040,00,00045,00,00050,00,000\\nS1 Punjab Local Assama Bola Thailand MaleSNP's\\nSample Genotype\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 72}, page_content='73  However, managing the sheer complexity of the data can pose significant challenges, especially when \\ndealing with millions of SNPs generated across multiple individuals. To simplify this intricate dataset, \\nit often turns to generating consensus sequences. T hese consensus sequences distill the wealth of SNP \\ninformation into a more manageable form by representing the most common nucleotide observed at \\neach position across all individuals. Typically visualized with the x -axis denoting different sample \\ngenotypes  and the y -axis indicating SNP positions, consensus sequences provide a simplified overview \\nof genetic variation, with lines or curves illustrating the predominant nucleotide at each position. By \\nfocusing on these consensus sequences, we could more effecti vely analyze the dataset, identifying \\nprevalent variations, uncovering patterns, and elucidating associations with traits or diseases. \\nUltimately, this approach facilitates a deeper understanding of genetic diversity within populations, \\nenabling researcher s to extract meaningful insights from complex genomic data.  \\n4.7 Identifying SSRs  \\nResults obtained from the SSR identification process concerning mulberry genetics would provide \\ninsights into the distribution and characteristics of microsatellite markers within the mulberry genome, \\nwhich could be valuable for various genetic studies, br eeding programs, and conservation efforts related \\nto mulberry species.  \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 17: The graph displays the frequency count of SSRs (Simple Sequence Repeats) on different \\nchromosome segments or types. The Y -axis represents the frequency in counts per megabase \\n(Count/Mb), ranging from 0 to 6000. The X -axis labels correspond to various chr omosome segments, \\nsuch as “JN108091.1,” “JN126801.1,” “JN144241,” and so forth. There are eight bars in total, each of \\na different color. The bar labeled “JN167781” stands out with the highest frequency count, reaching up \\nto 6000.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 73}, page_content='74   \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 18: The Picture depicts a bar graph showing the top -grouped motifs in DNA sequences. These \\nmotifs are short, recurring patterns in DNA that are presumed to have a biological function. The graph \\nhighlights the total occurrence and diversity of these motifs.  \\n \\n \\n4.8 Database development  \\nThe Indian Mulberry Linkage Mapping Database (IMLM -DB) represents a significant milestone in the \\ngenetic and breeding research of Morus indica , a species with diverse therapeutic and economic \\nimportance. The database focuses on Single Nucleotide Polymorphisms (SNPs) and Simple Sequence \\nRepeats (SSRs), crucial molecular markers increasingly utilized in genetic diversity analysis, marker -\\nassisted selection, genome and QTL mapping, and conservation genetics. A genotyping array \\ncomprising of 25K SNPs has been d eveloped to identify markers associated with specific traits crucial \\nfor mulberry leaf production enhancement. This array has been designed through comprehensive \\nbioinformatic analysis of the whole genome and transcriptome data obtained from the network pr oject \\nNW1. Approximately 350 germplasm accessions have been genotyped using the developed array. \\nPromising trait donor genotypes have been identified through rigorous analysis of genotypic data, \\nfacilitating targeted breeding efforts aimed at enhancing yie ld potential, drought adaptation, and disease \\nresistance in mulberry plants. A comprehensive database of SSR and SNP markers has been \\nestablished, providing researchers with valuable genetic information for various applications in \\nmulberry breeding and con servation. SNP markers identified through whole genome alignment of 20 \\ncontrasting lines have been cataloged, contributing to the development of robust DNA -based molecular'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 74}, page_content='75  markers for accelerating trait introgression and improving mulberry leaf production. The project has \\nprovided genotyping support to other subprojects, enabling the demonstration of effective Linkage \\nDisequilibrium (LD) decay and the construction of a dense  linkage map using codominant markers like \\nSSRs and SNPs.  \\nhttps://mulbery.netlify.app/       \\n \\n \\n4.9 Consensus Generation  \\nThe VCF files yielded an extensive dataset with 5 million SNPs. To distill the most pertinent SNP \\nvariations within a sequence, consensus generation was undertaken. Through this process, a contig file \\nwas produced, encapsulating the most specific SNP varia tions identified from the dataset.  \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 19: Contig file generated after consensus generation  \\n \\n \\n4.10 Primer Design  \\nThe Primer3 Primer design suite was harnessed to craft primers for amplification. Leveraging the start \\nand end regions of the contig file, primers were meticulously designed to target specific regions within \\nthe sequence.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 75}, page_content='76   \\n \\nFigure 20: The picture depicts the primers that are designed using the primer3 design suite  \\n \\nhttps://docs.google.com/spreadsheets/d/1QoLKp1zCKGdwNFQH9RIrk3f0bnHzlDu2/edit?usp=sharing&ouid=\\n114613627565011743972&rtpof=true&sd=true  \\n \\n \\n4.11 Genotype/phenotype relation  \\nGene Mi00022 (GO:0046872 - Response to Water Deprivation), Alleles: Reference allele T, \\nAlternative allele A, Breed: Punjab Local, Phenotype Associated: Drought. The presence of the \\nalternative allele A in gene Mi00022 in the Punjab Local breed suggests a potential association with \\ndrought tolerance or response to water deprivation. This genotype -phenotype correlation implies that \\nindividuals carrying the alternative allele may exhibit enhanced resilience or adaptation to drought \\nconditions compared to thos e with the reference allele T. Gene Mi00043 (GO:0009399 - Nitrogen \\nFixation), Alleles: Reference allele A, Alternative allele T, Breed: Assama Bola, Phenotype Associated: \\nNUE (Nutrient Use Efficiency). The alternative allele T in gene Mi00043, found in the  Assama Bola \\nbreed, is linked to Nitrogen Fixation, which is associated with Nutrient Use Efficiency (NUE). This \\ngenotype -phenotype correlation suggests that individuals carrying the alternative allele may possess \\ntraits conducive to efficient nitrogen uti lization, potentially leading to improved NUE in agricultural \\nsettings. Gene Mi00041 (GO:0016168 - Chlorophyll Binding), Alleles: Reference allele A, Alternative \\nallele G, Breed: S1, Phenotype Associated: Leaf yield. The alternative allele G in gene Mi0004 1, \\nobserved in the S1 breed, is associated with Chlorophyll Binding, a process essential for photosynthesis \\nand leaf development. The correlation with Leaf yield suggests that individuals carrying this allele may \\nexhibit traits conducive to higher leaf pro ductivity, potentially leading to increased overall yield in \\nagricultural contexts. These genotype -phenotype correlations highlight the intricate relationships'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 76}, page_content=\"77  between genetic variations and observable traits in plant populations. Understanding such correlations \\nis essential for targeted breeding efforts aimed at enhancing desirable traits, such as drought tolerance, \\nnutrient use efficiency, and yield, to improve  agricultural sustainability and productivity.  \\n \\n \\nFigure 21: Gene Information and Phenotype Associations  \\n \\n \\nhttps://docs.google.com/spreadsheets/d/1uGmqXDQL1pjPTfvxf_xt9GQMKrV3WfnxDhvCjfyePwQ/edit\\n#gid=0  \\n \\n \\n \\n \\nWhen plants face water scarcity, known as drought stress, it triggers complex changes in their \\nphysiology. This includes alterations in leaf water levels and adjustments to cope with the lack of water. \\nDrought stress reduces photosynthesis and worsens with  severe stress, affecting plant growth and yield. \\nCertain plants tolerate drought better, showing fewer negative effects on their membranes and \\nmaintaining cell stability. These physiological adjustments help plants endure drought conditions, but \\nsevere st ress can still reduce crop yield. Breeding for drought tolerance is crucial for mitigating these \\neffects and ensuring food security amidst changing climate conditions.  \\nIn farming, nitrogen fixation by plants like legumes turns atmospheric nitrogen into usable forms, \\nenriching soil fertility and reducing the need for synthetic fertilizers. Nitrogen use efficiency (NUE) \\ngauges how well crops utilize nitrogen. This balance influences agricultural sustainability, impacting \\nenvironmental health. Nitrogen fixation can enhance NUE by supplementing nitrogen supply, lessening \\nreliance on synthetic fertilizers, and curbing environmental harm. Yet, the interaction between nitrogen \\nfixation and NUE hinges on various factors like crop type and management practices. Understanding \\nthis link guides sustainable nitrogen management, vital for balanced agricultural productivity and \\nenvironmental stewardship.  \\nThe connection between chlorophyll binding and leaf yield is crucial for a plant's ability to thrive \\nthrough photosynthesis. Chlorophyll, essential for this process, captures light energy for plant growth. \\nLeaf characteristics like size and chlorophyll con tent affect how efficiently plants can absorb and use \\nlight. Genetic differences in these traits impact yield and seed quality. For example, leaf shape and size \\ninfluence how well plants can distribute light and utilize it effectively. These traits are con trolled by\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 77}, page_content=\"78  complex genetic factors, as seen in studies identifying specific genetic regions linked to leaf traits and \\nchlorophyll content.  \\nNecrosis, the death of plant tissue, often caused by diseases like root rot, severely impacts plant health \\nand productivity. Root rot, typically triggered by pathogens like Phytophthora, attacks plant roots, \\nhindering water and nutrient absorption, and harming growth and yield. Moreover, it worsens soil \\nerosion and nutrient loss. On the other hand, carbon fixation, where plants convert carbon dioxide into \\nusable forms, aids growth and soil fertility. It indirectly boosts leaf yie ld by enhancing nutrient \\navailability and reducing fertilizer needs. However, the link between carbon fixation and yield is \\ncomplex, varying with plant type, environment, and farming methods. Understanding these connections \\nis vital for sustainable agricul ture, ensuring healthier plants and higher yields.  \\n \\n4.12 Phenotype/phenotype relation  \\nIn this phenotype -phenotype correlation matrix, we observe intriguing relationships among the factors \\nof Drought, Leaf Yield, NUE (Nutrient Use Efficiency), and Root Rot, shedding light on their \\ninterconnectedness and potential implications for plant physi ology and agriculture. The strong negative \\ncorrelation between Drought and NUE ( -0.95) underscores a significant relationship. As drought \\nconditions escalate, there is a notable decline in nutrient use efficiency. This suggests that genetic traits \\ninvolved  in drought response may influence how efficiently plants utilize nutrients, highlighting the \\nintricate interplay between environmental stressors and physiological processes. The robust negative \\ncorrelation between Leaf Yield and Root Rot ( -1) reveals a co mpelling association. As leaf yield \\nincreases, there's a corresponding decrease in root rot incidence, and vice versa. This implies that \\ngenetic factors influencing leaf productivity might also confer resistance or susceptibility to root \\npathogens, or conv ersely, root health might affect the plant's capacity to produce leaves. This intricate \\nrelationship underscores the importance of considering both above -ground and below -ground traits in \\ncrop improvement efforts. While weaker, the correlations between oth er factors hint at additional \\nphenotype -phenotype relationships. Although less pronounced, these associations may still offer \\nvaluable insights into how different traits interact and influence each other under varying environmental \\nconditions or genetic ba ckgrounds.\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 78}, page_content='79   \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 22: Correlation Matrix of Agricultural Factors  \\n \\n4.13 Genotyping Chip Generation  \\nThe table below contains data related to genotyping, particularly focusing on single nucleotide \\npolymorphisms (SNPs) generated using the Illumina platform. The table includes various columns such \\nas Illumina ID, Name, Quality, Illumina Strand, SNP, Address  A ID, Primers, #genomeBuildk, Ploidy, \\nSpecies, and Source. Quality Column Indicates the quality of SNPs, categorized into \"High\" and \"Very \\nHigh,\" suggesting the reliability and accuracy of the SNP data. The SNP Column Lists different types \\nof single nucle otide polymorphisms, which are the variations at a single position in a DNA sequence \\namong individuals. Ploidy Column All entries are marked as \"polyploid,\" indicating that the organisms \\nbeing studied have multiple sets of chromosomes, typical in many plan t species. Species Column \\nspecifies \" Morus indica 6K_ARRAY,\" indicating that the genotyping chip used is designed specifically \\nfor the species Morus indica  or a related organism, and it\\'s tailored to interrogate approximately 6000 \\nSNPs. The SNPs generated f rom this data are likely specific to Morus indica  or a closely related species, \\nand they have been obtained using a genotyping chip designed for this species. The \"Source\" column \\nmight provide additional information on the origin or sample type of the genetic material used for \\ngenotyping.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 79}, page_content='80   \\n \\nhttps://docs.google.com/spreadsheets/d/1DscSqLq64GhHlxUNnJu7mjt6i5EtUjA9/edit?usp=drive_link&\\nouid=104528999025534933339&rtpof=true&sd=true  \\n \\n \\n \\n \\n \\n \\n \\n4.14 Linkage Mapping  \\nDrought Resistance , this trait is highly concentrated in LG-2, LG -4, and LG -5. Genes associated with \\ndrought resistance are likely clustered within these linkage groups.  By identifying specific SNPs (single \\nnucleotide polymorphisms) within these regions, researchers can pinpoint candidate genes responsible \\nfor drought tolerance . The gene  IDs were homologicall y mapped with th e gene names.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 80}, page_content='81   \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\n \\nFigure 23: Marker -Trait association or linkage mapping  \\n \\n \\n \\n \\n4.15 Utilizing Genotype -Phenotype Correlation and Specific SNPs for Informed Breeding \\nTechniques in Agriculture  \\nFarmers can use genotype -phenotype correlations to prioritize breeding efforts aimed at improving \\ntraits with a strong positive correlation, such as root rot resistance and nitrogen use efficiency. By \\nselecting plants with favorable genotypes associated wi th these traits, breeders can develop cultivars \\nbetter suited to combat root rot and optimize nitrogen utilization, leading to improved crop health and \\nproductivity. Given the positive correlation between nitrogen use efficiency and root rot resistance,'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 81}, page_content='82  farmers can focus on breeding for cultivars that exhibit enhanced nitrogen use efficiency. This can \\ninvolve selecting plants with specific SNPs associated with efficient nitrogen uptake, utilization, and \\nassimilation, ultimately leading to improved crop pe rformance and reduced environmental impact. \\nFarmers can prioritize breeding goals based on the strength of genotype -phenotype correlations. For \\ninstance, while there may be a strong correlation between root rot resistance and nitrogen use efficiency, \\nthe low correlation between nitrogen use efficiency and leaf yield suggests that breeding efforts may \\nneed to be diversified to address multiple traits simultaneously, rather than focusing solely on \\nincreasing leaf yield. The genotype and phenotype mapping resu lts, which include ontologies related \\nto water deprivation, nitrogen fixation, necrosis, and carbon fixation, provide valuable insights into the \\nunderlying biological processes influencing trait variation. Farmers can use this information to inform \\nbreedin g strategies aimed at enhancing resilience to water stress, improving nitrogen fixation \\nefficiency, mitigating necrosis -related diseases, and optimizing carbon assimilation for improved yield \\nand quality [154,155].'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 82}, page_content='83  CHAPTER 5  \\nCONCLUSION AND FUTURE ASPECT  \\n \\n5.1 Conclusion  \\nIn the culmination of this study, a comprehensive exploration of the genomic landscape of Morus indica  \\nhas been undertaken, elucidating its multifaceted implications for agriculture and sericulture. Through \\nmeticulous data preprocessing, stringent quality assessment, comprehensive variant processing, and \\nrigorous genotype -phenotype correlation analysis, in valuable insights into the genetic underpinnings \\nof crucial traits essential for the successful cultivation of mulberry have been unearthed. This study ha s \\nprovided a comprehensive understanding of the genetic architecture governing traits such as drought \\nresistance, nitrogen use efficiency, and leaf yield, paving the way for the development of targeted \\nbreeding techniques aimed at enhancing the resilience and productivity of mulberry cultivars. By \\nestablishing robust databases, conducting meticulous linkage mapping efforts, and pinpointing specific \\nsingle nucleotide polymorphisms (SNPs), stakeholders have been equipped with indispensable \\nresources to inform  breeding strategies and expedite cultivar improvement programs.  \\nFurthermore, this study underscores the critical importance of addressing existing research gaps, \\nparticularly in developing Indian -specific genomic resources and molecular markers. By bridging these \\nlacunae, this study aims to empower farmers with actiona ble insights and tailored solutions for \\nsustainable mulberry cultivation practices, fostering agricultural resilience and bolstering food security. \\nIntegrating genotype -phenotype and phenotype -phenotype relations serves as a linchpin in this \\nendeavor, enab ling farmers to make informed decisions and select breeds that exhibit optimal traits \\nsuited to their specific agro ecological contexts. Through this strategic approach, this study endeavors \\nto optimize mulberry cultivation, maximize productivity, and elev ate the economic prosperity of \\nfarming communities. This study serves as a cornerstone for continued advancements in mulberry \\ngenomics and its transformative application in agricultural innovation. The development of primers for \\nchip development represents  a pivotal step forward, poised to amplify the utilization of genomic \\ninformation for precision breeding and cultivar enhancement. By leveraging the insights gleaned from \\nthis study and embracing emerging genomic paradigms, we are poised to catalyze a para digm shift in \\nmulberry cultivation, ensuring its enduring prominence and prosperity in the global agricultural \\nlandscape.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 83}, page_content='84   \\n5.2 Future Aspects  \\nThe future holds exciting prospects for refining and advancing conservation and breeding strategies, \\nall underpinned by the integration of comprehensive genetic insights. A key direction lies in the \\nexploration of quantitative trait loci (QTL) analysis, a method aimed at identifyin g specific genomic \\nregions associated with various agronomic traits crucial for mulberry cultivation. By deciphering the \\ngenetic basis of traits such as disease resistance, yield potential, and quality characteristics, breeders \\ncan target these regions for  precise trait enhancement, thereby accelerating the development of superior \\nmulberry cultivars. Moreover, the application of a metabolic approach presents a promising avenue for \\nunderstanding the intricate biochemical pathways underlying trait variation i n mulberry. By integrating \\nmetabolomic data with genomic information, researchers can unravel the complex interactions between \\ngenes and metabolites, providing deeper insights into the physiological processes governing mulberry \\ngrowth, development, and pro ductivity.  \\nIn parallel, the continued advancement of genetic monitoring programs holds immense potential for \\npreserving the genetic diversity of mulberry populations. By incorporating real -time genomic data \\nthrough high -throughput sequencing technologies and bioinfor matics tools, conservation efforts can be \\nbolstered, enabling timely interventions to mitigate threats and conserve the invaluable genetic \\nresources harbored within mulberry germplasm. The expansion of marker -assisted breeding \\napproaches, leveraging single  nucleotide polymorphisms (SNPs) and simple sequence repeats (SSRs), \\noffers an efficient means to accelerate cultivar improvement. Marker -assisted selection (MAS) and \\ngenomic selection techniques empower breeders to harness genetic information for the iden tification \\nand introgression of desirable traits into mulberry germplasm, thereby enhancing productivity, \\nresilience, and quality characteristics. In the realm of functional genomics, ongoing studies are poised \\nto elucidate the intricate biological mechani sms governing trait variation in mulberry. By integrating \\ngenomic data with transcriptomic, proteomic, and metabolomic datasets, researchers can unravel gene \\nexpression patterns, protein function, and metabolic pathways associated with critical agronomic t raits. \\nThis comprehensive understanding paves the way for targeted manipulation of the mulberry genome \\nto enhance desired traits, offering unprecedented opportunities for precision breeding. Moreover, the \\nexploration of genome editing technologies, such as  CRISPR -Cas9, heralds a new era of precise genetic'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 84}, page_content='85  manipulation in mulberry cultivation. By enabling targeted gene knockout, gene editing for trait \\nimprovement, and allele replacement, genome editing holds immense potential for enhancing mulberry \\nproductivity, resilience, and quality characteristics in a p recise and controlled manner. The integration \\nof these diverse approaches and technologies promises to revolutionize mulberry cultivation, ensuring \\nits sustainability and resilience in the face of evolving agricultural challenges. By embracing \\ninterdiscipl inary collaboration and leveraging cutting -edge genetic tools, the future of mulberry \\nagriculture is poised for remarkable advancements, driving innovation and prosperity in the global \\nsericulture industry.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 85}, page_content=\"86  CHAPTER 6  \\nREFERENCES  \\n1) Encyclopedia. (1998). JAMA, 279(17), 1409. https://doi.org/10.1001/jama.279.17.1409 -jbk0506 -6-\\n1 00/00544 Recent condition and trend of explosibility monitoring techniques for coal mine in China. \\n(2000). Fuel and Energy Abstracts, 41(1), 57. https://doi. org/10.1016/0140 -6701(00)95325 -6  \\n2) Yongkang Group. (2009). In Geological Formation Names of China (1866 –2000) (pp. 1345 -1345): \\nSpringer Berlin Heidelberg.  \\n3) West Java. (2016). In 2015 Annual Competitiveness Analysis and Development Strategies for \\nIndonesian Provinces (pp. 223 -234): WORLD SCIENTIFIC.  \\nhttp://ljournal.ru/wp -content/uploads/2017/03/a -2017 -023.pdf. (2017). http://dx.doi.org/10.18411/a -\\n2017 -023 \\n4) Identification of Sex -specific Molecular Markers in Barbel and Nile Carp using SCoT and ISSR \\nMarkers. (2021). Assiut Journal of Agricultural Sciences, 52(1), 60 -73. \\nhttps://doi.org/10.21608/ajas.2021.171568  \\n5) Morus indica. (2022). CABI Publishing. https://doi.org/10.1079/cabicompendium.69105335  \\n \\n6) Acharya, R., Bagchi, T., & Gangopadhyay, D. (2022). Mulberry as a Valuable Resource for \\nFood and Pharmaceutical Industries: A Review. In Medicinal Plants: IntechOpen.  \\n7) Akhunov , E., Nicolet, C., & Dvorak, J. (2009). Single nucleotide polymorphism genotyping in \\npolyploid wheat with the Illumina GoldenGate assay. TAG. Theoretical and applied genetics. \\nTheoretische und angewandte Genetik, 119(3), 507 -517. https://doi.org/10.1007/s0 0122 -009-\\n1059 -5  \\n8) Allen, A., Islamovic, E., Kaur, J., Gold, S., Shah, D., & Smith, T. J. (2011). Transgenic maize \\nplants expressing the Totivirus antifungal protein, KP4, are highly resistant to corn smut. Plant \\nBiotechnology Journal, 9(8), 857 -864. https://doi.org/10.1111/ j.1467 -7652.2011.00590.x  \\n9) Allen, A. M., Winfield, M. O., Burridge, A. J., Downie, R. C., Benbow, H. R., Barker, G. L. \\nA., Wilkinson, P. A., Coghill, J., Waterfall, C., Davassi, A., Scopes, G., Pirani, A., Webster, T., \\nBrew, F., Bloor, C., Griffiths, S., Bentley, A. R., Alda, M., Ja ck, P., . . . Edwards, K. J. (2017). \\nCharacterization of a Wheat Breeders' Array suitable for high -throughput SNP genotyping of \\nglobal accessions of hexaploid bread wheat (Triticum aestivum). Plant biotechnology journal, \\n15(3), 390 -401. https://doi.org/10. 1111/pbi.12635  \\n10) Arcade, A., Labourdette, A., Falque, M., Mangin, B., Chardon, F., Charcosset, A., & Joets, J. \\n(2004). BioMercator: integrating genetic maps and QTL towards discovery of candidate genes. \\nBioinformatics, 20(14), 2324 -2326. https://doi.org/10.1093/bioinformat ics/bth230\"),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 86}, page_content='87  11) Asama, J., Suzuki, H., & Oiwa, T. (2016). Development of a magnetically suspended double -\\nlayered cylinder with 1 -DOF active positioning control. Mechanical Engineering Journal, 3(1), \\n15-00544 -00515 -00544. https://doi.org/10.1299/mej.15 -00544  \\n12) Aversano, R., Ercolano, M., Caruso, I., Fasano, C., Rosellini, D., & Carputo, D. (2014). \\nMolecular Tools for Exploring Polyploid Genomes in Plants. In The Role of Bioinformatics in \\nAgriculture (pp. 73 -97): Apple Academic Press.  \\n13) Awasthi, A. K., Nagaraja, G. M., Naik, G. V., Kanginakudru, S., Thangavelu, K., & Nagaraju, \\nJ. (2004). Genetic diversity and relationships in mulberry (genus Morus) as revealed by RAPD \\nand ISSR marker assays. BMC genetics, 5, 1 -1. https://doi.org/10.1186/1 471-2156 -5-1  \\n14) Baciu, E. -D., Baci, G. -M., Moise, A. R., & Dezmirean, D. S. (2023). A Status Review on the \\nImportance of Mulberry (Morus spp.) and Prospects towards Its Cultivation in a Controlled \\nEnvironment. Horticulturae, 9(4), 444. https://doi.org/10.3390/horticultura e9040444  \\n15) Banerjee, R., Chattopadhyay, S., Saha, A. K., & Kumar, S. N. (2014). Association of \\nmorphological and biochemical features of mulberry lines with resistance to bacterial leaf spot. \\nArchives Of Phytopathology And Plant Protection, 47(16), 2020 -2032. \\nhttps:/ /doi.org/10.1080/03235408.2013.868693  \\n16) Boro, H., Das, S., & Middha, S. K. (2021). The therapeutic potential and the health benefits of \\nMorus indica Linn.: a mini review. Advances in Traditional Medicine, 21(2), 241 -252. \\nhttps://doi.org/10.1007/s13596 -020-00544 -5  \\n17) Boro, H., Das, S., & Middha, S. K. (2021). The therapeutic potential and the health benefits of \\nMorus indica Linn.: a mini review. Advances in Traditional Medicine, 21(2), 241 -252. \\nhttps://doi.org/10.1007/s13596 -020-00544 -5  \\n18) Bukhari, R., & Sharma, A. (2022). IMPACT OF COVID -19 ON THE INDIAN \\nSERICULTURE INDUSTRY. UTTAR PRADESH JOURNAL OF ZOOLOGY, 9 -19. \\nhttps://doi.org/10.56557/upjoz/2022/v43i203191  \\n19) Burkart -Waco, D., Henry, I. M., Ngo, K., Comai, L., & Tai, T. H. (2016). Determining Mutation \\nDensity Using Restriction Enzyme Sequence Comparative Analysis (RESCAN). In \\nBiotechnologies for Plant Mutation Breeding (pp. 305 -321): Springer International Publ ishing.  \\n20) Bus, A., Hecht, J., Huettel, B., Reinhardt, R., & Stich, B. (2012). High -throughput \\npolymorphism detection and genotyping in Brassica napus using next -generation RAD \\nsequencing. BMC genomics, 13, 281 -281. https://doi.org/10.1186/1471 -2164 -13-281  \\n21) Cai, M., Mu, L., Wang, Z. -L., Liu, J. -Y., Liu, T. -L., Wanapat, M., & Huang, B. -Z. (2019). \\nAssessment of mulberry leaf as a potential feed supplement for animal feeding in P.R. China. \\nAsian -Australasian journal of animal sciences, 32(8), 1145 -1152.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 87}, page_content='88  https://doi.org/10.5713/ajas.18.0671  \\n22) Cavanagh, C. R., Chao, S., Wang, S., Huang, B. E., Stephen, S., Kiani, S., Forrest, K., \\nSaintenac, C., Brown -Guedira, G. L., Akhunova, A., See, D., Bai, G., Pumphrey, M., Tomar, \\nL., Wong, D., Kong, S., Reynolds, M., da Silva, M. L., Bockelman, H., . . . Ak hunov, E. (2013). \\nGenome -wide comparative diversity uncovers multiple targets of selection for improvement in \\nhexaploid wheat landraces and cultivars. Proceedings of the National Academy of Sciences of \\nthe United States of America, 110(20), 8057 -8062. http s://doi.org/10.1073/pnas.1217133110  \\n23) Checker, V. G., Saeed, B., & Khurana, P. (2012). Analysis of expressed sequence tags from \\nmulberry (Morus indica) roots and implications for comparative transcriptomics and marker \\nidentification. Tree Genetics &amp; Genomes, 8(6), 1437 -1450. \\nhttps://doi.or g/10.1007/s11295 -012-0531 -6  \\n24) Checker, V. G., Saeed, B., & Khurana, P. (2012). Analysis of expressed sequence tags from \\nmulberry (Morus indica) roots and implications for comparative transcriptomics and marker \\nidentification. Tree Genetics &amp; Genomes, 8(6), 1437 -1450. \\nhttps://doi.or g/10.1007/s11295 -012-0531 -6  \\n25) Cheema, J., Ellis, T. H. N., & Dicks, J. (2010). THREaD Mapper Studio: a novel, visual web \\nserver for the estimation of genetic linkage maps. Nucleic acids research, 38(Web Server issue), \\nW188 -W193. https://doi.org/10.1093/nar/gkq430  \\n26) Chen, C., You, L. -J., Abbasi, A. M., Fu, X., & Liu, R. H. (2015). Optimization for ultrasound \\nextraction of polysaccharides from mulberry fruits with antioxidant and hyperglycemic activity \\nin vitro. Carbohydrate Polymers, 130, 122 -132. https://doi.org/10.1 016/j.carbpol.2015.05.003  \\n27) Chen, C., You, L. -J., Abbasi, A. M., Fu, X., Liu, R. H., & Li, C. (2016). Characterization of \\npolysaccharide fractions in mulberry fruit and assessment of their antioxidant and \\nhypoglycemic activities in vitro. Food &amp; Function, 7(1), 530 -539. \\nhttps://d oi.org/10.1039/c5fo01114k  \\n28) Chen, C., Zhang, B., Fu, X., & Liu, R. H. (2016). A novel polysaccharide isolated from \\nmulberry fruits (Murus alba L.) and its selenide derivative: structural characterization and \\nbiological activities. Food &amp; Function, 7(6), 2886 -2897. \\nhttps://doi.org /10.1039/c6fo00370b  \\n29) Chen, S., Huang, Z., Dai, Y., Qin, S., Gao, Y., Zhang, L., Gao, Y., & Chen, J. (2013). The \\ndevelopment of 7E chromosome -specific molecular markers for Thinopyrum elongatum based \\non SLAF -seq technology. PloS one, 8(6), e65122 -e65122. \\nhttps://doi.org/10.1371 /journal.pone.0065122  \\n30) Chen, Z. J. (2010). Molecular mechanisms of polyploidy and hybrid vigor. Trends in plant'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 88}, page_content='89  science, 15(2), 57 -71. https://doi.org/10.1016/j.tplants.2009.12.003  \\n31) Choulet, F., Wicker, T., Rustenholz, C., Paux, E., Salse, J., Leroy, P., Schlub, S., Le Paslier, \\nM.-C., Magdelenat, G., Gonthier, C., Couloux, A., Budak, H., Breen, J., Pumphrey, M., Liu, S., \\nKong, X., Jia, J., Gut, M., Brunel, D., . . . Feuillet, C. (2010 ). Megabase level sequencing reveals \\ncontrasted organization and evolution patterns of the wheat gene and transposable element \\nspaces. The Plant cell, 22(6), 1686 -1701. https://doi.org/10.1105/tpc.110.074187  \\n32) Cifuentes, M., Eber, F., Lucas, M. -O., Lode, M., Chèvre, A. -M., & Jenczewski, E. (2010). \\nRepeated polyploidy drove different levels of crossover suppression between homoeologous \\nchromosomes in Brassica napus allohaploids. The Plant cell, 22(7), 2265 -2276. \\nhttps://doi.org/10.1105/tpc.109.072991  \\n33) Claros, M. G., Bautista, R., Guerrero -Fernández, D., Benzerki, H., Seoane, P., & Fernández -\\nPozo, N. (2012). Why assembling plant genome sequences is so challenging. Biology, 1(2), \\n439-459. https://doi.org/10.3390/biology1020439  \\n34) Comai, L. (2005). The advantages and disadvantages of being polyploid. Nature Reviews \\nGenetics, 6(11), 836 -846. https://doi.org/10.1038/nrg1711  \\n35) Dhanyalakshmi, K. H., & Nataraja, K. N. (2018). Mulberry (Morus spp.) has the features to \\ntreat as a potential perennial model system. Plant signaling & behavior, 13(8), e1491267 -\\ne1491267. https://doi.org/10.1080/15592324.2018.1491267  \\n36) Dhanyalakshmi, K. H., & Nataraja, K. N. (2018). Mulberry (Morus spp.) has the features to \\ntreat as a potential perennial model system. Plant signaling & behavior, 13(8), e1491267 -\\ne1491267. https://doi.org/10.1080/15592324.2018.1491267  \\n37) Duhan, N., Kaur, S., & Kaundal, R. (2023). ranchSATdb: A Genome -Wide Simple Sequence \\nRepeat (SSR) Markers Database of Livestock Species for Mutant Germplasm Characterization \\nand Improving Farm Animal Health. Genes, 14(7), 1481. \\nhttps://doi.org/10.3390/gene s14071481  \\n38) Elshire, R. J., Glaubitz, J. C., Sun, Q., Poland, J. A., Kawamoto, K., Buckler, E. S., & Mitchell, \\nS. E. (2011). A robust, simple genotyping -by-sequencing (GBS) approach for high diversity \\nspecies. PloS one, 6(5), e19379 -e19379. https://doi.org/10.1371/jou rnal.pone.0019379  \\n39) Endelman, J. B., & Plomion, C. (2014). LPmerge: an R package for merging genetic maps by \\nlinear programming. Bioinformatics, 30(11), 1623 -1624. \\nhttps://doi.org/10.1093/bioinformatics/btu091  \\n40) Faiz, A. U. H., & Faiz, L. Z. (2021). Phytochemical Screening, Antimicrobial Activity of \\nGemmotherapeutically White Mulberry (Morus Alba) Leaves. Journal of Bioresource \\nManagement, 8(2), 14 -19. https://doi.org/10.35691/jbm.1202.0176'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 89}, page_content='90  41) Farvid, M. S., Ding, M., Pan, A., Sun, Q., Chiuve, S. E., Steffen, L. M., Willett, W. C., & Hu, \\nF. B. (2014). Dietary linoleic acid and risk of coronary heart disease: a systematic review and \\nmeta -analysis of prospective cohort studies. Circulation, 130(18 ), 1568 -1578. \\nhttps://doi.org/10.1161/CIRCULATIONAHA.114.010236  \\n42) Gaeta, R. T., Pires, J. C., Iniguez -Luy, F., Leon, E., & Osborn, T. C. (2007). Genomic changes \\nin resynthesized Brassica napus and their effect on gene expression and phenotype. The Plant \\ncell, 19(11), 3403 -3417. https://doi.org/10.1105/tpc.107.054346  \\n43) Geng, X., Jiang, C., Yang, J., Wang, L., Wu, X., & Wei, W. (2016). Rapid Identification of \\nCandidate Genes for Seed Weight Using the SLAF -Seq Method in Brassica napus. PloS one, \\n11(1), e0147580 -e0147580. https://doi.org/10.1371/journal.pone.0147580  \\n44) Ghorpade, K., Netam, V., Desai, N., & Shewale, V. (2022). ASCERTAINING ROLE AND \\nIMPORTANCE OF ARTIFICIAL DIET AND BOTANICALS MEDIATED SERIFEED FOR \\nTHE STRENGTHENING OF SERICULTURE INDUSTRY. UTTAR PRADESH JOURNAL \\nOF ZOOLOGY, 46 -54. https://doi.org/10.56557/ upjoz/2022/v43i223230  \\n45) Giampieri, F., Tulipani, S., Alvarez -Suarez, J. M., Quiles, J. L., Mezzetti, B., & Battino, M. \\n(2012). The strawberry: Composition, nutritional quality, and impact on human health. \\nNutrition, 28(1), 9 -19. https://doi.org/10.1016/j.nut.2011.08.009  \\n46) Gnanesh, B. N., Arunakumar, G. S., Tejaswi, A., Supriya, M., Manojkumar, H. B., & Devi, S. \\nS. (2022). Characterization and Pathogenicity of Lasiodiplodia theobromae Causing Black Root \\nRot and Identification of Novel Sources of Resistance in Mulberry Collec tions. The plant \\npathology journal, 38(4), 272 -286. https://doi.org/10.5423/PPJ.OA.01.2022.0005  \\n47) Gnanesh, B. N., Mondal, R., G S, A., H B, M., Singh, P., M R, B., P, S., Burji, S. M., T, M., & \\nV, S. (2023). Genome size, genetic diversity, and phenotypic variability imply the effect of \\ngenetic variation instead of ploidy on trait plasticity in the cros s-pollinated tree species of \\nmulberry. PloS one, 18(8), e0289766 -e0289766. https://doi.org/10.1371/journal.pone.0289766  \\n48) Grześkowiak, J., Łochyńska, M., & Frankowski, J. (2022). Sericulture in Terms of Sustainable \\nDevelopment in Agriculture. Problemy Ekorozwoju, 17(2), 210 -217. \\nhttps://doi.org/10.35784/pe.2022.2.23  \\n49) Hamilton, J. P., & Robin Buell, C. (2012). Advances in plant genome sequencing. The Plant \\nJournal, 70(1), 177 -190. https://doi.org/10.1111/j.1365 -313x.2012.04894.x  \\n50) He, N., Zhang, C., Qi, X., Zhao, S., Tao, Y., Yang, G., Lee, T. -H., Wang, X., Cai, Q., Li, D., \\nLu, M., Liao, S., Luo, G., He, R., Tan, X., Xu, Y., Li, T., Zhao, A., Jia, L., . . . Xiang, Z. (2013). \\nDraft genome sequence of the mulberry tree Morus notabilis . Nature communications, 4, 2445 -\\n2445. https://doi.org/10.1038/ncomms3445'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 90}, page_content='91  51) Huang, H. -P., Ou, T. -T., & Wang, C. -J. (2013). Mulberry ( sang shèn zǐ) and its bioactive \\ncompounds, the chemoprevention effects and molecular mechanisms in vitro and in vivo. \\nJournal of traditional and complementary medicine, 3(1), 7 -15. https://doi.org/1 0.4103/2225 -\\n4110.106535  \\n52) Illahi, I., Mittal, V., & Sharmila, K. K. (2022). Integrated Pest and Disease Management \\n(IPDM) Approach in Mulberry Sericulture of Jammu &amp; Kashmir. Research Biotica, 4(3), \\n161-165. https://doi.org/10.54083/resbio/4.3.2022/161 -165  \\n53) Islam, M. S., Ahmed, F., & Rahman, M. M. (2023). Intercropping pulses with mulberry on \\nsericulture productivity and profitability. Bangladesh Journal of Agriculture, 47(2), 107 -118. \\nhttps://doi.org/10.3329/bjagri.v47i2.63330  \\n54) Ito, T., & Arai, N. (1966). Nutrition of the silkworm, Bombyx mori —XI. Requirements for \\naspartic and glutamic acids. Journal of Insect Physiology, 12(7), 861 -869. \\nhttps://doi.org/10.1016/0022 -1910(66)90036 -9  \\n55) Jain, M., Bansal, J., Rajkumar, M. S., Sharma, N., Khurana, J. P., & Khurana, P. (2022). Draft \\ngenome sequence of Indian mulberry (Morus indica) provides a resource for functional and \\ntranslational genomics. Genomics, 114(3), 110346. \\nhttps://doi.org/10.101 6/j.ygeno.2022.110346  \\n56) Jain, M., Bansal, J., Rajkumar, M. S., Sharma, N., Khurana, J. P., & Khurana, P. (2022). Draft \\ngenome sequence of Indian mulberry (Morus indica) provides a resource for functional and \\ntranslational genomics. Genomics, 114(3), 110346. \\nhttps://doi.org/10.101 6/j.ygeno.2022.110346  \\n57) Jain, M., Bansal, J., Rajkumar, M. S., Sharma, N., Khurana, J. P., & Khurana, P. (2022). Draft \\ngenome sequence of Indian mulberry (Morus indica) provides a resource for functional and \\ntranslational genomics. Genomics, 114(3), 110346. \\nhttps://doi.org/10.101 6/j.ygeno.2022.110346  \\n58) Jan, B., Parveen, R., Zahiruddin, S., Khan, M. U., Mohapatra, S., & Ahmad, S. (2021). \\nNutritional constituents of mulberry and their potential applications in food and \\npharmaceuticals: A review. Saudi journal of biological sciences, 28(7), 3909 -3921. \\nhttps ://doi.org/10.1016/j.sjbs.2021.03.056  \\n59) Jan, B., Parveen, R., Zahiruddin, S., Khan, M. U., Mohapatra, S., & Ahmad, S. (2021). \\nNutritional constituents of mulberry and their potential applications in food and \\npharmaceuticals: A review. Saudi journal of biological sciences, 28(7), 3909 -3921. \\nhttps ://doi.org/10.1016/j.sjbs.2021.03.056  \\n60) Jansen, J., de Jong, A. G., & van Ooijen, J. W. (2001). Constructing dense genetic linkage maps.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 91}, page_content='92  Theoretical and Applied Genetics, 102(6 -7), 1113 -1122. \\nhttps://doi.org/10.1007/s001220000489  \\n61) Jiao, W. -B., & Schneeberger, K. (2017). The impact of third generation genomic technologies \\non plant genome assembly. Current Opinion in Plant Biology, 36, 64 -70. \\nhttps://doi.org/10.1016/j.pbi.2017.02.002  \\n62) Jones, S. Identification of spectral disease signatures and resistant QTL for charcoal rot \\ninfection in soybean Iowa State University]. http://dx.doi.org/10.31274/etd -180810 -5913  \\n63) Ju, W. -T., Kwon, O. C., Kim, H. -B., Sung, G. -B., Kim, H. -W., & Kim, Y. -S. (2018). Qualitative \\nand quantitative analysis of flavonoids from 12 species of Korean mulberry leaves. Journal of \\nfood science and technology, 55(5), 1789 -1796. https://doi.org/10.10 07/s13197 -018-3093 -2  \\n64) Kaur, S., Francki, M. G., & Forster, J. W. (2011). Identification, characterization and \\ninterpretation of single‐nucleotide sequence variation in allopolyploid crop species. Plant \\nBiotechnology Journal, 10(2), 125 -138. https://doi.org/10.1111/j.1467 -7652.2 011.00644.x  \\n65) Kerstin, M. (2006). Food and Agriculture Organization of the United Nations (FAO). In Max \\nPlanck Encyclopedia of Public International Law. Oxford University Press. \\nhttps://doi.org/10.1093/law:epil/9780199231690/e462  \\n66) Kim, D., & Kang, K. -H. (2022). Anti -Inflammatory and Anti -Bacterial Potential of Mulberry \\nLeaf Extract on Oral Microorganisms. International journal of environmental research and \\npublic health, 19(9), 4984. https://doi.org/10.3390/ijerph19094984  \\n67) Kim, D., & Kang, K. -H. (2022). Anti -Inflammatory and Anti -Bacterial Potential of Mulberry \\nLeaf Extract on Oral Microorganisms. International journal of environmental research and \\npublic health, 19(9), 4984. https://doi.org/10.3390/ijerph19094984  \\n68) Kim, D., & Kang, K. -H. (2022). Anti -Inflammatory and Anti -Bacterial Potential of Mulberry \\nLeaf Extract on Oral Microorganisms. International journal of environmental research and \\npublic health, 19(9), 4984. https://doi.org/10.3390/ijerph19094984  \\n69) Kusano, G., Orihara, S., Tsukamoto, D., Shibano, M., Coskun, M., Guvenc, A., & Erdurak, C. \\nS. (2002). Five New Nortropane Alkaloids and Six New Amino Acids from the Fruit of Morus \\nalba LINNE Growing in Turkey. Chemical and Pharmaceutical Bulletin, 50(2), 1 85-192. \\nhttps://doi.org/10.1248/cpb.50.185  \\n70) Lai, J., Li, R., Xu, X., Jin, W., Xu, M., Zhao, H., Xiang, Z., Song, W., Ying, K., Zhang, M., \\nJiao, Y., Ni, P., Zhang, J., Li, D., Guo, X., Ye, K., Jian, M., Wang, B., Zheng, H., . . . Wang, J. \\n(2010). Genome -wide patterns of genetic variation among elite maize inbred lines. Nature \\ngenetics, 42(11), 1027 -1030. https://doi.org/10.1038/ng.684  \\n71) Le, H. D., Fallon, E. M., Kalish, B. T., de Meijer, V. E., Meisel, J. A., Gura, K. M., Nose, V.,'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 92}, page_content='93  Pan, A. H., Bistrian, B. R., & Puder, M. (2013). The effect of varying ratios of docosahexaenoic \\nacid and arachidonic acid in the prevention and reversal of biochemical essential fatty acid \\ndeficiency in a murine model. Metabolism: clinical and experimenta l, 62(4), 499 -508. \\nhttps://doi.org/10.1016/j.metabol.2012.10.003  \\n72) Lee, J. S., Synytsya, A., Kim, H. B., Choi, D. J., Lee, S., Lee, J., Kim, W. J., Jang, S., & Park, \\nY. I. (2013). Purification, characterization and immunomodulating activity of a pectic \\npolysaccharide isolated from Korean mulberry fruit Oddi (Morus alba L. ). International \\nImmunopharmacology, 17(3), 858 -866. https://doi.org/10.1016/j.intimp.2013.09.019  \\n73) Leitch, I. J., & Bennett, M. D. (1997). Polyploidy in angiosperms. Trends in Plant Science, \\n2(12), 470 -476. https://doi.org/10.1016/s1360 -1385(97)01154 -0  \\n74) Li, F., Fan, G., Lu, C., Xiao, G., Zou, C., Kohel, R. J., Ma, Z., Shang, H., Ma, X., Wu, J., Liang, \\nX., Huang, G., Percy, R. G., Liu, K., Yang, W., Chen, W., Du, X., Shi, C., Yuan, Y., . . . Yu, S. \\n(2015). Genome sequence of cultivated Upland cotton (Gossy pium  hirsutum TM -1) provides \\ninsights into genome evolution. Nature Biotechnology, 33(5), 524 -530. \\nhttps://doi.org/10.1038/nbt.3208  \\n75) Li, T., Qi, X., Zeng, Q., Xiang, Z., & He, N. (2014). MorusDB: a resource for mulberry \\ngenomics and genome biology. Database : the journal of biological databases and curation, \\n2014, bau054. https://doi.org/10.1093/database/bau054  \\n76) Li, W., Huang, L., & Gill, B. S. (2008). Recurrent deletions of puroindoline genes at the grain \\nhardness locus in four independent lineages of polyploid wheat. Plant physiology, 146(1), 200 -\\n212. https://doi.org/10.1104/pp.107.108852  \\n77) Liu, B., Vega, J. M., & Feldman, M. (1998). Rapid genomic changes in newly synthesized \\namphiploids of<i>Triticum</i>and<i>Aegilops</i>. II. Changes in low -copy coding DNA \\nsequences. Genome, 41(4), 535 -542. https://doi.org/10.1139/g98 -052  \\n78) Liu, C. -J., & Lin, J. -Y. (2014). Protective effects of strawberry and mulberry fruit \\npolysaccharides on inflammation and apoptosis in murine primary splenocytes. Journal of Food \\nand Drug Analysis, 22(2), 210 -219. https://doi.org/10.1016/j.jfda.2014.01.015  \\n79) Liu, Y., Salsman, E., Wang, R., Galagedara, N., Zhang, Q., Fiedler, J. D., Liu, Z., Xu, S., Faris, \\nJ. D., & Li, X. (2020). Meta -QTL analysis of tan spot resistance in wheat. Theoretical and \\nApplied Genetics, 133(8), 2363 -2375. https://doi.org/10.1007/s0012 2-020-03604 -1  \\n80) Liu, Y., Salsman, E., Wang, R., Galagedara, N., Zhang, Q., Fiedler, J. D., Liu, Z., Xu, S., Faris, \\nJ. D., & Li, X. (2020). Meta -QTL analysis of tan spot resistance in wheat. Theoretical and \\nApplied Genetics, 133(8), 2363 -2375. https://doi.org/10.1007/s0012 2-020-03604 -1  \\n81) Luk, K. F. Process development for separation and purification of schisandrin B stereoisomer'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 93}, page_content='94  from fructus schisandrae The Hong Kong University of Science and Technology Library]. \\nhttp://dx.doi.org/10.14711/thesis -b987389  \\n82) Maccaferri, M., Ricci, A., Salvi, S., Milner, S. G., Noli, E., Martelli, P. L., Casadio, R., \\nAkhunov, E., Scalabrin, S., Vendramin, V., Ammar, K., Blanco, A., Desiderio, F., Distelfeld, \\nA., Dubcovsky, J., Fahima, T., Faris, J., Korol, A., Massi, A., . . . Tuberosa, R. (2014). A high‐\\ndensity, <scp>SNP</scp>‐based consensus map of tetraploid wheat as a bridge to integrate \\ndurum and bread wheat genomics and breeding. Plant Biotechnology Journal, 13(5), 648 -663. \\nhttps://doi.org/10.1111/pbi.12288  \\n83) Madlung, A., Masuelli, R. W., Watson, B., Reynolds, S. H., Davison, J., & Comai, L. (2002). \\nRemodeling of DNA methylation and phenotypic and transcriptional changes in synthetic \\nArabidopsis allotetraploids. Plant physiology, 129(2), 733 -746. \\nhttps://doi.or g/10.1104/pp.003095  \\n84) Madlung, A., Tyagi, A. P., Watson, B., Jiang, H., Kagochi, T., Doerge, R. W., Martienssen, R., \\n& Comai, L. (2004). Genomic changes in synthetic Arabidopsis polyploids. The Plant Journal, \\n41(2), 221 -230. https://doi.org/10.1111/j.1365 -313x.2004.02297.x  \\n85) Manjunath, K. K., Bs, H., Nm, A., & Ha, B. (2023). Breeding for mulberry improvement: A \\nreview. The Pharma Innovation, 12(2), 477 -481. \\nhttps://doi.org/10.22271/tpi.2023.v12.i2f.18461  \\n86) Mardis, E. R. (2008). Next -Generation DNA Sequencing Methods. Annual Review of \\nGenomics and Human Genetics, 9(1), 387 -402. \\nhttps://doi.org/10.1146/annurev.genom.9.081307.164359  \\n87) Mason, A. S. (2014). SSR Genotyping. In Methods in Molecular Biology (pp. 77 -89): Springer \\nNew York.  \\n88) Mathew, F., Varghese, B., Joseph, D., Mathews, M. M., Sunny, B., & George, J. (2019). An \\nAppraisal of Pharmacological actions of <i>Morus indica</i>: The Indian mulberry with a \\ndetailed investigation on its Anti -diabetic potential. Research Journal of Phar macy and \\nTechnology, 12(8), 3654. https://doi.org/10.5958/0974 -360x.2019.00623.1  \\n89) Moffat, J. W. (2006). Scalar –tensor –vector gravity theory. Journal of Cosmology and \\nAstroparticle Physics, 2006(03), 004 -004. https://doi.org/10.1088/1475 -7516/2006/03/004  \\n90) Mushtaq, R., Qadiri, B., Lone, F. A., Raja, T. A., Singh, H., Pervez, A., & Sharma, R. (2023). \\nRole of Sericulture in Achieving Sustainable Development Goals. Problemy Ekorozwoju, \\n18(1), 199 -206. https://doi.org/10.35784/pe.2023.1.21  \\n91) Negro, C., Aprile, A., De Bellis, L., & Miceli, A. (2019). Nutraceutical Properties of Mulberries \\nGrown in Southern Italy (Apulia). Antioxidants (Basel, Switzerland), 8(7), 223.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 94}, page_content='95  https://doi.org/10.3390/antiox8070223  \\n92) Newell, W. R., Mott, R., Beck, S., & Lehrach, H. (1995). Construction of Genetic Maps Using \\nDistance Geometry. Genomics, 30(1), 59 -70. https://doi.org/10.1006/geno.1995.0009  \\n93) Nicolas, S. D., Leflon, M., Monod, H., Eber, F., Coriton, O., Huteau, V., Chèvre, A. -M., & \\nJenczewski, E. (2009). Genetic regulation of meiotic cross -overs between related genomes in \\nBrassica napus haploids and hybrids. The Plant cell, 21(2), 373 -385. \\nhttps://doi.org/10.1105/tpc.108.062273  \\n94) Ou, T., Zhang, M., Gao, H., Wang, F., Xu, W., Liu, X., Wang, L., Wang, R., & Xie, J. (2023). \\nStudy on the Potential for Stimulating Mulberry Growth and Drought Tolerance of Plant \\nGrowth -Promoting Fungi. International journal of molecular sciences, 24(4), 4 090. \\nhttps://doi.org/10.3390/ijms24044090  \\n95) Pada Bag, M. (2021). Use of Sericulture Byproduct as Feed for Tilapia – Oreochromis niloticus \\n(Linn.). Bioscience Biotechnology Research Communications, 14(2), 622 -627. \\nhttps://doi.org/10.21786/bbrc/14.2.28  \\n96) Park, K. M., You, J. S., Lee, H. Y., Baek, N. I., & Hwang, J. K. (2003). Kuwanon G: an \\nantibacterial agent from the root bark of Morus alba against oral pathogens. Journal of \\nEthnopharmacology, 84(2 -3), 181 -185. https://doi.org/10.1016/s0378 -8741(02)00318 -5  \\n97) Park, K. M., You, J. S., Lee, H. Y., Baek, N. I., & Hwang, J. K. (2003). Kuwanon G: an \\nantibacterial agent from the root bark of Morus alba against oral pathogens. Journal of \\nEthnopharmacology, 84(2 -3), 181 -185. https://doi.org/10.1016/s0378 -8741(02)00318 -5  \\n98) Pérez -Gregorio, M. R., Regueiro, J., Alonso -González, E., Pastrana -Castro, L. M., & Simal -\\nGándara, J. (2011). Influence of alcoholic fermentation process on antioxidant activity and \\nphenolic levels from mulberries (Morus nigra L.). LWT - Food Science and T echnology, 44(8), \\n1793 -1801. https://doi.org/10.1016/j.lwt.2011.03.007  \\n99) Pinto, M. V., H S, P., M S, R., R, T., & Naik, V. G. (2018). Association mapping of quantitative \\nresistance to charcoal root rot in mulberry germplasm. PloS one, 13(7), e0200099 -e0200099. \\nhttps://doi.org/10.1371/journal.pone.0200099  \\n100) Pinto, M. V., H S, P., M S, R., R, T., & Naik, V. G. (2018). Association mapping of \\nquantitative resistance to charcoal root rot in mulberry germplasm. PloS one, 13(7), e0200099 -\\ne0200099. https://doi.org/10.1371/journal.pone.0200099  \\n101) Pires, J. C., Lim, K. Y., Kovarík, A., Matyásek, R., Boyd, A., Leitch, A. R., Leitch, I. \\nJ., Bennett, M. D., Soltis, P. S., & Soltis, D. E. (2004). Molecular cytogenetic analysis of \\nrecently evolved<i>Tragopogon</i>(Asteraceae) allopolyploids reveal a kary otype that is \\nadditive of the diploid progenitors. American Journal of Botany, 91(7), 1022 -1035.'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 95}, page_content='96  https://doi.org/10.3732/ajb.91.7.1022  \\n102) Pootakham, W., Jomchai, N., Ruang -areerate, P., Shearman, J. R., Sonthirod, C., \\nSangsrakru, D., Tragoonrung, S., & Tangphatsornruang, S. (2015). Genome -wide SNP \\ndiscovery and identification of QTL associated with agronomic traits in oil palm using \\ngenotypi ng-by-sequencing (GBS). Genomics, 105(5 -6), 288 -295. \\nhttps://doi.org/10.1016/j.ygeno.2015.02.002  \\n103) Porceddu, A., Albertini, E., Barcaccia, G., Falistocco, E., & Falcinelli, M. (2002). \\nLinkage mapping in apomictic and sexual Kentucky bluegrass (Poa pratensis L.) genotypes \\nusing a two way pseudo -testcross strategy based on AFLP and SAMPL markers. Theoreti cal \\nand Applied Genetics, 104(2), 273 -280. https://doi.org/10.1007/s001220100659  \\n104) Radojkovic, M. Mulberry extracts (Morus spp., Morasceae), composition, activity and \\napplication National Library of Serbia]. http://dx.doi.org/10.2298/ns20121109radojkovic  \\n105) Ramsey, J., & Schemske, D. W. (2002). Neopolyploidy in Flowering Plants. Annual \\nReview of Ecology and Systematics, 33(1), 589 -639. \\nhttps://doi.org/10.1146/annurev.ecolsys.33.010802.150437  \\n106) Ray, P., Gogoi, S. N., Bandyopadhyay, S., Jena, R. K., Ramachandran, S., Sharma, G. \\nK., Sah, K. D., Singh, S. K., & Ray, S. K. (2021). Soil -cum-nutrient management package for \\nachieving high yield and quality of mulberry (<i>Morus indica</i> L.) leaf in ac id soil of North \\nEastern India. Journal of Plant Nutrition, 44(17), 2501 -2513. \\nhttps://doi.org/10.1080/01904167.2021.1921196  \\n107) Ray, P., Gogoi, S. N., Bandyopadhyay, S., Jena, R. K., Ramachandran, S., Sharma, G. \\nK., Sah, K. D., Singh, S. K., & Ray, S. K. (2021). Soil -cum-nutrient management package for \\nachieving high yield and quality of mulberry (<i>Morus indica</i> L.) leaf in ac id soil of North \\nEastern India. Journal of Plant Nutrition, 44(17), 2501 -2513. \\nhttps://doi.org/10.1080/01904167.2021.1921196  \\n108) Ray, P., Gogoi, S. N., Bandyopadhyay, S., Jena, R. K., Ramachandran, S., Sharma, G. \\nK., Sah, K. D., Singh, S. K., & Ray, S. K. (2021). Soil -cum-nutrient management package for \\nachieving high yield and quality of mulberry (<i>Morus indica</i> L.) leaf in ac id soil of North \\nEastern India. Journal of Plant Nutrition, 44(17), 2501 -2513. \\nhttps://doi.org/10.1080/01904167.2021.1921196  \\n109) Reyna -López, G. E., Simpson, J., & Ruiz -Herrera, J. (1997). Differences in DNA \\nmethylation patterns are detectable during the dimorphic transition of fungi by amplification of \\nrestriction polymorphisms. Molecular and General Genetics MGG, 253(6), 703 -710. \\nhttps://doi.org/10.1007/s004380050374'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 96}, page_content='97  110) Ronin, Y., Mester, D., Minkov, D., Belotserkovski, R., Jackson, B. N., Schnable, P. S., \\nAluru, S., & Korol, A. (2012). Two -phase analysis in consensus genetic mapping. G3 \\n(Bethesda, Md.), 2(5), 537 -549. https://doi.org/10.1534/g3.112.002428  \\n111) Rukmangada, M. S., Sumathy, R., Sivaprasad, V., & Naik, V. G. (2018). Genome -wide \\nidentification and characterization of growth -regulating factors in mulberry (Morus spp.). \\nTrees, 32(6), 1695 -1705. https://doi.org/10.1007/s00468 -018-1744 -6  \\n112) Saini, P., Rohela, G. K., Kumar, J. S., Shabnam, A. A., & Kumar, A. (2023). Cultivation, \\nUtilization, and Economic Benefits of Mulberry. In Compendium of Plant Genomes (pp. 13 -\\n56): Springer International Publishing.  \\n113) Saini, P., Rohela, G. K., Kumar, J. S., Shabnam, A. A., & Kumar, A. (2023). Cultivation, \\nUtilization, and Economic Benefits of Mulberry. In Compendium of Plant Genomes (pp. 13 -\\n56): Springer International Publishing.  \\n114) Sajeevan, R. S., Nataraja, K. N., Shivashankara, K. S., Pallavi, N., Gurumurthy, D. S., \\n& Shivanna, M. B. (2017). Expression of Arabidopsis SHN1 in Indian Mulberry (Morus indica \\nL.) Increases Leaf Surface Wax Content and Reduces Post -harvest Water Loss. Fr ontiers in \\nplant science, 8, 418 -418. https://doi.org/10.3389/fpls.2017.00418  \\n115) Sajeevan, R. S., Nataraja, K. N., Shivashankara, K. S., Pallavi, N., Gurumurthy, D. S., \\n& Shivanna, M. B. (2017). Expression of Arabidopsis SHN1 in Indian Mulberry (Morus indica \\nL.) Increases Leaf Surface Wax Content and Reduces Post -harvest Water Loss. Fr ontiers in \\nplant science, 8, 418 -418. https://doi.org/10.3389/fpls.2017.00418  \\n116) Salmon, A., Ainouche, M. L., & Wendel, J. F. (2005). Genetic and epigenetic \\nconsequences of recent hybridization and polyploidy in <i>Spartina</i> (Poaceae). Molecular \\nEcology, 14(4), 1163 -1175. https://doi.org/10.1111/j.1365 -294x.2005.02488.x  \\n117) San, B., & Yildirim, A. N. (2010). Phenolic, alpha -tocopherol, beta -carotene and fatty \\nacid composition of four promising jujube (Ziziphus jujuba Miller) selections. Journal of Food \\nComposition and Analysis, 23(7), 706 -710. https://doi.org/10.1016/j.jfca.2 010.02.008  \\n118) Sarkar, T., Ravindra, K. N., Doss, S. G., Kumar, P. M. P., & Tewary, P. (2021). In vitro \\nregeneration of mulberry plants from seedling explants of Morus indica cv. G4 through direct \\norganogenesis. Trees, 36(1), 113 -125. https://doi.org/10.1007/s00468 -021-02186 -9  \\n119) Sastry, G. R. K. (1984). Molecular biology of plant tumours. Physiological Plant \\nPathology, 24(1), 129 -130. https://doi.org/10.1016/0048 -4059(84)90080 -8  \\n120) Schadt, E. E., Turner, S., & Kasarskis, A. (2010). A window into third -generation \\nsequencing. Human Molecular Genetics, 19(R2), R227 -R240. \\nhttps://doi.org/10.1093/hmg/ddq416'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 97}, page_content='98  121) Shreelakshmi, S. V., Nazareth, M. S., Kumar, S. S., Giridhar, P., Prashanth, K. V. H., \\n& Shetty, N. P. (2021). Physicochemical Composition and Characterization of Bioactive \\nCompounds of Mulberry (Morus indica L.) Fruit During Ontogeny. Plant Foods for Huma n \\nNutrition, 76(3), 304 -310. https://doi.org/10.1007/s11130 -021-00909 -4  \\n122) Shulaev , V., Sargent, D. J., Crowhurst, R. N., Mockler, T. C., Folkerts, O., Delcher, A. \\nL., Jaiswal, P., Mockaitis, K., Liston, A., Mane, S. P., Burns, P., Davis, T. M., Slovin, J. P., \\nBassil, N., Hellens, R. P., Evans, C., Harkins, T., Kodira, C., Desany, B., .  . . Folta, K. M. \\n(2011). The genome of woodland strawberry (Fragaria vesca). Nature genetics, 43(2), 109 -116. \\nhttps://doi.org/10.1038/ng.740  \\n123) Somers, D. J., Isaac, P., & Edwards, K. (2004). A high -density microsatellite consensus \\nmap for bread wheat (Triticum aestivum L.). Theoretical and Applied Genetics, 109(6), 1105 -\\n1114. https://doi.org/10.1007/s00122 -004-1740 -7  \\n124) Song, K., Lu, P., Tang, K., & Osborn, T. C. (1995). Rapid genome change in synthetic \\npolyploids of Brassica and its implications for polyploid evolution. Proceedings of the National \\nAcademy of Sciences of the United States of America, 92(17), 7719 -7723. \\nhttps://doi.org/10.1073/pnas.92.17.7719  \\n125) Stam, P. (1993). Construction of integrated genetic linkage maps by means of a new \\ncomputer package: JoinMap. The Plant Journal, 3(5), 739 -744. https://doi.org/10.1046/j.1365 -\\n313x.1993.03050739.x  \\n126) Sun, X., Liu, D., Zhang, X., Li, W., Liu, H., Hong, W., Jiang, C., Guan, N., Ma, C., \\nZeng, H., Xu, C., Song, J., Huang, L., Wang, C., Shi, J., Wang, R., Zheng, X., Lu, C., Wang, \\nX., & Zheng, H. (2013). SLAF -seq: an efficient method of large -scale de novo S NP discovery \\nand genotyping using high -throughput sequencing. PloS one, 8(3), e58700 -e58700. \\nhttps://doi.org/10.1371/journal.pone.0058700  \\n127) Szadkowski, E., Eber, F., Huteau, V., Lodé, M., Huneau, C., Belcram, H., Coriton, O., \\nManzanares‐Dauleux, M. J., Delourme, R., King, G. J., Chalhoub, B., Jenczewski, E., & \\nChèvre, A. M. (2010). The first meiosis of resynthesized <i>Brassica napus</i>, a ge nome \\nblender. New Phytologist, 186(1), 102 -112. https://doi.org/10.1111/j.1469 -8137.2010.03182.x  \\n128) Tate, J. A., Soltis, D. E., & Soltis, P. S. (2005). Polyploidy in Plants. In The Evolution \\nof the Genome (pp. 371 -426): Elsevier.  \\n129) Teh, C. -K., Ong, A. -L., Kwong, Q. -B., Apparow, S., Chew, F. -T., Mayes, S., Mohamed, \\nM., Appleton, D., & Kulaveerasingam, H. (2016). Genome -wide association study identifies \\nthree key loci for high mesocarp oil content in perennial crop oil palm. Scientific  reports, 6, \\n19075 -19075. https://doi.org/10.1038/srep19075'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 98}, page_content='99  130) Thabti, I., Elfalleh, W., Tlili, N., Ziadi, M., Campos, M. G., & Ferchichi, A. (2013). \\nPhenols, Flavonoids, and Antioxidant and Antibacterial Activity of Leaves and Stem Bark \\nof<i>Morus</i>Species. International Journal of Food Properties, 17(4), 842 -854. \\nhttps://doi.org/10.1080/10942912.2012.660722  \\n131) Thaipitakwong, T., Numhom, S., & Aramwit, P. (2018). Mulberry leaves and their \\npotential effects against cardiometabolic risks: a review of chemical compositions, biological \\nproperties and clinical efficacy. Pharmaceutical biology, 56(1), 109 -118. \\nhttps:// doi.org/10.1080/13880209.2018.1424210  \\n132) Tikader, A., & Thangavelu, K. (2006). Grafting performance in some Mulberry \\ngermplasm collected through exploration. Indian Journal of Forestry, 29(1), 69 -71. \\nhttps://doi.org/10.54207/bsmps1000 -2006 -29xr38  \\n133) Town, C. D., Cheung, F., Maiti, R., Crabtree, J., Haas, B. J., Wortman, J. R., Hine, E. \\nE., Althoff, R., Arbogast, T. S., Tallon, L. J., Vigouroux, M., Trick, M., & Bancroft, I. (2006). \\nComparative genomics of Brassica oleracea and Arabidopsis thaliana rev eal gene loss, \\nfragmentation, and dispersal after polyploidy. The Plant cell, 18(6), 1348 -1359. \\nhttps://doi.org/10.1105/tpc.106.041665  \\n134) Van, O., W., J., & Jansen, J. (2013). Preface. In Genetic Mapping in Experimental \\nPopulations (pp. xi -xii): Cambridge University Press.  \\n135) Van Tassell, C. P., Smith, T. P. L., Matukumalli, L. K., Taylor, J. F., Schnabel, R. D., \\nLawley, C. T., Haudenschild, C. D., Moore, S. S., Warren, W. C., & Sonstegard, T. S. (2008). \\nSNP discovery and allele frequency estimation by deep sequencing of reduce d representation \\nlibraries. Nature Methods, 5(3), 247 -252. https://doi.org/10.1038/nmeth.1185  \\n136) Vijayan, K., Nair, C. V., & Chatterjee, S. N. (2005). Molecular characterization of \\nmulberry genetic resources indigenous to India. Genetic Resources and Crop Evolution, 52(1), \\n77-86. https://doi.org/10.1007/s10722 -005-0288 -y  \\n137) Vijayan, K., Ravikumar, G., & Tikader, A. (2018). Mulberry (Morus spp.) Breeding for \\nHigher Fruit Production. In Advances in Plant Breeding Strategies: Fruits (pp. 89 -130): \\nSpringer International Publishing.  \\n138) Vijayan, K., Srivastava, P. P., Raju, P. J., & Saratchandra, B. (2012). Breeding for higher \\nproductivity in mulberry. Czech Journal of Genetics and Plant Breeding, 48(4), 147 -156. \\nhttps://doi.org/10.17221/162/2011 -cjgpb  \\n139) Wang, S., Wong, D., Forrest, K., Allen, A., Chao, S., Huang, B. E., Maccaferri, M., \\nSalvi, S., Milner, S. G., Cattivelli, L., Mastrangelo, A. M., Whan, A., Stephen, S., Barker, G., \\nWieseke, R., Plieske, J., International Wheat Genome Sequencing, C., Lillem o, M., Mather, D.,'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 99}, page_content='100  . . . Akhunov, E. (2014). Characterization of polyploid wheat genomic diversity using a high -\\ndensity 90,000 single nucleotide polymorphism array. Plant biotechnology journal, 12(6), 787 -\\n796. https://doi.org/10.1111/pbi.12183  \\n140) Wang, Y., Jin, W., Che, Y., Huang, D., Wang, J., Zhao, M., & Sun, G. (2019). \\nAtmospheric Nitrogen Dioxide Improves Photosynthesis in Mulberry Leaves via Effective \\nUtilization of Excess Absorbed Light Energy. Forests, 10(4), 312. \\nhttps://doi.org/10.3390/f10 040312  \\n141) Wang, Y., Jin, W., Che, Y., Huang, D., Wang, J., Zhao, M., & Sun, G. (2019). \\nAtmospheric Nitrogen Dioxide Improves Photosynthesis in Mulberry Leaves via Effective \\nUtilization of Excess Absorbed Light Energy. Forests, 10(4), 312. \\nhttps://doi.org/10.3390/f10 040312  \\n142) Wang, Z., Hu, H., Jiang, X., Tao, Y., Lin, Y., Wu, F., Hou, S., Liu, S., Li, C., Chen, G., \\n& Liu, Y. (2020). Identification and Validation of a Novel Major Quantitative Trait Locus for \\nPlant Height in Common Wheat (Triticum aestivum L.). Frontiers in genet ics, 11, 602495 -\\n602495. https://doi.org/10.3389/fgene.2020.602495  \\n143) Wen, P., Hu, T. -G., Linhardt, R. J., Liao, S. -T., Wu, H., & Zou, Y. -X. (2019). Mulberry: \\nA review of bioactive compounds and advanced processing technology. Trends in Food Science \\n&amp; Technology, 83, 138 -158. https://doi.org/10.1016/j.tifs.2018.11.017  \\n144) Whelan, J. (2008). The health implications of changing linoleic acid intakes. \\nProstaglandins, Leukotrienes and Essential Fatty Acids, 79(3 -5), 165 -167. \\nhttps://doi.org/10.1016/j.plefa.2008.09.013  \\n145) Wingen , L. U., West, C., Leverington -Waite, M., Collier, S., Orford, S., Goram, R., \\nYang, C. -Y., King, J., Allen, A. M., Burridge, A., Edwards, K. J., & Griffiths, S. (2017). Wheat \\nLandrace Genome Diversity. Genetics, 205(4), 1657 -1676. \\nhttps://doi.org/10.1534/g enetics.116.194688  \\n146) Won -Jin, C., Ji -Ho, C., Won -Taek, C., Seung -Hee, K., Jong -Seok, K., Shi -Jong, L., & \\nTae-Kyung, Y. Hydrogen effect on 670 nm AlGaInP visible laser during high temperature \\noperation Proceedings of IEEE 14th International Semiconductor Laser Conference,  \\nhttp://dx.doi.org/10.1109/islc.1994.519152  \\n147) Xie, J. -H., Jin, M. -L., Morris, G. A., Zha, X. -Q., Chen, H. -Q., Yi, Y., Li, J. -E., Wang, \\nZ.-J., Gao, J., Nie, S. -P., Shang, P., & Xie, M. -Y. (2015). Advances on Bioactive \\nPolysaccharides from Medicinal Plants. Critical Reviews in Food Science and Nutrition , \\n56(sup1), S60 -S84. https://doi.org/10.1080/10408398.2015.1069255  \\n148) Yang, X., Yang, L., & Zheng, H. (2010). Hypolipidemic and antioxidant effects of'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 100}, page_content='101  mulberry (Morus alba L.) fruit in hyperlipidaemia rats. Food and Chemical Toxicology, 48(8 -\\n9), 2374 -2379. https://doi.org/10.1016/j.fct.2010.05.074  \\n149) Yonghui, W., Close, T. J., & Lonardi, S. (2011). Accurate Construction of Consensus \\nGenetic Maps via Integer Linear Programming. IEEE/ACM Transactions on Computational \\nBiology and Bioinformatics, 8(2), 381 -394. https://doi.org/10.1109/tcbb.2010.35  \\n150) Yuan, Q., & Zhao, L. (2017). The Mulberry (<i>Morus alba</i> L.) Fruit —A Review \\nof Characteristic Components and Health Benefits. Journal of Agricultural and Food Chemistry, \\n65(48), 10383 -10394. https://doi.org/10.1021/acs.jafc.7b03614  \\n151) Zhang, Y., Wang, L., Xin, H., Li, D., Ma, C., Ding, X., Hong, W., & Zhang, X. (2013). \\nConstruction of a high -density genetic map for sesame based on large scale marker development \\nby specific length amplified fragment (SLAF) sequencing. BMC Plant Biology, 13, 141 -141. \\nhttps://doi.org/10.1186/1471 -2229 -13-141  \\n152) Zhang, Z., Shang, H., Shi, Y., Huang, L., Li, J., Ge, Q., Gong, J., Liu, A., Chen, T., \\nWang, D., Wang, Y., Palanga, K. K., Muhammad, J., Li, W., Lu, Q., Deng, X., Tan, Y., Song, \\nW., Cai, J., . . . Yuan, Y. (2016). Construction of a high -density genetic map  by specific locus \\namplified fragment sequencing (SLAF -seq) and its application to Quantitative Trait Loci (QTL) \\nanalysis for boll weight in upland cotton (Gossypium hirsutum.). BMC Plant Biology, 16, 79 -\\n79. https://doi.org/10.1186/s12870 -016-0741 -4  \\n153) Zhang, Z., Shang, H., Shi, Y., Huang, L., Li, J., Ge, Q., Gong, J., Liu, A., Chen, T., \\nWang, D., Wang, Y., Palanga, K. K., Muhammad, J., Li, W., Lu, Q., Deng, X., Tan, Y., Song, \\nW., Cai, J., . . . Yuan, Y. (2016). Construction of a high -density genetic map  by specific locus \\namplified fragment sequencing (SLAF -seq) and its application to Quantitative Trait Loci (QTL) \\nanalysis for boll weight in upland cotton (Gossypium hirsutum.). BMC Plant Biology, 16, 79 -\\n79. https://doi.org/10.1186/s12870 -016-0741 -4  \\n154) Zhao, L., Wu, M., Xiao, C., Yang, L., Zhou, L., Gao, N., Li, Z., Chen, J., Chen, J., Liu, \\nJ., Qin, H., & Zhao, J. (2015). Discovery of an intrinsic tenase complex inhibitor: Pure \\nnonasaccharide from fucosylated glycosaminoglycan. Proceedings of the Nationa l Academy of \\nSciences of the United States of America, 112(27), 8284 -8289. \\nhttps://doi.org/10.1073/pnas.1504229112  \\n155) Zhao, M., Shu, G., Hu, Y., Cao, G., & Wang, Y. (2023). Pattern and variation in simple \\nsequence repeat (SSR) at different genomic regions and its implications to maize evolution and \\nbreeding. BMC genomics, 24(1), 136 -136. https://doi.org/10.1186/s12864 -023-09156 -0'),\n",
       " Document(metadata={'source': 'SINSHA_THESIS_final[1].pdf', 'page': 101}, page_content='102')]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from langchain_community.document_loaders import PyPDFLoader\n",
    "loader = PyPDFLoader(\"SINSHA_THESIS_final[1].pdf\")\n",
    "pages=loader.load_and_split()\n",
    "pages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Answer the question based on the following text. \n",
      "If you can't answer the question, please write \"I don't know.\"\n",
      "\n",
      "context=What is the main idea of this text?\n",
      "\n",
      "\n",
      "Question: What is the main idea?\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from langchain.prompts import PromptTemplate\n",
    "template = \"\"\"\n",
    "Answer the question based on the following text. \n",
    "If you can't answer the question, please write \"I don't know.\"\n",
    "\n",
    "context={context}\n",
    "\n",
    "\n",
    "Question: {question}\n",
    "\"\"\"\n",
    "prompt = PromptTemplate.from_template(template)\n",
    "print (prompt.format(context= \"What is the main idea of this text?\", question= \"What is the main idea?\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "chain= prompt | model | parser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'title': 'PromptInput',\n",
       " 'type': 'object',\n",
       " 'properties': {'context': {'title': 'Context', 'type': 'string'},\n",
       "  'question': {'title': 'Question', 'type': 'string'}},\n",
       " 'required': ['context', 'question']}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chain.input_schema.schema()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'The main idea of this text is to encourage and empower someone to be confident and unafraid to show their worth. It\\'s urging them to \"own the night\" and showcase their talents or personality like a firework that explodes in a burst of color and energy.'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chain.invoke(\n",
    "    {\n",
    "        \"context\": \"You just got to ignite the light and let it shine, just own the night like the Fourth of July. 'Cause, baby, you're a firework. Come on, show 'em what you're worth. Make 'em go, 'Aah, aah, aah' as you shoot across the sky-y-y. Baby, you're a firework. Come on, let your colors burst. Make 'em go, 'Aah, aah, aah'. You're gonna leave 'em all in awe, awe, awe.\", \n",
    "        \"question\": \"What is the main idea of this text?\"\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install --force-reinstall torch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting torch==2.3.0\n",
      "  Downloading torch-2.3.0-cp311-cp311-win_amd64.whl.metadata (26 kB)\n",
      "Collecting torchvision==0.18.0\n",
      "  Downloading torchvision-0.18.0-cp311-cp311-win_amd64.whl.metadata (6.6 kB)\n",
      "Collecting torchaudio==2.3.0\n",
      "  Downloading torchaudio-2.3.0-cp311-cp311-win_amd64.whl.metadata (6.4 kB)\n",
      "Requirement already satisfied: filelock in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.3.0) (3.15.4)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.3.0) (4.12.2)\n",
      "Requirement already satisfied: sympy in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.3.0) (1.13.2)\n",
      "Requirement already satisfied: networkx in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.3.0) (3.3)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.3.0) (3.1.4)\n",
      "Requirement already satisfied: fsspec in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.3.0) (2024.6.1)\n",
      "Collecting mkl<=2021.4.0,>=2021.1.1 (from torch==2.3.0)\n",
      "  Downloading mkl-2021.4.0-py2.py3-none-win_amd64.whl.metadata (1.4 kB)\n",
      "Requirement already satisfied: numpy in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torchvision==0.18.0) (1.26.4)\n",
      "Collecting pillow!=8.3.*,>=5.3.0 (from torchvision==0.18.0)\n",
      "  Downloading pillow-10.4.0-cp311-cp311-win_amd64.whl.metadata (9.3 kB)\n",
      "Collecting intel-openmp==2021.* (from mkl<=2021.4.0,>=2021.1.1->torch==2.3.0)\n",
      "  Downloading intel_openmp-2021.4.0-py2.py3-none-win_amd64.whl.metadata (1.2 kB)\n",
      "Collecting tbb==2021.* (from mkl<=2021.4.0,>=2021.1.1->torch==2.3.0)\n",
      "  Downloading tbb-2021.13.1-py3-none-win_amd64.whl.metadata (1.1 kB)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from jinja2->torch==2.3.0) (2.1.5)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from sympy->torch==2.3.0) (1.3.0)\n",
      "Downloading torch-2.3.0-cp311-cp311-win_amd64.whl (159.8 MB)\n",
      "   ---------------------------------------- 0.0/159.8 MB ? eta -:--:--\n",
      "   ---------------------------------------- 1.0/159.8 MB 5.6 MB/s eta 0:00:29\n",
      "   ---------------------------------------- 1.0/159.8 MB 5.6 MB/s eta 0:00:29\n",
      "    --------------------------------------- 2.6/159.8 MB 4.0 MB/s eta 0:00:40\n",
      "    --------------------------------------- 2.9/159.8 MB 3.4 MB/s eta 0:00:46\n",
      "    --------------------------------------- 3.9/159.8 MB 3.6 MB/s eta 0:00:44\n",
      "   - -------------------------------------- 4.7/159.8 MB 3.7 MB/s eta 0:00:43\n",
      "   - -------------------------------------- 5.5/159.8 MB 3.7 MB/s eta 0:00:42\n",
      "   - -------------------------------------- 6.3/159.8 MB 3.7 MB/s eta 0:00:42\n",
      "   - -------------------------------------- 7.1/159.8 MB 3.8 MB/s eta 0:00:41\n",
      "   - -------------------------------------- 7.9/159.8 MB 3.8 MB/s eta 0:00:41\n",
      "   -- ------------------------------------- 8.7/159.8 MB 3.8 MB/s eta 0:00:40\n",
      "   -- ------------------------------------- 9.7/159.8 MB 3.8 MB/s eta 0:00:40\n",
      "   -- ------------------------------------- 10.5/159.8 MB 3.8 MB/s eta 0:00:40\n",
      "   -- ------------------------------------- 11.3/159.8 MB 3.8 MB/s eta 0:00:40\n",
      "   --- ------------------------------------ 12.1/159.8 MB 3.8 MB/s eta 0:00:39\n",
      "   --- ------------------------------------ 12.8/159.8 MB 3.8 MB/s eta 0:00:39\n",
      "   --- ------------------------------------ 13.6/159.8 MB 3.8 MB/s eta 0:00:39\n",
      "   --- ------------------------------------ 14.4/159.8 MB 3.8 MB/s eta 0:00:39\n",
      "   --- ------------------------------------ 15.2/159.8 MB 3.8 MB/s eta 0:00:38\n",
      "   ---- ----------------------------------- 16.0/159.8 MB 3.8 MB/s eta 0:00:38\n",
      "   ---- ----------------------------------- 16.8/159.8 MB 3.8 MB/s eta 0:00:38\n",
      "   ---- ----------------------------------- 17.6/159.8 MB 3.8 MB/s eta 0:00:38\n",
      "   ---- ----------------------------------- 18.4/159.8 MB 3.8 MB/s eta 0:00:38\n",
      "   ---- ----------------------------------- 19.4/159.8 MB 3.8 MB/s eta 0:00:37\n",
      "   ----- ---------------------------------- 20.2/159.8 MB 3.8 MB/s eta 0:00:37\n",
      "   ----- ---------------------------------- 21.0/159.8 MB 3.8 MB/s eta 0:00:37\n",
      "   ----- ---------------------------------- 21.8/159.8 MB 3.8 MB/s eta 0:00:37\n",
      "   ----- ---------------------------------- 22.8/159.8 MB 3.8 MB/s eta 0:00:36\n",
      "   ----- ---------------------------------- 23.6/159.8 MB 3.8 MB/s eta 0:00:36\n",
      "   ------ --------------------------------- 24.4/159.8 MB 3.8 MB/s eta 0:00:36\n",
      "   ------ --------------------------------- 25.2/159.8 MB 3.8 MB/s eta 0:00:36\n",
      "   ------ --------------------------------- 26.0/159.8 MB 3.8 MB/s eta 0:00:35\n",
      "   ------ --------------------------------- 26.7/159.8 MB 3.8 MB/s eta 0:00:35\n",
      "   ------ --------------------------------- 27.8/159.8 MB 3.8 MB/s eta 0:00:35\n",
      "   ------- -------------------------------- 28.6/159.8 MB 3.8 MB/s eta 0:00:35\n",
      "   ------- -------------------------------- 29.4/159.8 MB 3.8 MB/s eta 0:00:34\n",
      "   ------- -------------------------------- 30.1/159.8 MB 3.8 MB/s eta 0:00:34\n",
      "   ------- -------------------------------- 30.9/159.8 MB 3.8 MB/s eta 0:00:34\n",
      "   ------- -------------------------------- 31.7/159.8 MB 3.8 MB/s eta 0:00:34\n",
      "   -------- ------------------------------- 32.5/159.8 MB 3.8 MB/s eta 0:00:34\n",
      "   -------- ------------------------------- 33.3/159.8 MB 3.9 MB/s eta 0:00:33\n",
      "   -------- ------------------------------- 34.1/159.8 MB 3.9 MB/s eta 0:00:33\n",
      "   -------- ------------------------------- 34.9/159.8 MB 3.9 MB/s eta 0:00:33\n",
      "   -------- ------------------------------- 35.7/159.8 MB 3.9 MB/s eta 0:00:33\n",
      "   --------- ------------------------------ 36.4/159.8 MB 3.9 MB/s eta 0:00:33\n",
      "   --------- ------------------------------ 37.5/159.8 MB 3.9 MB/s eta 0:00:32\n",
      "   --------- ------------------------------ 38.3/159.8 MB 3.9 MB/s eta 0:00:32\n",
      "   --------- ------------------------------ 39.1/159.8 MB 3.9 MB/s eta 0:00:32\n",
      "   --------- ------------------------------ 39.8/159.8 MB 3.9 MB/s eta 0:00:32\n",
      "   ---------- ----------------------------- 40.6/159.8 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------- ----------------------------- 41.4/159.8 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------- ----------------------------- 42.5/159.8 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------- ----------------------------- 43.0/159.8 MB 3.9 MB/s eta 0:00:31\n",
      "   ----------- ---------------------------- 44.0/159.8 MB 3.9 MB/s eta 0:00:31\n",
      "   ----------- ---------------------------- 44.8/159.8 MB 3.9 MB/s eta 0:00:30\n",
      "   ----------- ---------------------------- 45.6/159.8 MB 3.9 MB/s eta 0:00:30\n",
      "   ----------- ---------------------------- 46.4/159.8 MB 3.9 MB/s eta 0:00:30\n",
      "   ----------- ---------------------------- 47.2/159.8 MB 3.9 MB/s eta 0:00:30\n",
      "   ------------ --------------------------- 48.0/159.8 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------ --------------------------- 48.8/159.8 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------ --------------------------- 49.5/159.8 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------ --------------------------- 50.6/159.8 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------ --------------------------- 51.4/159.8 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------- -------------------------- 52.2/159.8 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------- -------------------------- 53.0/159.8 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------- -------------------------- 53.7/159.8 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------- -------------------------- 54.8/159.8 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------- -------------------------- 55.6/159.8 MB 3.9 MB/s eta 0:00:27\n",
      "   -------------- ------------------------- 56.1/159.8 MB 3.9 MB/s eta 0:00:27\n",
      "   -------------- ------------------------- 57.1/159.8 MB 3.9 MB/s eta 0:00:27\n",
      "   -------------- ------------------------- 57.9/159.8 MB 3.9 MB/s eta 0:00:27\n",
      "   -------------- ------------------------- 58.7/159.8 MB 3.9 MB/s eta 0:00:27\n",
      "   -------------- ------------------------- 59.5/159.8 MB 3.9 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 60.6/159.8 MB 3.9 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 61.3/159.8 MB 3.9 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 62.1/159.8 MB 3.9 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 62.4/159.8 MB 3.8 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 62.9/159.8 MB 3.8 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 63.2/159.8 MB 3.8 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 63.4/159.8 MB 3.8 MB/s eta 0:00:26\n",
      "   --------------- ------------------------ 63.7/159.8 MB 3.7 MB/s eta 0:00:26\n",
      "   ---------------- ----------------------- 64.0/159.8 MB 3.7 MB/s eta 0:00:26\n",
      "   ---------------- ----------------------- 64.2/159.8 MB 3.7 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 64.2/159.8 MB 3.7 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 64.5/159.8 MB 3.6 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 64.7/159.8 MB 3.6 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 64.7/159.8 MB 3.6 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 65.0/159.8 MB 3.5 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 65.3/159.8 MB 3.5 MB/s eta 0:00:28\n",
      "   ---------------- ----------------------- 65.5/159.8 MB 3.5 MB/s eta 0:00:28\n",
      "   ---------------- ----------------------- 66.6/159.8 MB 3.5 MB/s eta 0:00:27\n",
      "   ---------------- ----------------------- 67.6/159.8 MB 3.5 MB/s eta 0:00:27\n",
      "   ----------------- ---------------------- 68.4/159.8 MB 3.5 MB/s eta 0:00:27\n",
      "   ----------------- ---------------------- 69.2/159.8 MB 3.5 MB/s eta 0:00:26\n",
      "   ----------------- ---------------------- 70.0/159.8 MB 3.5 MB/s eta 0:00:26\n",
      "   ----------------- ---------------------- 70.8/159.8 MB 3.5 MB/s eta 0:00:26\n",
      "   ----------------- ---------------------- 71.6/159.8 MB 3.5 MB/s eta 0:00:26\n",
      "   ------------------ --------------------- 72.4/159.8 MB 3.5 MB/s eta 0:00:25\n",
      "   ------------------ --------------------- 73.1/159.8 MB 3.5 MB/s eta 0:00:25\n",
      "   ------------------ --------------------- 74.2/159.8 MB 3.5 MB/s eta 0:00:25\n",
      "   ------------------ --------------------- 75.0/159.8 MB 3.5 MB/s eta 0:00:25\n",
      "   ------------------ --------------------- 75.8/159.8 MB 3.5 MB/s eta 0:00:24\n",
      "   ------------------- -------------------- 76.5/159.8 MB 3.5 MB/s eta 0:00:24\n",
      "   ------------------- -------------------- 77.3/159.8 MB 3.5 MB/s eta 0:00:24\n",
      "   ------------------- -------------------- 78.1/159.8 MB 3.5 MB/s eta 0:00:24\n",
      "   ------------------- -------------------- 78.9/159.8 MB 3.5 MB/s eta 0:00:23\n",
      "   -------------------- ------------------- 80.0/159.8 MB 3.5 MB/s eta 0:00:23\n",
      "   -------------------- ------------------- 80.7/159.8 MB 3.5 MB/s eta 0:00:23\n",
      "   -------------------- ------------------- 81.5/159.8 MB 3.5 MB/s eta 0:00:23\n",
      "   -------------------- ------------------- 82.3/159.8 MB 3.5 MB/s eta 0:00:22\n",
      "   -------------------- ------------------- 83.1/159.8 MB 3.5 MB/s eta 0:00:22\n",
      "   -------------------- ------------------- 83.9/159.8 MB 3.5 MB/s eta 0:00:22\n",
      "   --------------------- ------------------ 84.7/159.8 MB 3.5 MB/s eta 0:00:22\n",
      "   --------------------- ------------------ 85.7/159.8 MB 3.5 MB/s eta 0:00:21\n",
      "   --------------------- ------------------ 86.5/159.8 MB 3.6 MB/s eta 0:00:21\n",
      "   --------------------- ------------------ 87.3/159.8 MB 3.6 MB/s eta 0:00:21\n",
      "   ---------------------- ----------------- 88.1/159.8 MB 3.6 MB/s eta 0:00:21\n",
      "   ---------------------- ----------------- 88.9/159.8 MB 3.6 MB/s eta 0:00:20\n",
      "   ---------------------- ----------------- 89.9/159.8 MB 3.6 MB/s eta 0:00:20\n",
      "   ---------------------- ----------------- 90.7/159.8 MB 3.6 MB/s eta 0:00:20\n",
      "   ---------------------- ----------------- 91.5/159.8 MB 3.6 MB/s eta 0:00:20\n",
      "   ----------------------- ---------------- 92.3/159.8 MB 3.6 MB/s eta 0:00:19\n",
      "   ----------------------- ---------------- 93.1/159.8 MB 3.6 MB/s eta 0:00:19\n",
      "   ----------------------- ---------------- 93.8/159.8 MB 3.6 MB/s eta 0:00:19\n",
      "   ----------------------- ---------------- 94.6/159.8 MB 3.6 MB/s eta 0:00:19\n",
      "   ----------------------- ---------------- 95.4/159.8 MB 3.6 MB/s eta 0:00:18\n",
      "   ------------------------ --------------- 96.2/159.8 MB 3.6 MB/s eta 0:00:18\n",
      "   ------------------------ --------------- 97.3/159.8 MB 3.6 MB/s eta 0:00:18\n",
      "   ------------------------ --------------- 98.0/159.8 MB 3.6 MB/s eta 0:00:18\n",
      "   ------------------------ --------------- 98.8/159.8 MB 3.6 MB/s eta 0:00:17\n",
      "   ------------------------ --------------- 99.6/159.8 MB 3.6 MB/s eta 0:00:17\n",
      "   ------------------------- -------------- 100.4/159.8 MB 3.6 MB/s eta 0:00:17\n",
      "   ------------------------- -------------- 101.2/159.8 MB 3.6 MB/s eta 0:00:17\n",
      "   ------------------------- -------------- 102.0/159.8 MB 3.6 MB/s eta 0:00:17\n",
      "   ------------------------- -------------- 103.0/159.8 MB 3.6 MB/s eta 0:00:16\n",
      "   ------------------------- -------------- 103.8/159.8 MB 3.6 MB/s eta 0:00:16\n",
      "   -------------------------- ------------- 104.6/159.8 MB 3.6 MB/s eta 0:00:16\n",
      "   -------------------------- ------------- 105.4/159.8 MB 3.6 MB/s eta 0:00:16\n",
      "   -------------------------- ------------- 105.9/159.8 MB 3.6 MB/s eta 0:00:15\n",
      "   -------------------------- ------------- 107.0/159.8 MB 3.6 MB/s eta 0:00:15\n",
      "   -------------------------- ------------- 107.7/159.8 MB 3.6 MB/s eta 0:00:15\n",
      "   --------------------------- ------------ 108.8/159.8 MB 3.6 MB/s eta 0:00:15\n",
      "   --------------------------- ------------ 109.6/159.8 MB 3.6 MB/s eta 0:00:14\n",
      "   --------------------------- ------------ 110.4/159.8 MB 3.6 MB/s eta 0:00:14\n",
      "   --------------------------- ------------ 111.1/159.8 MB 3.6 MB/s eta 0:00:14\n",
      "   ---------------------------- ----------- 111.9/159.8 MB 3.6 MB/s eta 0:00:14\n",
      "   ---------------------------- ----------- 112.7/159.8 MB 3.6 MB/s eta 0:00:13\n",
      "   ---------------------------- ----------- 113.5/159.8 MB 3.6 MB/s eta 0:00:13\n",
      "   ---------------------------- ----------- 114.6/159.8 MB 3.6 MB/s eta 0:00:13\n",
      "   ---------------------------- ----------- 115.3/159.8 MB 3.6 MB/s eta 0:00:13\n",
      "   ----------------------------- ---------- 116.1/159.8 MB 3.6 MB/s eta 0:00:13\n",
      "   ----------------------------- ---------- 116.9/159.8 MB 3.6 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 117.7/159.8 MB 3.6 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 118.5/159.8 MB 3.6 MB/s eta 0:00:12\n",
      "   ----------------------------- ---------- 119.3/159.8 MB 3.6 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 120.1/159.8 MB 3.6 MB/s eta 0:00:11\n",
      "   ------------------------------ --------- 120.8/159.8 MB 3.6 MB/s eta 0:00:11\n",
      "   ------------------------------ --------- 121.9/159.8 MB 3.6 MB/s eta 0:00:11\n",
      "   ------------------------------ --------- 122.7/159.8 MB 3.6 MB/s eta 0:00:11\n",
      "   ------------------------------ --------- 123.5/159.8 MB 3.6 MB/s eta 0:00:11\n",
      "   ------------------------------- -------- 124.3/159.8 MB 3.6 MB/s eta 0:00:10\n",
      "   ------------------------------- -------- 125.0/159.8 MB 3.6 MB/s eta 0:00:10\n",
      "   ------------------------------- -------- 125.8/159.8 MB 3.6 MB/s eta 0:00:10\n",
      "   ------------------------------- -------- 126.6/159.8 MB 3.6 MB/s eta 0:00:10\n",
      "   ------------------------------- -------- 127.4/159.8 MB 3.6 MB/s eta 0:00:09\n",
      "   -------------------------------- ------- 128.2/159.8 MB 3.6 MB/s eta 0:00:09\n",
      "   -------------------------------- ------- 129.0/159.8 MB 3.6 MB/s eta 0:00:09\n",
      "   -------------------------------- ------- 129.8/159.8 MB 3.6 MB/s eta 0:00:09\n",
      "   -------------------------------- ------- 130.5/159.8 MB 3.6 MB/s eta 0:00:09\n",
      "   -------------------------------- ------- 131.3/159.8 MB 3.6 MB/s eta 0:00:08\n",
      "   --------------------------------- ------ 132.1/159.8 MB 3.6 MB/s eta 0:00:08\n",
      "   --------------------------------- ------ 132.9/159.8 MB 3.6 MB/s eta 0:00:08\n",
      "   --------------------------------- ------ 133.7/159.8 MB 3.6 MB/s eta 0:00:08\n",
      "   --------------------------------- ------ 134.5/159.8 MB 3.6 MB/s eta 0:00:07\n",
      "   --------------------------------- ------ 135.3/159.8 MB 3.6 MB/s eta 0:00:07\n",
      "   ---------------------------------- ----- 136.3/159.8 MB 3.6 MB/s eta 0:00:07\n",
      "   ---------------------------------- ----- 137.1/159.8 MB 3.6 MB/s eta 0:00:07\n",
      "   ---------------------------------- ----- 137.9/159.8 MB 3.6 MB/s eta 0:00:07\n",
      "   ---------------------------------- ----- 138.7/159.8 MB 3.6 MB/s eta 0:00:06\n",
      "   ---------------------------------- ----- 139.5/159.8 MB 3.6 MB/s eta 0:00:06\n",
      "   ----------------------------------- ---- 140.2/159.8 MB 3.6 MB/s eta 0:00:06\n",
      "   ----------------------------------- ---- 141.0/159.8 MB 3.6 MB/s eta 0:00:06\n",
      "   ----------------------------------- ---- 141.8/159.8 MB 3.6 MB/s eta 0:00:05\n",
      "   ----------------------------------- ---- 142.9/159.8 MB 3.6 MB/s eta 0:00:05\n",
      "   ----------------------------------- ---- 143.7/159.8 MB 3.6 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 144.4/159.8 MB 3.6 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 145.2/159.8 MB 3.6 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 146.0/159.8 MB 3.6 MB/s eta 0:00:04\n",
      "   ------------------------------------ --- 146.8/159.8 MB 3.6 MB/s eta 0:00:04\n",
      "   ------------------------------------ --- 147.6/159.8 MB 3.6 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 148.4/159.8 MB 3.6 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 149.2/159.8 MB 3.6 MB/s eta 0:00:03\n",
      "   ------------------------------------- -- 149.9/159.8 MB 3.6 MB/s eta 0:00:03\n",
      "   ------------------------------------- -- 150.7/159.8 MB 3.6 MB/s eta 0:00:03\n",
      "   ------------------------------------- -- 151.8/159.8 MB 3.6 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 152.6/159.8 MB 3.6 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 153.4/159.8 MB 3.6 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 154.1/159.8 MB 3.6 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 154.9/159.8 MB 3.6 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 155.7/159.8 MB 3.6 MB/s eta 0:00:02\n",
      "   ---------------------------------------  156.5/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  157.3/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  157.8/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  158.9/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  159.6/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  159.6/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------  159.6/159.8 MB 3.6 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 159.8/159.8 MB 3.6 MB/s eta 0:00:00\n",
      "Downloading torchvision-0.18.0-cp311-cp311-win_amd64.whl (1.2 MB)\n",
      "   ---------------------------------------- 0.0/1.2 MB ? eta -:--:--\n",
      "   ----------------------------------- ---- 1.0/1.2 MB 6.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 1.2/1.2 MB 4.9 MB/s eta 0:00:00\n",
      "Downloading torchaudio-2.3.0-cp311-cp311-win_amd64.whl (2.4 MB)\n",
      "   ---------------------------------------- 0.0/2.4 MB ? eta -:--:--\n",
      "   ------------- -------------------------- 0.8/2.4 MB 5.6 MB/s eta 0:00:01\n",
      "   -------------------------- ------------- 1.6/2.4 MB 4.4 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.4/2.4 MB 4.0 MB/s eta 0:00:00\n",
      "Downloading mkl-2021.4.0-py2.py3-none-win_amd64.whl (228.5 MB)\n",
      "   ---------------------------------------- 0.0/228.5 MB ? eta -:--:--\n",
      "   ---------------------------------------- 0.8/228.5 MB 4.8 MB/s eta 0:00:48\n",
      "   ---------------------------------------- 1.6/228.5 MB 4.2 MB/s eta 0:00:55\n",
      "   ---------------------------------------- 2.6/228.5 MB 4.1 MB/s eta 0:00:56\n",
      "    --------------------------------------- 3.4/228.5 MB 4.0 MB/s eta 0:00:56\n",
      "    --------------------------------------- 4.2/228.5 MB 3.9 MB/s eta 0:00:58\n",
      "    --------------------------------------- 5.0/228.5 MB 4.0 MB/s eta 0:00:57\n",
      "   - -------------------------------------- 5.8/228.5 MB 4.0 MB/s eta 0:00:57\n",
      "   - -------------------------------------- 6.6/228.5 MB 3.9 MB/s eta 0:00:57\n",
      "   - -------------------------------------- 7.3/228.5 MB 3.9 MB/s eta 0:00:57\n",
      "   - -------------------------------------- 8.1/228.5 MB 3.9 MB/s eta 0:00:57\n",
      "   - -------------------------------------- 8.9/228.5 MB 3.9 MB/s eta 0:00:56\n",
      "   - -------------------------------------- 9.7/228.5 MB 3.9 MB/s eta 0:00:56\n",
      "   - -------------------------------------- 10.5/228.5 MB 3.9 MB/s eta 0:00:56\n",
      "   - -------------------------------------- 11.3/228.5 MB 3.9 MB/s eta 0:00:56\n",
      "   -- ------------------------------------- 11.8/228.5 MB 3.9 MB/s eta 0:00:56\n",
      "   -- ------------------------------------- 12.8/228.5 MB 3.9 MB/s eta 0:00:56\n",
      "   -- ------------------------------------- 13.6/228.5 MB 3.9 MB/s eta 0:00:55\n",
      "   -- ------------------------------------- 14.7/228.5 MB 3.9 MB/s eta 0:00:55\n",
      "   -- ------------------------------------- 15.5/228.5 MB 3.9 MB/s eta 0:00:55\n",
      "   -- ------------------------------------- 16.3/228.5 MB 3.9 MB/s eta 0:00:55\n",
      "   -- ------------------------------------- 17.0/228.5 MB 3.9 MB/s eta 0:00:55\n",
      "   --- ------------------------------------ 17.8/228.5 MB 3.9 MB/s eta 0:00:54\n",
      "   --- ------------------------------------ 18.6/228.5 MB 3.9 MB/s eta 0:00:54\n",
      "   --- ------------------------------------ 19.4/228.5 MB 3.9 MB/s eta 0:00:54\n",
      "   --- ------------------------------------ 20.2/228.5 MB 3.9 MB/s eta 0:00:54\n",
      "   --- ------------------------------------ 21.2/228.5 MB 3.9 MB/s eta 0:00:54\n",
      "   --- ------------------------------------ 21.8/228.5 MB 3.9 MB/s eta 0:00:54\n",
      "   --- ------------------------------------ 22.8/228.5 MB 3.9 MB/s eta 0:00:53\n",
      "   ---- ----------------------------------- 23.6/228.5 MB 3.9 MB/s eta 0:00:53\n",
      "   ---- ----------------------------------- 24.4/228.5 MB 3.9 MB/s eta 0:00:53\n",
      "   ---- ----------------------------------- 25.2/228.5 MB 3.9 MB/s eta 0:00:53\n",
      "   ---- ----------------------------------- 26.0/228.5 MB 3.9 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 26.7/228.5 MB 3.9 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 27.5/228.5 MB 3.9 MB/s eta 0:00:52\n",
      "   ---- ----------------------------------- 28.3/228.5 MB 3.9 MB/s eta 0:00:52\n",
      "   ----- ---------------------------------- 29.1/228.5 MB 3.9 MB/s eta 0:00:52\n",
      "   ----- ---------------------------------- 29.9/228.5 MB 3.9 MB/s eta 0:00:52\n",
      "   ----- ---------------------------------- 30.9/228.5 MB 3.9 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 31.7/228.5 MB 3.9 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 32.5/228.5 MB 3.9 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 33.3/228.5 MB 3.9 MB/s eta 0:00:51\n",
      "   ----- ---------------------------------- 34.1/228.5 MB 3.9 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 35.1/228.5 MB 3.9 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 35.9/228.5 MB 3.9 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 36.7/228.5 MB 3.9 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 37.5/228.5 MB 3.9 MB/s eta 0:00:50\n",
      "   ------ --------------------------------- 38.3/228.5 MB 3.9 MB/s eta 0:00:49\n",
      "   ------ --------------------------------- 39.1/228.5 MB 3.9 MB/s eta 0:00:49\n",
      "   ------ --------------------------------- 39.8/228.5 MB 3.9 MB/s eta 0:00:49\n",
      "   ------- -------------------------------- 40.6/228.5 MB 3.9 MB/s eta 0:00:49\n",
      "   ------- -------------------------------- 41.7/228.5 MB 3.9 MB/s eta 0:00:49\n",
      "   ------- -------------------------------- 42.5/228.5 MB 3.9 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 43.3/228.5 MB 3.9 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 44.0/228.5 MB 3.9 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 44.8/228.5 MB 3.9 MB/s eta 0:00:48\n",
      "   ------- -------------------------------- 45.6/228.5 MB 3.9 MB/s eta 0:00:48\n",
      "   -------- ------------------------------- 46.4/228.5 MB 3.9 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 47.2/228.5 MB 3.9 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 48.0/228.5 MB 3.9 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 48.8/228.5 MB 3.9 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 49.5/228.5 MB 3.9 MB/s eta 0:00:47\n",
      "   -------- ------------------------------- 50.6/228.5 MB 3.9 MB/s eta 0:00:46\n",
      "   -------- ------------------------------- 51.4/228.5 MB 3.9 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 51.9/228.5 MB 3.9 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 53.0/228.5 MB 3.9 MB/s eta 0:00:46\n",
      "   --------- ------------------------------ 53.7/228.5 MB 3.9 MB/s eta 0:00:45\n",
      "   --------- ------------------------------ 54.5/228.5 MB 3.9 MB/s eta 0:00:45\n",
      "   --------- ------------------------------ 55.3/228.5 MB 3.9 MB/s eta 0:00:45\n",
      "   --------- ------------------------------ 56.1/228.5 MB 3.9 MB/s eta 0:00:45\n",
      "   --------- ------------------------------ 56.9/228.5 MB 3.9 MB/s eta 0:00:45\n",
      "   ---------- ----------------------------- 57.7/228.5 MB 3.9 MB/s eta 0:00:44\n",
      "   ---------- ----------------------------- 58.7/228.5 MB 3.9 MB/s eta 0:00:44\n",
      "   ---------- ----------------------------- 59.5/228.5 MB 3.9 MB/s eta 0:00:44\n",
      "   ---------- ----------------------------- 60.3/228.5 MB 3.9 MB/s eta 0:00:44\n",
      "   ---------- ----------------------------- 61.1/228.5 MB 3.9 MB/s eta 0:00:44\n",
      "   ---------- ----------------------------- 61.9/228.5 MB 3.9 MB/s eta 0:00:43\n",
      "   ---------- ----------------------------- 62.7/228.5 MB 3.9 MB/s eta 0:00:43\n",
      "   ----------- ---------------------------- 63.4/228.5 MB 3.9 MB/s eta 0:00:43\n",
      "   ----------- ---------------------------- 64.2/228.5 MB 3.9 MB/s eta 0:00:43\n",
      "   ----------- ---------------------------- 65.0/228.5 MB 3.9 MB/s eta 0:00:43\n",
      "   ----------- ---------------------------- 65.8/228.5 MB 3.9 MB/s eta 0:00:42\n",
      "   ----------- ---------------------------- 66.6/228.5 MB 3.9 MB/s eta 0:00:42\n",
      "   ----------- ---------------------------- 67.4/228.5 MB 3.9 MB/s eta 0:00:42\n",
      "   ----------- ---------------------------- 68.2/228.5 MB 3.9 MB/s eta 0:00:42\n",
      "   ------------ --------------------------- 68.9/228.5 MB 3.9 MB/s eta 0:00:42\n",
      "   ------------ --------------------------- 70.0/228.5 MB 3.9 MB/s eta 0:00:41\n",
      "   ------------ --------------------------- 70.8/228.5 MB 3.9 MB/s eta 0:00:41\n",
      "   ------------ --------------------------- 71.6/228.5 MB 3.9 MB/s eta 0:00:41\n",
      "   ------------ --------------------------- 72.4/228.5 MB 3.9 MB/s eta 0:00:41\n",
      "   ------------ --------------------------- 73.1/228.5 MB 3.9 MB/s eta 0:00:40\n",
      "   ------------ --------------------------- 73.9/228.5 MB 3.9 MB/s eta 0:00:40\n",
      "   ------------- -------------------------- 75.0/228.5 MB 3.9 MB/s eta 0:00:40\n",
      "   ------------- -------------------------- 75.8/228.5 MB 3.9 MB/s eta 0:00:40\n",
      "   ------------- -------------------------- 76.8/228.5 MB 3.9 MB/s eta 0:00:40\n",
      "   ------------- -------------------------- 77.6/228.5 MB 3.9 MB/s eta 0:00:39\n",
      "   ------------- -------------------------- 78.4/228.5 MB 3.9 MB/s eta 0:00:39\n",
      "   ------------- -------------------------- 79.2/228.5 MB 3.9 MB/s eta 0:00:39\n",
      "   -------------- ------------------------- 80.2/228.5 MB 3.9 MB/s eta 0:00:39\n",
      "   -------------- ------------------------- 81.0/228.5 MB 3.9 MB/s eta 0:00:38\n",
      "   -------------- ------------------------- 81.8/228.5 MB 3.9 MB/s eta 0:00:38\n",
      "   -------------- ------------------------- 82.6/228.5 MB 3.9 MB/s eta 0:00:38\n",
      "   -------------- ------------------------- 83.4/228.5 MB 3.9 MB/s eta 0:00:38\n",
      "   -------------- ------------------------- 84.1/228.5 MB 3.9 MB/s eta 0:00:38\n",
      "   -------------- ------------------------- 84.9/228.5 MB 3.9 MB/s eta 0:00:37\n",
      "   --------------- ------------------------ 85.7/228.5 MB 3.9 MB/s eta 0:00:37\n",
      "   --------------- ------------------------ 86.8/228.5 MB 3.9 MB/s eta 0:00:37\n",
      "   --------------- ------------------------ 87.6/228.5 MB 3.9 MB/s eta 0:00:37\n",
      "   --------------- ------------------------ 88.3/228.5 MB 3.9 MB/s eta 0:00:37\n",
      "   --------------- ------------------------ 89.1/228.5 MB 3.9 MB/s eta 0:00:36\n",
      "   --------------- ------------------------ 89.9/228.5 MB 3.9 MB/s eta 0:00:36\n",
      "   --------------- ------------------------ 90.7/228.5 MB 3.9 MB/s eta 0:00:36\n",
      "   ---------------- ----------------------- 91.5/228.5 MB 3.9 MB/s eta 0:00:36\n",
      "   ---------------- ----------------------- 92.5/228.5 MB 3.9 MB/s eta 0:00:35\n",
      "   ---------------- ----------------------- 93.3/228.5 MB 3.9 MB/s eta 0:00:35\n",
      "   ---------------- ----------------------- 93.8/228.5 MB 3.9 MB/s eta 0:00:35\n",
      "   ---------------- ----------------------- 94.9/228.5 MB 3.9 MB/s eta 0:00:35\n",
      "   ---------------- ----------------------- 95.7/228.5 MB 3.9 MB/s eta 0:00:35\n",
      "   ---------------- ----------------------- 96.5/228.5 MB 3.9 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 97.3/228.5 MB 3.9 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 98.0/228.5 MB 3.9 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 98.8/228.5 MB 3.9 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 99.9/228.5 MB 3.9 MB/s eta 0:00:34\n",
      "   ----------------- ---------------------- 100.7/228.5 MB 3.9 MB/s eta 0:00:33\n",
      "   ----------------- ---------------------- 101.4/228.5 MB 3.9 MB/s eta 0:00:33\n",
      "   ----------------- ---------------------- 102.2/228.5 MB 3.9 MB/s eta 0:00:33\n",
      "   ------------------ --------------------- 103.0/228.5 MB 3.9 MB/s eta 0:00:33\n",
      "   ------------------ --------------------- 104.1/228.5 MB 3.9 MB/s eta 0:00:33\n",
      "   ------------------ --------------------- 104.6/228.5 MB 3.9 MB/s eta 0:00:32\n",
      "   ------------------ --------------------- 105.4/228.5 MB 3.9 MB/s eta 0:00:32\n",
      "   ------------------ --------------------- 106.4/228.5 MB 3.9 MB/s eta 0:00:32\n",
      "   ------------------ --------------------- 107.2/228.5 MB 3.9 MB/s eta 0:00:32\n",
      "   ------------------ --------------------- 108.0/228.5 MB 3.9 MB/s eta 0:00:32\n",
      "   ------------------- -------------------- 108.8/228.5 MB 3.9 MB/s eta 0:00:31\n",
      "   ------------------- -------------------- 109.6/228.5 MB 3.9 MB/s eta 0:00:31\n",
      "   ------------------- -------------------- 110.4/228.5 MB 3.9 MB/s eta 0:00:31\n",
      "   ------------------- -------------------- 111.1/228.5 MB 3.9 MB/s eta 0:00:31\n",
      "   ------------------- -------------------- 111.9/228.5 MB 3.9 MB/s eta 0:00:31\n",
      "   ------------------- -------------------- 112.7/228.5 MB 3.9 MB/s eta 0:00:30\n",
      "   ------------------- -------------------- 113.5/228.5 MB 3.9 MB/s eta 0:00:30\n",
      "   -------------------- ------------------- 114.3/228.5 MB 3.9 MB/s eta 0:00:30\n",
      "   -------------------- ------------------- 115.1/228.5 MB 3.9 MB/s eta 0:00:30\n",
      "   -------------------- ------------------- 115.9/228.5 MB 3.9 MB/s eta 0:00:30\n",
      "   -------------------- ------------------- 116.7/228.5 MB 3.9 MB/s eta 0:00:29\n",
      "   -------------------- ------------------- 117.7/228.5 MB 3.9 MB/s eta 0:00:29\n",
      "   -------------------- ------------------- 118.5/228.5 MB 3.9 MB/s eta 0:00:29\n",
      "   -------------------- ------------------- 119.3/228.5 MB 3.9 MB/s eta 0:00:29\n",
      "   --------------------- ------------------ 120.1/228.5 MB 3.9 MB/s eta 0:00:28\n",
      "   --------------------- ------------------ 120.8/228.5 MB 3.9 MB/s eta 0:00:28\n",
      "   --------------------- ------------------ 121.6/228.5 MB 3.9 MB/s eta 0:00:28\n",
      "   --------------------- ------------------ 122.4/228.5 MB 3.9 MB/s eta 0:00:28\n",
      "   --------------------- ------------------ 123.2/228.5 MB 3.9 MB/s eta 0:00:28\n",
      "   --------------------- ------------------ 124.0/228.5 MB 3.9 MB/s eta 0:00:27\n",
      "   --------------------- ------------------ 125.0/228.5 MB 3.9 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 125.8/228.5 MB 3.9 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 126.6/228.5 MB 3.9 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 127.4/228.5 MB 3.9 MB/s eta 0:00:27\n",
      "   ---------------------- ----------------- 128.2/228.5 MB 3.9 MB/s eta 0:00:26\n",
      "   ---------------------- ----------------- 129.0/228.5 MB 3.9 MB/s eta 0:00:26\n",
      "   ---------------------- ----------------- 129.8/228.5 MB 3.9 MB/s eta 0:00:26\n",
      "   ---------------------- ----------------- 130.5/228.5 MB 3.9 MB/s eta 0:00:26\n",
      "   ---------------------- ----------------- 131.3/228.5 MB 3.9 MB/s eta 0:00:26\n",
      "   ----------------------- ---------------- 132.1/228.5 MB 3.9 MB/s eta 0:00:25\n",
      "   ----------------------- ---------------- 132.9/228.5 MB 3.9 MB/s eta 0:00:25\n",
      "   ----------------------- ---------------- 134.0/228.5 MB 3.9 MB/s eta 0:00:25\n",
      "   ----------------------- ---------------- 134.7/228.5 MB 3.9 MB/s eta 0:00:25\n",
      "   ----------------------- ---------------- 135.5/228.5 MB 3.9 MB/s eta 0:00:24\n",
      "   ----------------------- ---------------- 136.3/228.5 MB 3.9 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 137.1/228.5 MB 3.9 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 137.9/228.5 MB 3.9 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 138.7/228.5 MB 3.9 MB/s eta 0:00:24\n",
      "   ------------------------ --------------- 139.5/228.5 MB 3.9 MB/s eta 0:00:23\n",
      "   ------------------------ --------------- 140.5/228.5 MB 3.9 MB/s eta 0:00:23\n",
      "   ------------------------ --------------- 141.3/228.5 MB 3.9 MB/s eta 0:00:23\n",
      "   ------------------------ --------------- 142.1/228.5 MB 3.9 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 142.9/228.5 MB 3.9 MB/s eta 0:00:23\n",
      "   ------------------------- -------------- 143.7/228.5 MB 3.9 MB/s eta 0:00:22\n",
      "   ------------------------- -------------- 144.4/228.5 MB 3.9 MB/s eta 0:00:22\n",
      "   ------------------------- -------------- 145.2/228.5 MB 3.9 MB/s eta 0:00:22\n",
      "   ------------------------- -------------- 146.0/228.5 MB 3.9 MB/s eta 0:00:22\n",
      "   ------------------------- -------------- 146.8/228.5 MB 3.9 MB/s eta 0:00:22\n",
      "   ------------------------- -------------- 147.6/228.5 MB 3.9 MB/s eta 0:00:21\n",
      "   ------------------------- -------------- 148.4/228.5 MB 3.9 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 149.4/228.5 MB 3.9 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 150.2/228.5 MB 3.9 MB/s eta 0:00:21\n",
      "   -------------------------- ------------- 151.0/228.5 MB 3.9 MB/s eta 0:00:20\n",
      "   -------------------------- ------------- 151.8/228.5 MB 3.9 MB/s eta 0:00:20\n",
      "   -------------------------- ------------- 152.6/228.5 MB 3.9 MB/s eta 0:00:20\n",
      "   -------------------------- ------------- 153.4/228.5 MB 3.9 MB/s eta 0:00:20\n",
      "   --------------------------- ------------ 154.4/228.5 MB 3.9 MB/s eta 0:00:20\n",
      "   --------------------------- ------------ 155.2/228.5 MB 3.9 MB/s eta 0:00:19\n",
      "   --------------------------- ------------ 156.0/228.5 MB 3.9 MB/s eta 0:00:19\n",
      "   --------------------------- ------------ 156.8/228.5 MB 3.9 MB/s eta 0:00:19\n",
      "   --------------------------- ------------ 157.5/228.5 MB 3.9 MB/s eta 0:00:19\n",
      "   --------------------------- ------------ 158.3/228.5 MB 3.9 MB/s eta 0:00:19\n",
      "   --------------------------- ------------ 159.1/228.5 MB 3.9 MB/s eta 0:00:18\n",
      "   --------------------------- ------------ 159.9/228.5 MB 3.9 MB/s eta 0:00:18\n",
      "   ---------------------------- ----------- 160.7/228.5 MB 3.9 MB/s eta 0:00:18\n",
      "   ---------------------------- ----------- 161.5/228.5 MB 3.9 MB/s eta 0:00:18\n",
      "   ---------------------------- ----------- 162.3/228.5 MB 3.9 MB/s eta 0:00:18\n",
      "   ---------------------------- ----------- 163.3/228.5 MB 3.9 MB/s eta 0:00:17\n",
      "   ---------------------------- ----------- 164.1/228.5 MB 3.9 MB/s eta 0:00:17\n",
      "   ---------------------------- ----------- 164.6/228.5 MB 3.9 MB/s eta 0:00:17\n",
      "   ----------------------------- ---------- 165.7/228.5 MB 3.9 MB/s eta 0:00:17\n",
      "   ----------------------------- ---------- 166.5/228.5 MB 3.9 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 167.2/228.5 MB 3.9 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 168.0/228.5 MB 3.9 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 168.8/228.5 MB 3.9 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 169.6/228.5 MB 3.9 MB/s eta 0:00:16\n",
      "   ----------------------------- ---------- 170.4/228.5 MB 3.9 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 171.4/228.5 MB 3.9 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 172.2/228.5 MB 3.9 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 173.0/228.5 MB 3.9 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 173.8/228.5 MB 3.9 MB/s eta 0:00:15\n",
      "   ------------------------------ --------- 174.6/228.5 MB 3.9 MB/s eta 0:00:14\n",
      "   ------------------------------ --------- 175.4/228.5 MB 3.9 MB/s eta 0:00:14\n",
      "   ------------------------------ --------- 176.2/228.5 MB 3.9 MB/s eta 0:00:14\n",
      "   ------------------------------ --------- 176.9/228.5 MB 3.9 MB/s eta 0:00:14\n",
      "   ------------------------------- -------- 177.7/228.5 MB 3.9 MB/s eta 0:00:14\n",
      "   ------------------------------- -------- 178.8/228.5 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 179.6/228.5 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 180.4/228.5 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 180.9/228.5 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 181.4/228.5 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------- -------- 182.7/228.5 MB 3.9 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 183.5/228.5 MB 3.9 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 184.3/228.5 MB 3.9 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 185.1/228.5 MB 3.9 MB/s eta 0:00:12\n",
      "   -------------------------------- ------- 185.9/228.5 MB 3.9 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 186.6/228.5 MB 3.9 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 187.4/228.5 MB 3.9 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 188.2/228.5 MB 3.9 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 189.3/228.5 MB 3.9 MB/s eta 0:00:11\n",
      "   --------------------------------- ------ 190.1/228.5 MB 3.9 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 190.8/228.5 MB 3.9 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 191.6/228.5 MB 3.9 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 192.4/228.5 MB 3.9 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 193.2/228.5 MB 3.9 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 194.0/228.5 MB 3.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 194.8/228.5 MB 3.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 195.6/228.5 MB 3.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 196.3/228.5 MB 3.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 197.1/228.5 MB 3.9 MB/s eta 0:00:09\n",
      "   ---------------------------------- ----- 197.9/228.5 MB 3.9 MB/s eta 0:00:08\n",
      "   ---------------------------------- ----- 198.7/228.5 MB 3.9 MB/s eta 0:00:08\n",
      "   ---------------------------------- ----- 199.8/228.5 MB 3.9 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 200.5/228.5 MB 3.9 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 201.3/228.5 MB 3.9 MB/s eta 0:00:08\n",
      "   ----------------------------------- ---- 202.1/228.5 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 202.9/228.5 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 203.7/228.5 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 204.5/228.5 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 205.3/228.5 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 206.0/228.5 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 206.8/228.5 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 207.9/228.5 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 208.7/228.5 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 209.5/228.5 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 210.2/228.5 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 211.0/228.5 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------- -- 211.8/228.5 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------- -- 212.6/228.5 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------- -- 213.4/228.5 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 214.4/228.5 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 215.2/228.5 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 216.0/228.5 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 216.8/228.5 MB 3.9 MB/s eta 0:00:04\n",
      "   -------------------------------------- - 217.6/228.5 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 218.4/228.5 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 219.2/228.5 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 219.9/228.5 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 221.0/228.5 MB 3.9 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 221.8/228.5 MB 3.9 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 222.3/228.5 MB 3.9 MB/s eta 0:00:02\n",
      "   ---------------------------------------  223.3/228.5 MB 3.9 MB/s eta 0:00:02\n",
      "   ---------------------------------------  224.1/228.5 MB 3.9 MB/s eta 0:00:02\n",
      "   ---------------------------------------  224.9/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  225.7/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  226.8/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  227.5/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  228.3/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  228.3/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  228.3/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  228.3/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  228.3/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  228.3/228.5 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 228.5/228.5 MB 3.7 MB/s eta 0:00:00\n",
      "Downloading intel_openmp-2021.4.0-py2.py3-none-win_amd64.whl (3.5 MB)\n",
      "   ---------------------------------------- 0.0/3.5 MB ? eta -:--:--\n",
      "   ----------- ---------------------------- 1.0/3.5 MB 6.3 MB/s eta 0:00:01\n",
      "   -------------------- ------------------- 1.8/3.5 MB 4.6 MB/s eta 0:00:01\n",
      "   ----------------------------- ---------- 2.6/3.5 MB 4.4 MB/s eta 0:00:01\n",
      "   -------------------------------------- - 3.4/3.5 MB 4.3 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 3.5/3.5 MB 4.0 MB/s eta 0:00:00\n",
      "Downloading tbb-2021.13.1-py3-none-win_amd64.whl (286 kB)\n",
      "Downloading pillow-10.4.0-cp311-cp311-win_amd64.whl (2.6 MB)\n",
      "   ---------------------------------------- 0.0/2.6 MB ? eta -:--:--\n",
      "   ------------ --------------------------- 0.8/2.6 MB 4.2 MB/s eta 0:00:01\n",
      "   ------------------------ --------------- 1.6/2.6 MB 4.0 MB/s eta 0:00:01\n",
      "   ------------------------------------ --- 2.4/2.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 2.6/2.6 MB 3.7 MB/s eta 0:00:00\n",
      "Installing collected packages: tbb, intel-openmp, pillow, mkl, torch, torchvision, torchaudio\n",
      "  Attempting uninstall: torch\n",
      "    Found existing installation: torch 2.4.0\n",
      "    Uninstalling torch-2.4.0:\n",
      "      Successfully uninstalled torch-2.4.0\n",
      "Successfully installed intel-openmp-2021.4.0 mkl-2021.4.0 pillow-10.4.0 tbb-2021.13.1 torch-2.3.0 torchaudio-2.3.0 torchvision-0.18.0\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install torch==2.3.0 torchvision==0.18.0 torchaudio==2.3.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting torch==2.2.1\n",
      "  Downloading torch-2.2.1-cp311-cp311-win_amd64.whl.metadata (26 kB)\n",
      "Requirement already satisfied: filelock in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.2.1) (3.15.4)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.2.1) (4.12.2)\n",
      "Requirement already satisfied: sympy in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.2.1) (1.13.2)\n",
      "Requirement already satisfied: networkx in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.2.1) (3.3)\n",
      "Requirement already satisfied: jinja2 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.2.1) (3.1.4)\n",
      "Requirement already satisfied: fsspec in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from torch==2.2.1) (2024.6.1)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from jinja2->torch==2.2.1) (2.1.5)\n",
      "Requirement already satisfied: mpmath<1.4,>=1.1.0 in c:\\users\\akank\\desktop\\rag-model\\venv\\lib\\site-packages (from sympy->torch==2.2.1) (1.3.0)\n",
      "Downloading torch-2.2.1-cp311-cp311-win_amd64.whl (198.6 MB)\n",
      "   ---------------------------------------- 0.0/198.6 MB ? eta -:--:--\n",
      "   ---------------------------------------- 1.0/198.6 MB 6.3 MB/s eta 0:00:32\n",
      "   ---------------------------------------- 1.8/198.6 MB 4.0 MB/s eta 0:00:49\n",
      "    --------------------------------------- 2.6/198.6 MB 4.0 MB/s eta 0:00:50\n",
      "    --------------------------------------- 3.1/198.6 MB 3.6 MB/s eta 0:00:56\n",
      "    --------------------------------------- 3.9/198.6 MB 3.6 MB/s eta 0:00:54\n",
      "    --------------------------------------- 4.5/198.6 MB 3.5 MB/s eta 0:00:55\n",
      "   - -------------------------------------- 5.5/198.6 MB 3.7 MB/s eta 0:00:53\n",
      "   - -------------------------------------- 6.3/198.6 MB 3.7 MB/s eta 0:00:52\n",
      "   - -------------------------------------- 7.1/198.6 MB 3.7 MB/s eta 0:00:52\n",
      "   - -------------------------------------- 7.9/198.6 MB 3.7 MB/s eta 0:00:52\n",
      "   - -------------------------------------- 8.7/198.6 MB 3.7 MB/s eta 0:00:51\n",
      "   - -------------------------------------- 9.7/198.6 MB 3.8 MB/s eta 0:00:51\n",
      "   -- ------------------------------------- 10.5/198.6 MB 3.8 MB/s eta 0:00:51\n",
      "   -- ------------------------------------- 11.3/198.6 MB 3.8 MB/s eta 0:00:50\n",
      "   -- ------------------------------------- 12.1/198.6 MB 3.8 MB/s eta 0:00:50\n",
      "   -- ------------------------------------- 12.8/198.6 MB 3.8 MB/s eta 0:00:50\n",
      "   -- ------------------------------------- 13.6/198.6 MB 3.8 MB/s eta 0:00:50\n",
      "   -- ------------------------------------- 14.4/198.6 MB 3.8 MB/s eta 0:00:49\n",
      "   --- ------------------------------------ 15.2/198.6 MB 3.8 MB/s eta 0:00:49\n",
      "   --- ------------------------------------ 16.3/198.6 MB 3.8 MB/s eta 0:00:49\n",
      "   --- ------------------------------------ 17.0/198.6 MB 3.8 MB/s eta 0:00:48\n",
      "   --- ------------------------------------ 17.8/198.6 MB 3.8 MB/s eta 0:00:48\n",
      "   --- ------------------------------------ 18.6/198.6 MB 3.8 MB/s eta 0:00:48\n",
      "   --- ------------------------------------ 19.4/198.6 MB 3.8 MB/s eta 0:00:48\n",
      "   ---- ----------------------------------- 20.2/198.6 MB 3.8 MB/s eta 0:00:47\n",
      "   ---- ----------------------------------- 21.0/198.6 MB 3.8 MB/s eta 0:00:47\n",
      "   ---- ----------------------------------- 21.8/198.6 MB 3.8 MB/s eta 0:00:47\n",
      "   ---- ----------------------------------- 22.5/198.6 MB 3.8 MB/s eta 0:00:47\n",
      "   ---- ----------------------------------- 23.6/198.6 MB 3.8 MB/s eta 0:00:46\n",
      "   ---- ----------------------------------- 24.4/198.6 MB 3.8 MB/s eta 0:00:46\n",
      "   ----- ---------------------------------- 25.2/198.6 MB 3.8 MB/s eta 0:00:46\n",
      "   ----- ---------------------------------- 26.0/198.6 MB 3.8 MB/s eta 0:00:46\n",
      "   ----- ---------------------------------- 26.7/198.6 MB 3.8 MB/s eta 0:00:45\n",
      "   ----- ---------------------------------- 27.5/198.6 MB 3.8 MB/s eta 0:00:45\n",
      "   ----- ---------------------------------- 28.3/198.6 MB 3.8 MB/s eta 0:00:45\n",
      "   ----- ---------------------------------- 29.1/198.6 MB 3.8 MB/s eta 0:00:45\n",
      "   ------ --------------------------------- 29.9/198.6 MB 3.8 MB/s eta 0:00:45\n",
      "   ------ --------------------------------- 30.7/198.6 MB 3.8 MB/s eta 0:00:44\n",
      "   ------ --------------------------------- 31.5/198.6 MB 3.8 MB/s eta 0:00:44\n",
      "   ------ --------------------------------- 32.2/198.6 MB 3.8 MB/s eta 0:00:44\n",
      "   ------ --------------------------------- 33.0/198.6 MB 3.8 MB/s eta 0:00:44\n",
      "   ------ --------------------------------- 34.1/198.6 MB 3.8 MB/s eta 0:00:43\n",
      "   ------- -------------------------------- 34.9/198.6 MB 3.8 MB/s eta 0:00:43\n",
      "   ------- -------------------------------- 35.7/198.6 MB 3.8 MB/s eta 0:00:43\n",
      "   ------- -------------------------------- 36.4/198.6 MB 3.8 MB/s eta 0:00:43\n",
      "   ------- -------------------------------- 37.2/198.6 MB 3.8 MB/s eta 0:00:43\n",
      "   ------- -------------------------------- 38.0/198.6 MB 3.8 MB/s eta 0:00:42\n",
      "   ------- -------------------------------- 38.8/198.6 MB 3.8 MB/s eta 0:00:42\n",
      "   ------- -------------------------------- 39.6/198.6 MB 3.8 MB/s eta 0:00:42\n",
      "   -------- ------------------------------- 40.4/198.6 MB 3.8 MB/s eta 0:00:42\n",
      "   -------- ------------------------------- 41.2/198.6 MB 3.8 MB/s eta 0:00:42\n",
      "   -------- ------------------------------- 41.7/198.6 MB 3.8 MB/s eta 0:00:41\n",
      "   -------- ------------------------------- 42.7/198.6 MB 3.8 MB/s eta 0:00:41\n",
      "   -------- ------------------------------- 43.5/198.6 MB 3.8 MB/s eta 0:00:41\n",
      "   -------- ------------------------------- 44.3/198.6 MB 3.8 MB/s eta 0:00:41\n",
      "   --------- ------------------------------ 45.1/198.6 MB 3.8 MB/s eta 0:00:40\n",
      "   --------- ------------------------------ 45.9/198.6 MB 3.8 MB/s eta 0:00:40\n",
      "   --------- ------------------------------ 46.7/198.6 MB 3.8 MB/s eta 0:00:40\n",
      "   --------- ------------------------------ 47.4/198.6 MB 3.8 MB/s eta 0:00:40\n",
      "   --------- ------------------------------ 48.2/198.6 MB 3.8 MB/s eta 0:00:40\n",
      "   --------- ------------------------------ 49.0/198.6 MB 3.8 MB/s eta 0:00:39\n",
      "   ---------- ----------------------------- 49.8/198.6 MB 3.8 MB/s eta 0:00:39\n",
      "   ---------- ----------------------------- 50.6/198.6 MB 3.8 MB/s eta 0:00:39\n",
      "   ---------- ----------------------------- 51.4/198.6 MB 3.8 MB/s eta 0:00:39\n",
      "   ---------- ----------------------------- 52.2/198.6 MB 3.8 MB/s eta 0:00:39\n",
      "   ---------- ----------------------------- 53.0/198.6 MB 3.8 MB/s eta 0:00:38\n",
      "   ---------- ----------------------------- 53.7/198.6 MB 3.8 MB/s eta 0:00:38\n",
      "   ----------- ---------------------------- 54.8/198.6 MB 3.8 MB/s eta 0:00:38\n",
      "   ----------- ---------------------------- 55.6/198.6 MB 3.8 MB/s eta 0:00:38\n",
      "   ----------- ---------------------------- 56.4/198.6 MB 3.8 MB/s eta 0:00:37\n",
      "   ----------- ---------------------------- 57.1/198.6 MB 3.8 MB/s eta 0:00:37\n",
      "   ----------- ---------------------------- 57.9/198.6 MB 3.8 MB/s eta 0:00:37\n",
      "   ----------- ---------------------------- 58.7/198.6 MB 3.8 MB/s eta 0:00:37\n",
      "   ----------- ---------------------------- 59.5/198.6 MB 3.8 MB/s eta 0:00:37\n",
      "   ------------ --------------------------- 60.3/198.6 MB 3.8 MB/s eta 0:00:36\n",
      "   ------------ --------------------------- 61.1/198.6 MB 3.9 MB/s eta 0:00:36\n",
      "   ------------ --------------------------- 61.9/198.6 MB 3.9 MB/s eta 0:00:36\n",
      "   ------------ --------------------------- 62.7/198.6 MB 3.9 MB/s eta 0:00:36\n",
      "   ------------ --------------------------- 63.4/198.6 MB 3.8 MB/s eta 0:00:36\n",
      "   ------------ --------------------------- 64.2/198.6 MB 3.8 MB/s eta 0:00:35\n",
      "   ------------- -------------------------- 65.0/198.6 MB 3.8 MB/s eta 0:00:35\n",
      "   ------------- -------------------------- 65.8/198.6 MB 3.8 MB/s eta 0:00:35\n",
      "   ------------- -------------------------- 66.6/198.6 MB 3.8 MB/s eta 0:00:35\n",
      "   ------------- -------------------------- 67.4/198.6 MB 3.8 MB/s eta 0:00:35\n",
      "   ------------- -------------------------- 68.2/198.6 MB 3.8 MB/s eta 0:00:34\n",
      "   ------------- -------------------------- 68.9/198.6 MB 3.8 MB/s eta 0:00:34\n",
      "   -------------- ------------------------- 69.7/198.6 MB 3.8 MB/s eta 0:00:34\n",
      "   -------------- ------------------------- 70.5/198.6 MB 3.8 MB/s eta 0:00:34\n",
      "   -------------- ------------------------- 71.3/198.6 MB 3.8 MB/s eta 0:00:34\n",
      "   -------------- ------------------------- 72.1/198.6 MB 3.9 MB/s eta 0:00:33\n",
      "   -------------- ------------------------- 72.9/198.6 MB 3.9 MB/s eta 0:00:33\n",
      "   -------------- ------------------------- 73.7/198.6 MB 3.9 MB/s eta 0:00:33\n",
      "   -------------- ------------------------- 74.4/198.6 MB 3.9 MB/s eta 0:00:33\n",
      "   --------------- ------------------------ 75.5/198.6 MB 3.9 MB/s eta 0:00:32\n",
      "   --------------- ------------------------ 76.3/198.6 MB 3.9 MB/s eta 0:00:32\n",
      "   --------------- ------------------------ 77.1/198.6 MB 3.9 MB/s eta 0:00:32\n",
      "   --------------- ------------------------ 77.9/198.6 MB 3.9 MB/s eta 0:00:32\n",
      "   --------------- ------------------------ 78.6/198.6 MB 3.9 MB/s eta 0:00:32\n",
      "   ---------------- ----------------------- 79.4/198.6 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------------- ----------------------- 80.2/198.6 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------------- ----------------------- 81.0/198.6 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------------- ----------------------- 81.8/198.6 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------------- ----------------------- 82.6/198.6 MB 3.9 MB/s eta 0:00:31\n",
      "   ---------------- ----------------------- 83.4/198.6 MB 3.9 MB/s eta 0:00:30\n",
      "   ---------------- ----------------------- 84.1/198.6 MB 3.9 MB/s eta 0:00:30\n",
      "   ----------------- ---------------------- 84.9/198.6 MB 3.9 MB/s eta 0:00:30\n",
      "   ----------------- ---------------------- 85.7/198.6 MB 3.9 MB/s eta 0:00:30\n",
      "   ----------------- ---------------------- 86.8/198.6 MB 3.9 MB/s eta 0:00:29\n",
      "   ----------------- ---------------------- 87.6/198.6 MB 3.9 MB/s eta 0:00:29\n",
      "   ----------------- ---------------------- 88.3/198.6 MB 3.9 MB/s eta 0:00:29\n",
      "   ----------------- ---------------------- 89.1/198.6 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------------ --------------------- 89.9/198.6 MB 3.9 MB/s eta 0:00:29\n",
      "   ------------------ --------------------- 90.7/198.6 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------------ --------------------- 91.5/198.6 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------------ --------------------- 92.3/198.6 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------------ --------------------- 93.1/198.6 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------------ --------------------- 93.8/198.6 MB 3.9 MB/s eta 0:00:28\n",
      "   ------------------- -------------------- 94.6/198.6 MB 3.9 MB/s eta 0:00:27\n",
      "   ------------------- -------------------- 95.4/198.6 MB 3.9 MB/s eta 0:00:27\n",
      "   ------------------- -------------------- 96.2/198.6 MB 3.9 MB/s eta 0:00:27\n",
      "   ------------------- -------------------- 97.0/198.6 MB 3.9 MB/s eta 0:00:27\n",
      "   ------------------- -------------------- 97.8/198.6 MB 3.9 MB/s eta 0:00:27\n",
      "   ------------------- -------------------- 98.6/198.6 MB 3.9 MB/s eta 0:00:26\n",
      "   -------------------- ------------------- 99.6/198.6 MB 3.9 MB/s eta 0:00:26\n",
      "   -------------------- ------------------- 100.4/198.6 MB 3.9 MB/s eta 0:00:26\n",
      "   -------------------- ------------------- 101.2/198.6 MB 3.9 MB/s eta 0:00:26\n",
      "   -------------------- ------------------- 102.0/198.6 MB 3.9 MB/s eta 0:00:26\n",
      "   -------------------- ------------------- 102.8/198.6 MB 3.9 MB/s eta 0:00:25\n",
      "   -------------------- ------------------- 103.8/198.6 MB 3.9 MB/s eta 0:00:25\n",
      "   --------------------- ------------------ 104.6/198.6 MB 3.9 MB/s eta 0:00:25\n",
      "   --------------------- ------------------ 105.4/198.6 MB 3.9 MB/s eta 0:00:25\n",
      "   --------------------- ------------------ 106.2/198.6 MB 3.9 MB/s eta 0:00:24\n",
      "   --------------------- ------------------ 107.0/198.6 MB 3.9 MB/s eta 0:00:24\n",
      "   --------------------- ------------------ 107.7/198.6 MB 3.9 MB/s eta 0:00:24\n",
      "   --------------------- ------------------ 108.5/198.6 MB 3.9 MB/s eta 0:00:24\n",
      "   ---------------------- ----------------- 109.6/198.6 MB 3.9 MB/s eta 0:00:24\n",
      "   ---------------------- ----------------- 110.4/198.6 MB 3.9 MB/s eta 0:00:23\n",
      "   ---------------------- ----------------- 111.1/198.6 MB 3.9 MB/s eta 0:00:23\n",
      "   ---------------------- ----------------- 111.9/198.6 MB 3.9 MB/s eta 0:00:23\n",
      "   ---------------------- ----------------- 112.7/198.6 MB 3.9 MB/s eta 0:00:23\n",
      "   ---------------------- ----------------- 113.5/198.6 MB 3.9 MB/s eta 0:00:23\n",
      "   ----------------------- ---------------- 114.3/198.6 MB 3.9 MB/s eta 0:00:22\n",
      "   ----------------------- ---------------- 115.1/198.6 MB 3.9 MB/s eta 0:00:22\n",
      "   ----------------------- ---------------- 115.9/198.6 MB 3.9 MB/s eta 0:00:22\n",
      "   ----------------------- ---------------- 116.4/198.6 MB 3.9 MB/s eta 0:00:22\n",
      "   ----------------------- ---------------- 117.4/198.6 MB 3.9 MB/s eta 0:00:21\n",
      "   ----------------------- ---------------- 118.2/198.6 MB 3.9 MB/s eta 0:00:21\n",
      "   ----------------------- ---------------- 119.0/198.6 MB 3.9 MB/s eta 0:00:21\n",
      "   ------------------------ --------------- 119.8/198.6 MB 3.9 MB/s eta 0:00:21\n",
      "   ------------------------ --------------- 120.6/198.6 MB 3.9 MB/s eta 0:00:21\n",
      "   ------------------------ --------------- 121.6/198.6 MB 3.9 MB/s eta 0:00:20\n",
      "   ------------------------ --------------- 122.4/198.6 MB 3.9 MB/s eta 0:00:20\n",
      "   ------------------------ --------------- 123.2/198.6 MB 3.9 MB/s eta 0:00:20\n",
      "   ------------------------ --------------- 124.0/198.6 MB 3.9 MB/s eta 0:00:20\n",
      "   ------------------------- -------------- 124.8/198.6 MB 3.9 MB/s eta 0:00:20\n",
      "   ------------------------- -------------- 125.6/198.6 MB 3.9 MB/s eta 0:00:19\n",
      "   ------------------------- -------------- 126.4/198.6 MB 3.9 MB/s eta 0:00:19\n",
      "   ------------------------- -------------- 127.1/198.6 MB 3.9 MB/s eta 0:00:19\n",
      "   ------------------------- -------------- 127.9/198.6 MB 3.9 MB/s eta 0:00:19\n",
      "   ------------------------- -------------- 128.7/198.6 MB 3.9 MB/s eta 0:00:19\n",
      "   -------------------------- ------------- 129.5/198.6 MB 3.9 MB/s eta 0:00:18\n",
      "   -------------------------- ------------- 130.3/198.6 MB 3.9 MB/s eta 0:00:18\n",
      "   -------------------------- ------------- 131.1/198.6 MB 3.9 MB/s eta 0:00:18\n",
      "   -------------------------- ------------- 131.9/198.6 MB 3.9 MB/s eta 0:00:18\n",
      "   -------------------------- ------------- 132.9/198.6 MB 3.9 MB/s eta 0:00:17\n",
      "   -------------------------- ------------- 133.7/198.6 MB 3.9 MB/s eta 0:00:17\n",
      "   --------------------------- ------------ 134.5/198.6 MB 3.9 MB/s eta 0:00:17\n",
      "   --------------------------- ------------ 135.3/198.6 MB 3.9 MB/s eta 0:00:17\n",
      "   --------------------------- ------------ 136.1/198.6 MB 3.9 MB/s eta 0:00:17\n",
      "   --------------------------- ------------ 136.8/198.6 MB 3.9 MB/s eta 0:00:16\n",
      "   --------------------------- ------------ 137.6/198.6 MB 3.9 MB/s eta 0:00:16\n",
      "   --------------------------- ------------ 138.4/198.6 MB 3.9 MB/s eta 0:00:16\n",
      "   ---------------------------- ----------- 139.2/198.6 MB 3.9 MB/s eta 0:00:16\n",
      "   ---------------------------- ----------- 140.0/198.6 MB 3.9 MB/s eta 0:00:16\n",
      "   ---------------------------- ----------- 141.0/198.6 MB 3.9 MB/s eta 0:00:15\n",
      "   ---------------------------- ----------- 141.8/198.6 MB 3.9 MB/s eta 0:00:15\n",
      "   ---------------------------- ----------- 142.6/198.6 MB 3.9 MB/s eta 0:00:15\n",
      "   ---------------------------- ----------- 143.4/198.6 MB 3.9 MB/s eta 0:00:15\n",
      "   ----------------------------- ---------- 144.2/198.6 MB 3.9 MB/s eta 0:00:15\n",
      "   ----------------------------- ---------- 145.0/198.6 MB 3.9 MB/s eta 0:00:14\n",
      "   ----------------------------- ---------- 145.8/198.6 MB 3.9 MB/s eta 0:00:14\n",
      "   ----------------------------- ---------- 146.5/198.6 MB 3.9 MB/s eta 0:00:14\n",
      "   ----------------------------- ---------- 147.6/198.6 MB 3.9 MB/s eta 0:00:14\n",
      "   ----------------------------- ---------- 148.4/198.6 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------ --------- 149.2/198.6 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------ --------- 149.9/198.6 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------ --------- 150.7/198.6 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------ --------- 151.5/198.6 MB 3.9 MB/s eta 0:00:13\n",
      "   ------------------------------ --------- 152.3/198.6 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 153.1/198.6 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------ --------- 153.9/198.6 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 154.9/198.6 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 155.7/198.6 MB 3.9 MB/s eta 0:00:12\n",
      "   ------------------------------- -------- 156.5/198.6 MB 3.9 MB/s eta 0:00:11\n",
      "   ------------------------------- -------- 157.0/198.6 MB 3.9 MB/s eta 0:00:11\n",
      "   ------------------------------- -------- 158.1/198.6 MB 3.9 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 158.9/198.6 MB 3.9 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 159.6/198.6 MB 3.9 MB/s eta 0:00:11\n",
      "   -------------------------------- ------- 160.4/198.6 MB 3.9 MB/s eta 0:00:10\n",
      "   -------------------------------- ------- 161.2/198.6 MB 3.9 MB/s eta 0:00:10\n",
      "   -------------------------------- ------- 162.3/198.6 MB 3.9 MB/s eta 0:00:10\n",
      "   -------------------------------- ------- 163.1/198.6 MB 3.9 MB/s eta 0:00:10\n",
      "   --------------------------------- ------ 163.8/198.6 MB 3.9 MB/s eta 0:00:09\n",
      "   --------------------------------- ------ 164.6/198.6 MB 3.9 MB/s eta 0:00:09\n",
      "   --------------------------------- ------ 165.4/198.6 MB 3.9 MB/s eta 0:00:09\n",
      "   --------------------------------- ------ 166.2/198.6 MB 3.9 MB/s eta 0:00:09\n",
      "   --------------------------------- ------ 167.0/198.6 MB 3.9 MB/s eta 0:00:09\n",
      "   --------------------------------- ------ 167.8/198.6 MB 3.9 MB/s eta 0:00:08\n",
      "   --------------------------------- ------ 168.6/198.6 MB 3.9 MB/s eta 0:00:08\n",
      "   ---------------------------------- ----- 169.3/198.6 MB 3.9 MB/s eta 0:00:08\n",
      "   ---------------------------------- ----- 170.1/198.6 MB 3.9 MB/s eta 0:00:08\n",
      "   ---------------------------------- ----- 171.2/198.6 MB 3.9 MB/s eta 0:00:08\n",
      "   ---------------------------------- ----- 172.0/198.6 MB 3.9 MB/s eta 0:00:07\n",
      "   ---------------------------------- ----- 172.8/198.6 MB 3.9 MB/s eta 0:00:07\n",
      "   ---------------------------------- ----- 173.0/198.6 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 174.3/198.6 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 175.1/198.6 MB 3.9 MB/s eta 0:00:07\n",
      "   ----------------------------------- ---- 175.9/198.6 MB 3.9 MB/s eta 0:00:06\n",
      "   ----------------------------------- ---- 176.7/198.6 MB 3.9 MB/s eta 0:00:06\n",
      "   ----------------------------------- ---- 177.5/198.6 MB 3.9 MB/s eta 0:00:06\n",
      "   ----------------------------------- ---- 178.3/198.6 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 179.0/198.6 MB 3.9 MB/s eta 0:00:06\n",
      "   ------------------------------------ --- 179.8/198.6 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 180.9/198.6 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 181.7/198.6 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 182.5/198.6 MB 3.9 MB/s eta 0:00:05\n",
      "   ------------------------------------ --- 183.2/198.6 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 184.0/198.6 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 184.8/198.6 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 185.6/198.6 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 186.4/198.6 MB 3.9 MB/s eta 0:00:04\n",
      "   ------------------------------------- -- 187.2/198.6 MB 3.9 MB/s eta 0:00:03\n",
      "   ------------------------------------- -- 188.2/198.6 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 189.0/198.6 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 189.8/198.6 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 190.6/198.6 MB 3.9 MB/s eta 0:00:03\n",
      "   -------------------------------------- - 191.4/198.6 MB 3.9 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 192.2/198.6 MB 3.9 MB/s eta 0:00:02\n",
      "   -------------------------------------- - 193.2/198.6 MB 3.9 MB/s eta 0:00:02\n",
      "   ---------------------------------------  194.0/198.6 MB 3.9 MB/s eta 0:00:02\n",
      "   ---------------------------------------  194.8/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  195.6/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  196.3/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  197.1/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  197.9/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  198.4/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  198.4/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  198.4/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------  198.4/198.6 MB 3.9 MB/s eta 0:00:01\n",
      "   ---------------------------------------- 198.6/198.6 MB 3.8 MB/s eta 0:00:00\n",
      "Installing collected packages: torch\n",
      "  Attempting uninstall: torch\n",
      "    Found existing installation: torch 2.3.0\n",
      "    Uninstalling torch-2.3.0:\n",
      "      Successfully uninstalled torch-2.3.0\n",
      "Successfully installed torch-2.2.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  WARNING: Failed to remove contents in a temporary directory 'C:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\~-rch'.\n",
      "  You can safely remove it manually.\n",
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "torchaudio 2.3.0 requires torch==2.3.0, but you have torch 2.2.1 which is incompatible.\n",
      "torchvision 0.18.0 requires torch==2.3.0, but you have torch 2.2.1 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "pip install torch==2.2.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "[WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\torch\\lib\\shm.dll\" or one of its dependencies.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[40], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mlangchain_community\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mvectorstores\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DocArrayInMemorySearch\n\u001b[1;32m----> 3\u001b[0m vectorstore \u001b[38;5;241m=\u001b[39m \u001b[43mDocArrayInMemorySearch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_documents\u001b[49m\u001b[43m(\u001b[49m\n\u001b[0;32m      4\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpages\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\n\u001b[0;32m      5\u001b[0m \u001b[43m    \u001b[49m\u001b[43membedding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43membeddings\u001b[49m\n\u001b[0;32m      6\u001b[0m \u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\langchain_core\\vectorstores\\base.py:833\u001b[0m, in \u001b[0;36mVectorStore.from_documents\u001b[1;34m(cls, documents, embedding, **kwargs)\u001b[0m\n\u001b[0;32m    831\u001b[0m texts \u001b[38;5;241m=\u001b[39m [d\u001b[38;5;241m.\u001b[39mpage_content \u001b[38;5;28;01mfor\u001b[39;00m d \u001b[38;5;129;01min\u001b[39;00m documents]\n\u001b[0;32m    832\u001b[0m metadatas \u001b[38;5;241m=\u001b[39m [d\u001b[38;5;241m.\u001b[39mmetadata \u001b[38;5;28;01mfor\u001b[39;00m d \u001b[38;5;129;01min\u001b[39;00m documents]\n\u001b[1;32m--> 833\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_texts\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtexts\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43membedding\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmetadatas\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmetadatas\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\langchain_community\\vectorstores\\docarray\\in_memory.py:69\u001b[0m, in \u001b[0;36mDocArrayInMemorySearch.from_texts\u001b[1;34m(cls, texts, embedding, metadatas, **kwargs)\u001b[0m\n\u001b[0;32m     47\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[0;32m     48\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfrom_texts\u001b[39m(\n\u001b[0;32m     49\u001b[0m     \u001b[38;5;28mcls\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     53\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m     54\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DocArrayInMemorySearch:\n\u001b[0;32m     55\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Create an DocArrayInMemorySearch store and insert data.\u001b[39;00m\n\u001b[0;32m     56\u001b[0m \n\u001b[0;32m     57\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     67\u001b[0m \u001b[38;5;124;03m        DocArrayInMemorySearch Vector Store\u001b[39;00m\n\u001b[0;32m     68\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m---> 69\u001b[0m     store \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mcls\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_params\u001b[49m\u001b[43m(\u001b[49m\u001b[43membedding\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     70\u001b[0m     store\u001b[38;5;241m.\u001b[39madd_texts(texts\u001b[38;5;241m=\u001b[39mtexts, metadatas\u001b[38;5;241m=\u001b[39mmetadatas)\n\u001b[0;32m     71\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m store\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\langchain_community\\vectorstores\\docarray\\in_memory.py:40\u001b[0m, in \u001b[0;36mDocArrayInMemorySearch.from_params\u001b[1;34m(cls, embedding, metric, **kwargs)\u001b[0m\n\u001b[0;32m     22\u001b[0m \u001b[38;5;129m@classmethod\u001b[39m\n\u001b[0;32m     23\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mfrom_params\u001b[39m(\n\u001b[0;32m     24\u001b[0m     \u001b[38;5;28mcls\u001b[39m,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     29\u001b[0m     \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs: Any,\n\u001b[0;32m     30\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m DocArrayInMemorySearch:\n\u001b[0;32m     31\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Initialize DocArrayInMemorySearch store.\u001b[39;00m\n\u001b[0;32m     32\u001b[0m \n\u001b[0;32m     33\u001b[0m \u001b[38;5;124;03m    Args:\u001b[39;00m\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     38\u001b[0m \u001b[38;5;124;03m        **kwargs: Other keyword arguments to be passed to the get_doc_cls method.\u001b[39;00m\n\u001b[0;32m     39\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[1;32m---> 40\u001b[0m     \u001b[43m_check_docarray_import\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     41\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mindex\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m InMemoryExactNNIndex\n\u001b[0;32m     43\u001b[0m     doc_cls \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mcls\u001b[39m\u001b[38;5;241m.\u001b[39m_get_doc_cls(space\u001b[38;5;241m=\u001b[39mmetric, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\langchain_community\\vectorstores\\docarray\\base.py:19\u001b[0m, in \u001b[0;36m_check_docarray_import\u001b[1;34m()\u001b[0m\n\u001b[0;32m     17\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_check_docarray_import\u001b[39m() \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m     18\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m---> 19\u001b[0m         \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\n\u001b[0;32m     21\u001b[0m         da_version \u001b[38;5;241m=\u001b[39m docarray\u001b[38;5;241m.\u001b[39m__version__\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     22\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mint\u001b[39m(da_version[\u001b[38;5;241m0\u001b[39m]) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mint\u001b[39m(da_version[\u001b[38;5;241m1\u001b[39m]) \u001b[38;5;241m<\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m31\u001b[39m:\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\__init__.py:5\u001b[0m\n\u001b[0;32m      1\u001b[0m __version__ \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m0.40.0\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mlogging\u001b[39;00m\n\u001b[1;32m----> 5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01marray\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DocList, DocVec\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseDoc\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_internal\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmisc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m _get_path_from_docarray_root_level\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\array\\__init__.py:1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01marray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01many_array\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AnyDocArray\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01marray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc_list\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc_list\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DocList\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01marray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc_vec\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc_vec\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DocVec\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\array\\any_array.py:23\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[0;32m      5\u001b[0m     TYPE_CHECKING,\n\u001b[0;32m      6\u001b[0m     Any,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m     18\u001b[0m     overload,\n\u001b[0;32m     19\u001b[0m )\n\u001b[0;32m     21\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[1;32m---> 23\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseDocWithoutId\n\u001b[0;32m     24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdisplay\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdocument_array_summary\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m DocArraySummary\n\u001b[0;32m     25\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mexceptions\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mexceptions\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m UnusableObjectError\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\base_doc\\__init__.py:1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01many_doc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AnyDoc\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_node\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseNode\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseDoc\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\base_doc\\any_doc.py:5\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Type\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_internal\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpydantic\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m is_pydantic_v2\n\u001b[1;32m----> 5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdoc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseDoc\n\u001b[0;32m      8\u001b[0m \u001b[38;5;28;01mclass\u001b[39;00m \u001b[38;5;21;01mAnyDoc\u001b[39;00m(BaseDoc):\n\u001b[0;32m      9\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     10\u001b[0m \u001b[38;5;124;03m    AnyDoc is a Document that is not tied to any schema\u001b[39;00m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\base_doc\\doc.py:36\u001b[0m\n\u001b[0;32m     34\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_node\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseNode\n\u001b[0;32m     35\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mio\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mjson\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m orjson_dumps_and_decode\n\u001b[1;32m---> 36\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmixins\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m IOMixin, UpdateMixin\n\u001b[0;32m     37\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ID\n\u001b[0;32m     38\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mabstract_tensor\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AbstractTensor\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\base_doc\\mixins\\__init__.py:1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmixins\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m IOMixin\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmixins\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mupdate\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m UpdateMixin\n\u001b[0;32m      4\u001b[0m __all__ \u001b[38;5;241m=\u001b[39m [\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mIOMixin\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mUpdateMixin\u001b[39m\u001b[38;5;124m'\u001b[39m]\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\base_doc\\mixins\\io.py:25\u001b[0m\n\u001b[0;32m     22\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping_inspect\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m get_args, is_union_type\n\u001b[0;32m     24\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_doc\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_node\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseNode\n\u001b[1;32m---> 25\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m NdArray\n\u001b[0;32m     26\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mproto_register\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m _PROTO_TYPE_NAME_TO_CLASS\n\u001b[0;32m     27\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_internal\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_typing\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m safe_issubclass\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\__init__.py:3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping_extensions\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m TYPE_CHECKING\n\u001b[1;32m----> 3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbytes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioBytes, ImageBytes, VideoBytes\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mid\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ID\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ImageNdArray, ImageTensor\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\bytes\\__init__.py:1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbytes\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio_bytes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioBytes\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbytes\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mimage_bytes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ImageBytes\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbytes\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mvideo_bytes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m VideoBytes\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\bytes\\audio_bytes.py:9\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbytes\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbase_bytes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m BaseBytes\n\u001b[0;32m      8\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mproto_register\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m _register_proto\n\u001b[1;32m----> 9\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioNdArray\n\u001b[0;32m     10\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_internal\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmisc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m import_library\n\u001b[0;32m     12\u001b[0m T \u001b[38;5;241m=\u001b[39m TypeVar(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mT\u001b[39m\u001b[38;5;124m'\u001b[39m, bound\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mAudioBytes\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\tensor\\__init__.py:5\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtypes\u001b[39;00m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping_extensions\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m TYPE_CHECKING\n\u001b[1;32m----> 5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioNdArray\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01membedding\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AnyEmbedding, NdArrayEmbedding\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mimage\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m ImageNdArray, ImageTensor\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\tensor\\audio\\__init__.py:4\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtypes\u001b[39;00m\n\u001b[0;32m      2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m TYPE_CHECKING\n\u001b[1;32m----> 4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio_ndarray\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioNdArray\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio_tensor\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioTensor\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_internal\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmisc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m (\n\u001b[0;32m      7\u001b[0m     _get_path_from_docarray_root_level,\n\u001b[0;32m      8\u001b[0m     import_library,\n\u001b[0;32m      9\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\tensor\\audio\\audio_ndarray.py:2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mproto_register\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m _register_proto\n\u001b[1;32m----> 2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mabstract_audio_tensor\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AbstractAudioTensor\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mndarray\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m NdArray\n\u001b[0;32m      6\u001b[0m \u001b[38;5;129m@_register_proto\u001b[39m(proto_type_name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124maudio_ndarray\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mclass\u001b[39;00m \u001b[38;5;21;01mAudioNdArray\u001b[39;00m(AbstractAudioTensor, NdArray):\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\typing\\tensor\\audio\\abstract_audio_tensor.py:6\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m TYPE_CHECKING, Any, BinaryIO, Dict, TypeVar, Union\n\u001b[0;32m      5\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtensor\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mabstract_tensor\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AbstractTensor\n\u001b[1;32m----> 6\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutils\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01m_internal\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmisc\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m import_library, is_notebook\n\u001b[0;32m      8\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m TYPE_CHECKING:\n\u001b[0;32m      9\u001b[0m     \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mdocarray\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mtyping\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mbytes\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01maudio_bytes\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m AudioBytes\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\docarray\\utils\\_internal\\misc.py:10\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mnumpy\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mnp\u001b[39;00m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m---> 10\u001b[0m     \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtorch\u001b[39;00m  \u001b[38;5;66;03m# noqa: F401\u001b[39;00m\n\u001b[0;32m     11\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mImportError\u001b[39;00m:\n\u001b[0;32m     12\u001b[0m     torch_imported \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\torch\\__init__.py:130\u001b[0m\n\u001b[0;32m    128\u001b[0m     err \u001b[38;5;241m=\u001b[39m ctypes\u001b[38;5;241m.\u001b[39mWinError(last_error)\n\u001b[0;32m    129\u001b[0m     err\u001b[38;5;241m.\u001b[39mstrerror \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m Error loading \u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mdll\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m or one of its dependencies.\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m--> 130\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m err\n\u001b[0;32m    131\u001b[0m \u001b[38;5;28;01melif\u001b[39;00m res \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m    132\u001b[0m     is_loaded \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[1;31mOSError\u001b[0m: [WinError 127] The specified procedure could not be found. Error loading \"c:\\Users\\akank\\Desktop\\RAG-model\\venv\\Lib\\site-packages\\torch\\lib\\shm.dll\" or one of its dependencies."
     ]
    }
   ],
   "source": [
    "from langchain_community.vectorstores import DocArrayInMemorySearch\n",
    "\n",
    "vectorstore = DocArrayInMemorySearch.from_documents(\n",
    "    pages, \n",
    "    embedding=embeddings\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "retriever = vectorstore.as_retriever()\n",
    "retriever.invoke(\"Quantitative Trait Loci (QTL) mapping\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from operator import itemgetter\n",
    "\n",
    "chain = (\n",
    "    {\n",
    "        \"context\":itemgetter(\"question\") | retriever,\n",
    "        \"question\": itemgetter(\"question\") \n",
    "    }\n",
    "    | prompt \n",
    "    | model \n",
    "    | parser \n",
    ")\n",
    "chain.invoke({\"question\": \"What is the main idea of this text?\"})"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
